{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "3cWk9QFMdqBR",
        "gun_cOiMdrEb",
        "bft77Czidp7B",
        "VKNCGeWynpT5",
        "lc-B0_P9nxVV",
        "uAJ82B5tn4r_",
        "FBeDM9CAdp3l",
        "ghFm2aL_l7BE",
        "-fFsYOJOdp0R",
        "tPyX5J7AdpvH",
        "Rc4LGpNCdpnb",
        "ilSzepaDfK7l",
        "a8N8DQczfQpk",
        "OMyyRtILr1L-",
        "TD0wHp2ifbDt",
        "oCoB6-_bfa6J"
      ],
      "authorship_tag": "ABX9TyPM/1bNc+CZQEy/jkV/OS/X"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Sequence Fasta Depurator - CODE FASTA SEQUENCE CLEANER (*FSC*)**"
      ],
      "metadata": {
        "id": "3cWk9QFMdqBR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Function to process the FASTA sequence\n",
        "def process_fasta(file_path):\n",
        "    # Open the file and read its content\n",
        "    with open(file_path, 'r') as fasta_file:\n",
        "        f1 = fasta_file.read()\n",
        "\n",
        "    # Concatenate all lines except the first (header)\n",
        "    data = \"\".join(f1.split(\"\\n\")[1:])\n",
        "\n",
        "    # Filter the resulting string to keep only A, T, G, and C\n",
        "    filtered_sequence = ''.join([base for base in data if base in 'ATGC'])\n",
        "\n",
        "    return filtered_sequence\n",
        "\n",
        "# Path to the FASTA file\n",
        "file_path = 'path/to/your/file.fasta'\n",
        "\n",
        "# Process the sequence and get the filtered result\n",
        "SEQUENCE_RESULT = process_fasta(file_path)\n",
        "\n",
        "print(SEQUENCE_RESULT)\n"
      ],
      "metadata": {
        "id": "D_K2CvlJhpdd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Relative-Frequency-Sequence - CODE Nucleotide Relative  Frequency Visualizer (*NRFV*)**"
      ],
      "metadata": {
        "id": "gun_cOiMdrEb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import matplotlib.pyplot as plt\n",
        "\n",
        "def calculate_base_content(dna_sequence):\n",
        "    # Calculate the length of the DNA sequence\n",
        "    length = len(dna_sequence)\n",
        "\n",
        "    # Count the occurrences of each base\n",
        "    a_count = dna_sequence.count('A')\n",
        "    t_count = dna_sequence.count('T')\n",
        "    g_count = dna_sequence.count('G')\n",
        "    c_count = dna_sequence.count('C')\n",
        "\n",
        "    # Calculate the total number of bases\n",
        "    total_bases = a_count + t_count + g_count + c_count\n",
        "\n",
        "    # Calculate the percentage of each base\n",
        "    a_percent = (a_count / total_bases) * 100\n",
        "    t_percent = (t_count / total_bases) * 100\n",
        "    g_percent = (g_count / total_bases) * 100\n",
        "    c_percent = (c_count / total_bases) * 100\n",
        "\n",
        "    return a_percent, t_percent, g_percent, c_percent\n",
        "\n",
        "def plot_base_content(a_percent, t_percent, g_percent, c_percent, species_name, save_path):\n",
        "    # Define the bases and their corresponding percentages and colors\n",
        "    bases = ['A', 'T', 'G', 'C']\n",
        "    percentages = [a_percent, t_percent, g_percent, c_percent]\n",
        "    colors = ['green', 'red', 'black', 'blue']\n",
        "\n",
        "    # Create a bar plot\n",
        "    plt.figure(figsize=(8, 6))\n",
        "    bars = plt.bar(bases, percentages, color=colors)\n",
        "\n",
        "    # Set plot title and labels\n",
        "    plt.title(f'Relative Frequency of Bases in {species_name} Gene')\n",
        "    plt.xlabel('Bases')\n",
        "    plt.ylabel('Relative Percentage (%)')\n",
        "    plt.ylim(0, 100)\n",
        "\n",
        "    # Add percentage labels to each bar\n",
        "    for bar, percentage in zip(bars, percentages):\n",
        "        plt.text(bar.get_x() + bar.get_width()/2, bar.get_height(), f'{percentage:.2f}%',\n",
        "                 ha='center', va='bottom')\n",
        "\n",
        "    # Save the plot to the specified path\n",
        "    plt.savefig(save_path)\n",
        "    plt.show()\n",
        "\n",
        "# Example DNA sequence and save path\n",
        "dna_sequence = \"your_dna_sequence_here\"  # Replace with your actual DNA sequence\n",
        "save_path = \"path/to/save/your_plot.png\"  # Replace with your actual save path\n",
        "\n",
        "# Calculate the relative content of bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "\n",
        "# Generate the bar plot\n",
        "species_name = \"Species_Name\"  # Replace with the actual species name\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, species_name, save_path)\n"
      ],
      "metadata": {
        "id": "CSPlI6aTiTa0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Frequence Chaos Representation Game (*FCGR*) - CODE DNA K-mer Fractal Generator (*DKFG*)**"
      ],
      "metadata": {
        "id": "bft77Czidp7B"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **FREQUENCY COUNTING FOR K-MERS**"
      ],
      "metadata": {
        "id": "VKNCGeWynpT5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define the count_kmers function\n",
        "def count_kmers(seq, k):\n",
        "    d = collections.defaultdict(int)\n",
        "    for i in range(len(seq) - (k - 1)):\n",
        "        d[seq[i:i+k]] += 1\n",
        "    return d# Define the count_kmers function\n",
        "def count_kmers(seq, k):\n",
        "    d = collections.defaultdict(int)\n",
        "    for i in range(len(seq) - (k - 1)):\n",
        "        d[seq[i:i+k]] += 1\n",
        "    return d"
      ],
      "metadata": {
        "id": "TXwpvi9Pl8-L"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **FCGR NUMERICAL MATRIX GENERATOR**"
      ],
      "metadata": {
        "id": "lc-B0_P9nxVV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define la funci√≥n chaos_game_representation\n",
        "def chaos_game_representation(frequencies, k):\n",
        "    array_size = int(math.sqrt(4**k))\n",
        "    chaos = []\n",
        "    for i in range(array_size):\n",
        "        chaos.append([0] * array_size)\n",
        "\n",
        "    maxx = array_size\n",
        "    maxy = array_size\n",
        "    posx = 1\n",
        "    posy = 1\n",
        "    for key, value in frequencies.items():\n",
        "        for char in key:\n",
        "            if char == \"A\":\n",
        "                posx += maxx / 2\n",
        "            elif char == \"T\":\n",
        "                posy += maxy / 2\n",
        "            elif char == \"G\":\n",
        "                posx += maxx / 2\n",
        "                posy += maxy / 2\n",
        "            maxx = maxx / 2\n",
        "            maxy /= 2\n",
        "        chaos[int(posy)-1][int(posx)-1] = value\n",
        "        maxx = array_size\n",
        "        maxy = array_size\n",
        "        posx = 1\n",
        "        posy = 1\n",
        "\n",
        "    return chaos"
      ],
      "metadata": {
        "id": "5o_Rx2BQmDCU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **FRACTAL IMAGE GENERATOR**"
      ],
      "metadata": {
        "id": "uAJ82B5tn4r_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Calculate k-mers count n is the size of the k-mer\n",
        "kmer_counts = count_kmers(sequence_data, n)\n",
        "# Generate chaos game representation\n",
        "chaos_representation = chaos_game_representation(kmer_counts, n)\n",
        "\n",
        "# Function to save an image with a specific name\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Adjust size and resolution here\n",
        "    ax.imshow(image, interpolation='nearest', cmap='flag_r')  # Adjust color map if needed\n",
        "    ax.set_xlabel('Total pixels for $2^7$ ')\n",
        "    ax.set_ylabel('Total pixels for $2^7$ ')\n",
        "    ax.set_title(f'FCGR for {file_name} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{file_name}.png'))\n",
        "    plt.close(fig)  # Close the current figure to free memory\n",
        "\n",
        "# Save the image\n",
        "save_image(chaos_representation, f'{file_name}_k7')\n",
        "\n",
        "# Display the image\n",
        "pylab.title(f'FCGR for {file_name} / 7-mers')\n",
        "pylab.imshow(chaos_representation, interpolation='nearest', cmap='flag_r')  # Adjust color map if needed\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "pylab.show()\n"
      ],
      "metadata": {
        "id": "J7aZJicVnQuS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software fractalandscape  - CODE Fractal Landscape Generator (*FLG*)**  "
      ],
      "metadata": {
        "id": "FBeDM9CAdp3l"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# GENERATING THE LANDSCAPE USING THE NUMERIC MATRIX GENERATED BY FCGR\n",
        "#SAVED FILE NAME\n",
        "name = \"NAME_SEQUENCE\"\n",
        "#SAVED FILE PATH\n",
        "output_directory_3 = \"/path/to/your/directory\"\n",
        "\n",
        "#NORMALIZED\n",
        "# Convert lists to NumPy arrays\n",
        "chaos_N_kn = np.array(chaos_N_kn)\n",
        "# Create a Normalize object for both images (using the minimum and maximum values of all images)\n",
        "global_min = min(chaos_N_kn.min())\n",
        "global_max = max(chaos_N_kn.max())\n",
        "global_normalization = Normalize(vmin=global_min, vmax=global_max)\n",
        "\n",
        "#COLOR MAP\n",
        "color_2 = cm.tab20c\n",
        "\n",
        "# Define the landscape size (resolution)\n",
        "resolution1 = len(chaos_N_kn)\n",
        "# Generate coordinates in the range [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Create a coordinate grid\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Use the heights from chaos_X_k7\n",
        "Z = np.array(chaos_N_kn).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configure the colormap and 3D visualization\n",
        "fig = plt.figure(figsize=(20, 15), dpi=res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Adjust elevation and viewing angle\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Create the three-dimensional surface\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configure labels\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {name} / 7-mers')\n",
        "\n",
        "# Add a color bar\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "color_bar = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "file_name = f'{name} Landscape 7-mers.png'\n",
        "full_path = os.path.join(output_directory_3, file_name)\n",
        "plt.savefig(full_path)\n",
        "\n",
        "# Show the image\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "wcHorGrhoZ6n"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE Nucleotide Counter and Visualizer (*NCV*)**"
      ],
      "metadata": {
        "id": "ghFm2aL_l7BE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import matplotlib.pyplot as plt\n",
        "\n",
        "def calculate_base_content_T(dna_sequence):\n",
        "    length = len(dna_sequence)\n",
        "    a_count = dna_sequence.count('A')\n",
        "    t_count = dna_sequence.count('T')\n",
        "    g_count = dna_sequence.count('G')\n",
        "    c_count = dna_sequence.count('C')\n",
        "    total_bases = length\n",
        "    return a_count, t_count, g_count, c_count\n",
        "\n",
        "def plot_base_content_T(a_count, t_count, g_count, c_count, species_name, save_path):\n",
        "    bases = ['A', 'T', 'G', 'C']\n",
        "    counts = [a_count, t_count, g_count, c_count]\n",
        "    colors = ['green', 'red', 'black', 'blue']\n",
        "\n",
        "    plt.figure(figsize=(8, 6))\n",
        "    bars = plt.bar(bases, counts, color=colors)\n",
        "    plt.title(f'Total Frequency of bases in {species_name} gene')\n",
        "    plt.xlabel('Bases')\n",
        "    plt.ylabel('Total Count')\n",
        "\n",
        "    for bar, count in zip(bars, counts):\n",
        "        plt.text(bar.get_x() + bar.get_width()/2, bar.get_height(), f'{count}',\n",
        "                 ha='center', va='bottom')\n",
        "\n",
        "    # Guardar la imagen en la ruta especificada\n",
        "    plt.savefig(save_path)\n",
        "    plt.show()\n",
        "\n",
        "# Example DNA sequence\n",
        "dna_sequence = 'your_dna_sequence_here'  # Ensure to define your DNA sequence properly\n",
        "# Define the path where you want to save the image\n",
        "save_path = 'path/to/save/your/image.png'\n",
        "# Define the species name\n",
        "species_name = 'Species_Name'\n",
        "\n",
        "# Calculate the base content percentages\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "# Generate and save the bar plot\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, species_name, save_path)"
      ],
      "metadata": {
        "id": "MbZLULXWj-SC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Frequency-Clustering of k-mers - CODE K-mer Frequency-Clustering (*KFC*)**"
      ],
      "metadata": {
        "id": "-fFsYOJOdp0R"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import shutil\n",
        "from collections import defaultdict\n",
        "from operator import itemgetter\n",
        "#SEQUENCE NAME\n",
        "NAME = \"SEQUENCE NAME\"\n",
        "# Let's say you have the dictionary 'dero'\n",
        "dero = \"dictionary\"\n",
        "\n",
        "# Create a defaultdict to group similar values\n",
        "groups = defaultdict(list)\n",
        "\n",
        "# Fill the dictionary with similar values\n",
        "for key, value in dero.items():\n",
        "    groups[value].append(key)\n",
        "\n",
        "# Create a counting dictionary to count occurrences in each group\n",
        "group_counts = {value: len(keys) for value, keys in groups.items()}\n",
        "\n",
        "# Sort the counting dictionary from highest to lowest by value\n",
        "sorted_group_counts = dict(sorted(group_counts.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "# Destination path\n",
        "destination_path = \"/path/to/your/directory\"\n",
        "\n",
        "# Save the output to a text file in Markdown format\n",
        "with open(f'{NAME}.txt', 'w') as file:\n",
        "    # Add header title\n",
        "    file.write(f\"## Frequency of occurrence of 7-mers for {NAME}\\n\\n\")\n",
        "    for value, count in sorted_group_counts.items():\n",
        "        # Use Markdown format for bold text\n",
        "        line = f\"**Frequency of appearance {value}:** {count} **Nucleotide k-mer Composition** - {groups[value]}\\n\"\n",
        "        file.write(line)\n",
        "\n",
        "# Move the file to the destination path\n",
        "shutil.move(f'{NAME}.txt', destination_path + f'{NAME}.txt')\n",
        "\n",
        "print(f\"Results saved in '{destination_path}{NAME}.txt'\")\n",
        "\n"
      ],
      "metadata": {
        "id": "lseK8d9trA_0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Histogram-Frequency-mers Histograma - CODE K-mer Frequency-Rank Plot (*KFRP*)**\n"
      ],
      "metadata": {
        "id": "tPyX5J7AdpvH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# List of your datasets\n",
        "datasets = [dataset1, dataset2, dataset3, dataset4, dataset5, dataset6, dataset7, dataset8,\n",
        "            dataset9, dataset10, dataset11, dataset12, dataset13, dataset14, dataset15, dataset16]\n",
        "\n",
        "\n",
        "species_names = [\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{G. gorilla \\ PRKN}$',\n",
        "    r'$\\mathit{P. paniscus \\ PRKN}$',\n",
        "    r'$\\mathit{P. troglodytes \\ PRKN}$',\n",
        "    r'$\\mathit{P. abelii \\ PRKN}$',\n",
        "    r'$\\mathit{P. pygmaeus \\ PRKN}$',\n",
        "    r'$\\mathit{M. fascicularis \\ PRKN}$',\n",
        "     r'$\\mathit{M. mulatta \\ PRKN}$',\n",
        "    r'$\\mathit{M. thibetana \\ PRKN}$',\n",
        "    r'$\\mathit{R. roxellana \\ PRKN}$',\n",
        "    r'$\\mathit{T. francoisi \\ PRKN}$',\n",
        "    r'$\\mathit{N. leucogenys \\ PRKN}$',\n",
        "    r'$\\mathit{S. syndactylus \\ PRKN}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$',\n",
        "]\n",
        "\n",
        "# List of species names corresponding to each dataset\n",
        "species_labels = [\n",
        "    '01 H. sapiens',\n",
        "    '02 G. gorilla',\n",
        "    '03 P. paniscus',\n",
        "    '04 P. troglodytes',\n",
        "    '05 P. abelii',\n",
        "    '06 P. pygmaeus',\n",
        "    '07 M. fascicularis',\n",
        "    '08 M. mulatta',\n",
        "    '09 M. thibetana',\n",
        "    '10 R. roxellana',\n",
        "    '11 T. francoisi',\n",
        "    '12 N. leucogenys',\n",
        "    '13 S. syndactylus',\n",
        "    '14 M. murinus',\n",
        "    '15 C. jacchus',\n",
        "    '16 L. catta',\n",
        "]\n",
        "\n",
        "# Base path of the folder where the images will be saved\n",
        "base_path = \"/path/to/your/directory\"\n",
        "\n",
        "# For each dataset, generate and save the plot\n",
        "for i, data in enumerate(datasets, start=1):\n",
        "    # Path of the specific folder for this species\n",
        "    species_folder = os.path.join(base_path, species_labels[i-1])\n",
        "\n",
        "    # If the folder does not exist, create it\n",
        "    if not os.path.exists(species_folder):\n",
        "        os.makedirs(species_folder)\n",
        "\n",
        "    # Change the current directory to the species folder\n",
        "    os.chdir(species_folder)\n",
        "\n",
        "    # Create a new figure with the desired size\n",
        "    plt.figure(figsize=(20, 11))\n",
        "    # Calculate frequencies and sort by frequency\n",
        "    freq_list = sorted(list(data.values()), reverse=True)\n",
        "\n",
        "    # Create a bar plot\n",
        "    plt.bar(\n",
        "        range(1, len(freq_list) + 1),\n",
        "        freq_list,\n",
        "        label=f'{species_names[i-1]}',  # Use the species name instead of \"Data\"\n",
        "        color='blue',    # Bar color\n",
        "        alpha=0.7        # Transparency\n",
        "    )\n",
        "\n",
        "    # Set labels and title\n",
        "    plt.xlabel('Range')\n",
        "    plt.ylabel('Frequency')\n",
        "    plt.title(f'Frequency Plot for {species_names[i-1]} of size 7-mers')\n",
        "\n",
        "    # Show legend\n",
        "    plt.legend()\n",
        "\n",
        "    # Save the plot with the desired name\n",
        "    file_name = species_labels[i-1] + \"_k7.png\"\n",
        "    plt.savefig(file_name)\n",
        "\n",
        "    # Close the figure to free memory\n",
        "    plt.close()\n",
        "\n",
        "    # Change the working directory back to the original\n",
        "    os.chdir(base_path)\n"
      ],
      "metadata": {
        "id": "cVr6Cn_puXg9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE Shannon Entropy Calculator and Visualizer (*SECV*)**\n"
      ],
      "metadata": {
        "id": "Rc4LGpNCdpnb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import collections\n",
        "import math\n",
        "import os\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "def entropy(sequence):\n",
        "    # Calculate the probability distribution of characters in the sequence.\n",
        "    counts = collections.Counter(sequence)\n",
        "    probabilities = {char: count / len(sequence) for char, count in counts.items()}\n",
        "\n",
        "    # Calculate Shannon entropy.\n",
        "    entropy = 0\n",
        "    for probability in probabilities.values():\n",
        "        if probability > 0:\n",
        "            entropy -= probability * math.log2(probability)\n",
        "\n",
        "    return round(entropy, 5)  # Reduce the result to 5 decimal places\n",
        "\n",
        "\n",
        "# Example entropy calculations (replace DATOS1, DATOS2, etc. with actual data sequences)\n",
        "entropy1 = entropy(DNA_SEQUENCE1)\n",
        "entropy2 = entropy(DNA_SEQUENCE2)\n",
        "entropy3 = entropy(DNA_SEQUENCE3)\n",
        "entropy4 = entropy(DNA_SEQUENCE4)\n",
        "entropy5 = entropy(DNA_SEQUENCE5)\n",
        "entropy6 = entropy(DNA_SEQUENCE6)\n",
        "entropy7 = entropy(DNA_SEQUENCE7)\n",
        "entropy8 = entropy(DNA_SEQUENCE8)\n",
        "entropy9 = entropy(DNA_SEQUENCE9)\n",
        "entropy10 = entropy(DNA_SEQUENCE10)\n",
        "entropy11 = entropy(DNA_SEQUENCE11)\n",
        "entropy12 = entropy(DNA_SEQUENCE12)\n",
        "entropy13 = entropy(DNA_SEQUENCE13)\n",
        "entropy14 = entropy(DNA_SEQUENCE14)\n",
        "entropy15 = entropy(DNA_SEQUENCE15)\n",
        "entropy16 = entropy(DNA_SEQUENCE16)\n",
        "\n",
        "# Desired path\n",
        "desired_path = \"/path/to/your/directory\"\n",
        "\n",
        "# Entropy data\n",
        "entropies = {\n",
        "    'H. sapiens': entropy1,\n",
        "    'G. gorilla': entropy2,\n",
        "    'P. paniscus': entropy3,\n",
        "    'P. troglodytes': entropy4,\n",
        "    'P. abeli': entropy5,\n",
        "    'P. pygmaeus': entropy6,\n",
        "    'M. fascicularis': entropy7,\n",
        "    'M. mulatta': entropy8,\n",
        "    'M. thibetana': entropy9,\n",
        "    'R. roxellana': entropy10,\n",
        "    'T. francoisi': entropy11,\n",
        "    'N. leucogenys': entropy12,\n",
        "    'S. syndactylus': entropy13,\n",
        "    'M. murinus': entropy14,\n",
        "    'C. jacchus': entropy15,\n",
        "    'L. catta': entropy16,\n",
        "}\n",
        "\n",
        "# Create a pandas DataFrame\n",
        "df = pd.DataFrame(list(entropies.items()), columns=['Primate species', 'Shannon entropy'])\n",
        "\n",
        "# Sort the DataFrame by entropy from lowest to highest\n",
        "df = df.sort_values(by='Shannon entropy')\n",
        "\n",
        "# Full path to save the Excel file\n",
        "excel_path = os.path.join(desired_path, 'Shannon_entropy_Primate_species.xlsx')\n",
        "\n",
        "# Save the DataFrame to an Excel file with bold and italic species names\n",
        "with pd.ExcelWriter(excel_path, engine='xlsxwriter') as writer:\n",
        "    df.style.applymap(lambda x: 'font-weight: bold; font-style: italic', subset=['Primate species']).to_excel(writer, index=False, sheet_name='Entropies')\n",
        "\n",
        "# Show success message\n",
        "print(f\"Excel file saved at: {excel_path}\")\n",
        "\n",
        "# Define the destination path\n",
        "destination_path = \"/path/to/your/directory\"\n",
        "\n",
        "# Sort the entropies from lowest to highest\n",
        "sorted_entropies = sorted(entropies.items(), key=lambda x: x[1])\n",
        "\n",
        "# Extract the species names and sorted entropies\n",
        "sorted_species, sorted_entropies = zip(*sorted_entropies)\n",
        "\n",
        "# Calculate the mean and standard deviation\n",
        "mean_entropy = np.mean(sorted_entropies)\n",
        "std_deviation = np.std(sorted_entropies)\n",
        "\n",
        "# Plot the entropies\n",
        "plt.figure(figsize=(13, 9))\n",
        "plt.plot(sorted_species, sorted_entropies, marker='o', color='blue', linestyle='-')\n",
        "\n",
        "# Add line for the mean\n",
        "plt.axhline(y=mean_entropy, color='red', linestyle='--', label=f'Average: {mean_entropy:.5f}')\n",
        "\n",
        "# Add lines for the standard deviation\n",
        "plt.axhline(y=mean_entropy + std_deviation, color='green', linestyle='--', label=f'Average + Standard deviation: {mean_entropy + std_deviation:.5f}')\n",
        "plt.axhline(y=mean_entropy - std_deviation, color='green', linestyle='--', label=f'Average - Standard deviation: {mean_entropy - std_deviation:.5f}')\n",
        "\n",
        "# Label the points with the entropy values\n",
        "for i, txt in enumerate(sorted_entropies):\n",
        "    plt.annotate(f'{txt:.5f}', (i, sorted_entropies[i]), textcoords=\"offset points\", xytext=(-4,-4), ha='left', va='top', rotation=340, fontsize=8)\n",
        "\n",
        "# Set the x-axis with the species names\n",
        "plt.xticks(range(len(sorted_species)), sorted_species, rotation=45, ha='right')\n",
        "plt.ylabel('Shannon Entropy', fontweight='bold')\n",
        "plt.title(r'Shannon Entropy for Different $\\mathit{PRKN}$ Gene from Primate Species')\n",
        "\n",
        "# Add a legend\n",
        "plt.legend()\n",
        "\n",
        "# Save the plot to the destination path\n",
        "if not os.path.exists(destination_path):\n",
        "    os.makedirs(destination_path)\n",
        "plt.savefig(os.path.join(destination_path, 'Entropy_primate_species.png'))\n",
        "\n",
        "# Show the plot\n",
        "plt.tight_layout()\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "WThwFyzzyXXR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Caos Game Representation (*CGR*) - CODE DNA Chaos Game Algorithm (*DCGA*)**\n"
      ],
      "metadata": {
        "id": "ilSzepaDfK7l"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import matplotlib.pyplot as plt\n",
        "import gc\n",
        "\n",
        "def generate_points(dna_sequence):\n",
        "    # Define the corners of the square\n",
        "    corners = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "\n",
        "    # Initial point\n",
        "    current_point = (0.5, 0.5)\n",
        "\n",
        "    # Store the coordinates of the points for plotting\n",
        "    x_values = [current_point[0]]\n",
        "    y_values = [current_point[1]]\n",
        "\n",
        "    # Iterate over the DNA sequence\n",
        "    for nucleotide in dna_sequence:\n",
        "        # Get the corner coordinate corresponding to the current nucleotide\n",
        "        target_corner = corners[nucleotide]\n",
        "\n",
        "        # Calculate the point halfway between the current point and the target corner\n",
        "        current_point = ((current_point[0] + target_corner[0]) / 2, (current_point[1] + target_corner[1]) / 2)\n",
        "\n",
        "        # Store the coordinates of the new point\n",
        "        x_values.append(current_point[0])\n",
        "        y_values.append(current_point[1])\n",
        "\n",
        "    return x_values, y_values\n",
        "\n",
        "# Example data\n",
        "dna_data = \"your_dna_sequence_here\"\n",
        "output_path = \"your_output_path_here\"\n",
        "figure_title = \"your_title_here\"\n",
        "figure_dpi = 300\n",
        "figure_size = (8, 6)\n",
        "point_size = 1\n",
        "marker_style = 'o'\n",
        "color_map = 'viridis'\n",
        "\n",
        "x1, y1 = generate_points(dna_data)\n",
        "\n",
        "# Normalize point indices using the reference scale\n",
        "reference_scale = (0, 100)\n",
        "normalized_indices = range(len(x1))\n",
        "\n",
        "# Define corners with labels\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "\n",
        "# Plot only the points without lines, with reduced size\n",
        "fig, ax = plt.subplots(figsize=figure_size, dpi=figure_dpi)\n",
        "scatter_plot = plt.scatter(x1, y1, c=normalized_indices, cmap=color_map, s=point_size, marker=marker_style)\n",
        "\n",
        "# Show corners with labels\n",
        "for nucleotide, (x, y) in corners_labels.items():\n",
        "    plt.text(x, y, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "plt.title(f\"CGR of {figure_title}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Add the color bar\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "\n",
        "# Save the figure\n",
        "plt.savefig(os.path.join(output_path, f\"CGR_of_{figure_title}.png\"), dpi=figure_dpi, bbox_inches='tight')\n",
        "\n",
        "plt.show()\n",
        "gc.collect()\n"
      ],
      "metadata": {
        "id": "COmDkxBXouuQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Dendrogram Similarity Generator - CODE Similarity Dendrogram (*SimDendro*)**\n"
      ],
      "metadata": {
        "id": "a8N8DQczfQpk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "from scipy.cluster.hierarchy import dendrogram, linkage, cophenet\n",
        "from scipy.spatial.distance import pdist\n",
        "\n",
        "# Assuming data_list_1 to data_list_16 are lists that need to be converted to NumPy arrays, each of the data_list corresponds to the numerical matrix generated by the FCGR\n",
        "data_array_1 = np.array(data_list_1)\n",
        "data_array_2 = np.array(data_list_2)\n",
        "data_array_3 = np.array(data_list_3)\n",
        "data_array_4 = np.array(data_list_4)\n",
        "data_array_5 = np.array(data_list_5)\n",
        "data_array_6 = np.array(data_list_6)\n",
        "data_array_7 = np.array(data_list_7)\n",
        "data_array_8 = np.array(data_list_8)\n",
        "data_array_9 = np.array(data_list_9)\n",
        "data_array_10 = np.array(data_list_10)\n",
        "data_array_11 = np.array(data_list_11)\n",
        "data_array_12 = np.array(data_list_12)\n",
        "data_array_13 = np.array(data_list_13)\n",
        "data_array_14 = np.array(data_list_14)\n",
        "data_array_15 = np.array(data_list_15)\n",
        "data_array_16 = np.array(data_list_16)\n",
        "\n",
        "# Concatenate the flattened matrices\n",
        "flattened_matrices = np.vstack([\n",
        "    data_array_1.flatten(),\n",
        "    data_array_2.flatten(),\n",
        "    data_array_3.flatten(),\n",
        "    data_array_4.flatten(),\n",
        "    data_array_5.flatten(),\n",
        "    data_array_6.flatten(),\n",
        "    data_array_7.flatten(),\n",
        "    data_array_8.flatten(),\n",
        "    data_array_9.flatten(),\n",
        "    data_array_10.flatten(),\n",
        "    data_array_11.flatten(),\n",
        "    data_array_12.flatten(),\n",
        "    data_array_13.flatten(),\n",
        "    data_array_14.flatten(),\n",
        "    data_array_15.flatten(),\n",
        "    data_array_16.flatten()\n",
        "])\n",
        "\n",
        "# Calculate the Euclidean distance between the flattened matrices\n",
        "distance_matrix = pdist(flattened_matrices, metric='euclidean')\n",
        "\n",
        "# Generate the linkage matrix\n",
        "Z = linkage(distance_matrix, method='average')\n",
        "\n",
        "# Convert the linkage matrix to type double\n",
        "Z = Z.astype('double')\n",
        "\n",
        "# Construct the dendrogram without squareform\n",
        "#plt.figure(figsize=(12, 15))\n",
        "dendrogram_object = dendrogram(Z, labels=[\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{M. fascicularis}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M. thibetana}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$',\n",
        "])\n",
        "# Rotate the x-axis labels 90 degrees\n",
        "plt.xticks(rotation=90, ha='center')\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "\n",
        "# Destination path\n",
        "destination_path = \"/path/to/your/directory/\"\n",
        "\n",
        "# Save the image to the destination path\n",
        "plt.savefig(destination_path + \"dendrogram1.png\")\n",
        "\n",
        "# Calculate the cophenetic index\n",
        "cophenet_index, _ = cophenet(Z, distance_matrix)\n",
        "\n",
        "# Add the cophenetic index as text on the figure\n",
        "plt.text(0.5, 0.95, f'Cophenetic index: {cophenet_index:.3f}', horizontalalignment='left', verticalalignment='center', transform=plt.gca().transAxes, fontsize=12)\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "\n",
        "# Show the dendrogram\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "UImYHiRBpXkG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE  Destagger convert Grayscale Fractal (*DCGF*)**"
      ],
      "metadata": {
        "id": "OMyyRtILr1L-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import gc\n",
        "import cv2\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Function to save an image with a specific name\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Adjust size and resolution here\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.axis('off')  # Remove the axis\n",
        "    fig.savefig(os.path.join(output_directory, f'{name}.png'), bbox_inches='tight', pad_inches=0)\n",
        "    plt.close(fig)  # Close the current figure to free memory\n",
        "\n",
        "# Directory where the images will be saved\n",
        "output_directory = \"/path/to/output/directory\"\n",
        "gc.collect()\n",
        "\n",
        "save_image(chaos_1_k7, f'{name_1}_7-mers')\n",
        "\n",
        "# Source folder\n",
        "source_folder = \"/path/to/source/folder\"\n",
        "\n",
        "# Check if the folder exists\n",
        "if os.path.exists(source_folder):\n",
        "    # Get the list of files in the folder\n",
        "    files = os.listdir(source_folder)\n",
        "\n",
        "    # List to store image paths\n",
        "    paths = []\n",
        "\n",
        "    # Iterate over each file and get the full path\n",
        "    for file in files:\n",
        "        full_path = os.path.join(source_folder, file)\n",
        "        # Check if the file is an image based on its extension\n",
        "        if any(file.lower().endswith(ext) for ext in ['.jpg', '.jpeg', '.png', '.gif', '.bmp']):\n",
        "            paths.append(full_path)\n",
        "\n",
        "    # Destination path for the generated images\n",
        "    destination_path = \"/path/to/destination/folder\"\n",
        "\n",
        "    # Iterate over the paths\n",
        "    for path in paths:\n",
        "        # Check if the path exists\n",
        "        if os.path.exists(path):\n",
        "            # Load the color image\n",
        "            img_color = cv2.imread(path)\n",
        "\n",
        "            # Transform to grayscale\n",
        "            img_gray = cv2.cvtColor(img_color, cv2.COLOR_BGR2GRAY)\n",
        "\n",
        "            # Save the generated grayscale image\n",
        "            file_name = os.path.basename(path)  # Get the file name from the path\n",
        "            output_name = f'{destination_path}/{file_name[:-4]}_binary.png'  # Create the output file name\n",
        "            cv2.imwrite(output_name, img_gray)\n",
        "\n",
        "        else:\n",
        "            print(f\"The path {path} does not exist.\")\n",
        "else:\n",
        "    print(f\"The folder {source_folder} does not exist.\")"
      ],
      "metadata": {
        "id": "8FVicv2Fspej"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Image Difference - CODE Fractal Image Difference Analyzer (*FIDA*)**"
      ],
      "metadata": {
        "id": "TD0wHp2ifbDt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "7\n",
        "import os\n",
        "import cv2\n",
        "import numpy as np\n",
        "\n",
        "# Folder containing the images\n",
        "folder = \"/path/to/images\"\n",
        "\n",
        "# Check if the folder exists\n",
        "if os.path.exists(folder):\n",
        "    # Get the list of files in the folder\n",
        "    files = os.listdir(folder)\n",
        "\n",
        "    # Initialize a list to store the image paths\n",
        "    paths = []\n",
        "\n",
        "    # Iterate over each file and get the full path\n",
        "    for file in files:\n",
        "        full_path = os.path.join(folder, file)\n",
        "        # Check if the file is an image based on its extension\n",
        "        if any(file.lower().endswith(ext) for ext in ['.jpg', '.jpeg', '.png', '.gif', '.bmp']):\n",
        "            paths.append(full_path)\n",
        "else:\n",
        "    print(f\"The folder {folder} does not exist.\")\n",
        "\n",
        "# Initialize a dictionary to store the comparison results\n",
        "comparison_results = {}\n",
        "\n",
        "# Iterate over the reference images\n",
        "for i, reference in enumerate(paths):\n",
        "    # Load the reference image\n",
        "    reference_image = cv2.imread(reference)\n",
        "\n",
        "    # Ensure that the reference image is not None\n",
        "    if reference_image is None:\n",
        "        print(f\"Could not load the reference image at {reference}\")\n",
        "        continue\n",
        "\n",
        "    # Calculate the total number of pixels in the reference image\n",
        "    total_reference_pixels = reference_image.shape[1] * reference_image.shape[0]\n",
        "\n",
        "    # Iterate over the comparison images\n",
        "    for j, comparison in enumerate(paths):\n",
        "        # Load the comparison image\n",
        "        comparison_image = cv2.imread(comparison)\n",
        "\n",
        "        # Ensure that both images have the same dimensions\n",
        "        comparison_image = cv2.resize(comparison_image, (reference_image.shape[1], reference_image.shape[0]))\n",
        "\n",
        "        # Compare the images pixel by pixel and get the difference\n",
        "        difference = cv2.absdiff(reference_image, comparison_image)\n",
        "\n",
        "        # Get the binary color intensity (black or white)\n",
        "        _, binary = cv2.threshold(difference, 30, 255, cv2.THRESH_BINARY)\n",
        "\n",
        "        # Calculate the number of different pixels\n",
        "        different_pixels = np.sum(binary == 255)\n",
        "\n",
        "        # Calculate the binary color intensity (percentage of different pixels)\n",
        "        binary_intensity = (different_pixels / total_reference_pixels) * 100\n",
        "\n",
        "        # Store the comparison result in the comparison results dictionary\n",
        "        key = f\"ID_{i+1}vs_{j+1}\"\n",
        "        comparison_results[key] = binary_intensity\n",
        "\n",
        "# Dynamically create variables for each combination of ID and store the corresponding value\n",
        "for key, value in comparison_results.items():\n",
        "    globals()[key] = value\n"
      ],
      "metadata": {
        "id": "EGVKUZ6sp1Nw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Definir la matriz de datos\n",
        "data = [\n",
        "    [ID_1vs_1, ID_2vs_1, ID_3vs_1, ID_4vs_1, ID_5vs_1, ID_6vs_1, ID_7vs_1, ID_8vs_1, ID_9vs_1, ID_10vs_1, ID_11vs_1, ID_12vs_1, ID_13vs_1, ID_14vs_1, ID_15vs_1, ID_16vs_1],\n",
        "    [ID_1vs_2, ID_2vs_2, ID_3vs_2, ID_4vs_2, ID_5vs_2, ID_6vs_2, ID_7vs_2, ID_8vs_2, ID_9vs_2, ID_10vs_2, ID_11vs_2, ID_12vs_2, ID_13vs_2, ID_14vs_2, ID_15vs_2, ID_16vs_2],\n",
        "    [ID_1vs_3, ID_2vs_3, ID_3vs_3, ID_4vs_3, ID_5vs_3, ID_6vs_3, ID_7vs_3, ID_8vs_3, ID_9vs_3, ID_10vs_3, ID_11vs_3, ID_12vs_3, ID_13vs_3, ID_14vs_3, ID_15vs_3, ID_16vs_3],\n",
        "    [ID_1vs_4, ID_2vs_4, ID_3vs_4, ID_4vs_4, ID_5vs_4, ID_6vs_4, ID_7vs_4, ID_8vs_4, ID_9vs_4, ID_10vs_4, ID_11vs_4, ID_12vs_4, ID_13vs_4, ID_14vs_4, ID_15vs_4, ID_16vs_4],\n",
        "    [ID_1vs_5, ID_2vs_5, ID_3vs_5, ID_4vs_5, ID_5vs_5, ID_6vs_5, ID_7vs_5, ID_8vs_5, ID_9vs_5, ID_10vs_5, ID_11vs_5, ID_12vs_5, ID_13vs_5, ID_14vs_5, ID_15vs_5, ID_16vs_5],\n",
        "    [ID_1vs_6, ID_2vs_6, ID_3vs_6, ID_4vs_6, ID_5vs_6, ID_6vs_6, ID_7vs_6, ID_8vs_6, ID_9vs_6, ID_10vs_6, ID_11vs_6, ID_12vs_6, ID_13vs_6, ID_14vs_6, ID_15vs_6, ID_16vs_6],\n",
        "    [ID_1vs_7, ID_2vs_7, ID_3vs_7, ID_4vs_7, ID_5vs_7, ID_6vs_7, ID_7vs_7, ID_8vs_7, ID_9vs_7, ID_10vs_7, ID_11vs_7, ID_12vs_7, ID_13vs_7, ID_14vs_7, ID_15vs_7, ID_16vs_7],\n",
        "    [ID_1vs_8, ID_2vs_8, ID_3vs_8, ID_4vs_8, ID_5vs_8, ID_6vs_8, ID_7vs_8, ID_8vs_8, ID_9vs_8, ID_10vs_8, ID_11vs_8, ID_12vs_8, ID_13vs_8, ID_14vs_8, ID_15vs_8, ID_16vs_8],\n",
        "    [ID_1vs_9, ID_2vs_9, ID_3vs_9, ID_4vs_9, ID_5vs_9, ID_6vs_9, ID_7vs_9, ID_8vs_9, ID_9vs_9, ID_10vs_9, ID_11vs_9, ID_12vs_9, ID_13vs_9, ID_14vs_9, ID_15vs_9, ID_16vs_9],\n",
        "    [ID_1vs_10, ID_2vs_10, ID_3vs_10, ID_4vs_10, ID_5vs_10, ID_6vs_10, ID_7vs_10, ID_8vs_10, ID_9vs_10, ID_10vs_10, ID_11vs_10, ID_12vs_10, ID_13vs_10, ID_14vs_10, ID_15vs_10, ID_16vs_10],\n",
        "    [ID_1vs_11, ID_2vs_11, ID_3vs_11, ID_4vs_11, ID_5vs_11, ID_6vs_11, ID_7vs_11, ID_8vs_11, ID_9vs_11, ID_10vs_11, ID_11vs_11, ID_12vs_11, ID_13vs_11, ID_14vs_11, ID_15vs_11, ID_16vs_11],\n",
        "    [ID_1vs_12, ID_2vs_12, ID_3vs_12, ID_4vs_12, ID_5vs_12, ID_6vs_12, ID_7vs_12, ID_8vs_12, ID_9vs_12, ID_10vs_12, ID_11vs_12, ID_12vs_12, ID_13vs_12, ID_14vs_12, ID_15vs_12, ID_16vs_12],\n",
        "    [ID_1vs_13, ID_2vs_13, ID_3vs_13, ID_4vs_13, ID_5vs_13, ID_6vs_13, ID_7vs_13, ID_8vs_13, ID_9vs_13, ID_10vs_13, ID_11vs_13, ID_12vs_13, ID_13vs_13, ID_14vs_13, ID_15vs_13, ID_16vs_13],\n",
        "    [ID_1vs_14, ID_2vs_14, ID_3vs_14, ID_4vs_14, ID_5vs_14, ID_6vs_14, ID_7vs_14, ID_8vs_14, ID_9vs_14, ID_10vs_14, ID_11vs_14, ID_12vs_14, ID_13vs_14, ID_14vs_14, ID_15vs_14, ID_16vs_14],\n",
        "    [ID_1vs_15, ID_2vs_15, ID_3vs_15, ID_4vs_15, ID_5vs_15, ID_6vs_15, ID_7vs_15, ID_8vs_15, ID_9vs_15, ID_10vs_15, ID_11vs_15, ID_12vs_15, ID_13vs_15, ID_14vs_15, ID_15vs_15, ID_16vs_15],\n",
        "    [ID_1vs_16, ID_2vs_16, ID_3vs_16, ID_4vs_16, ID_5vs_16, ID_6vs_16, ID_7vs_16, ID_8vs_16, ID_9vs_16, ID_10vs_16, ID_11vs_16, ID_12vs_16, ID_13vs_16, ID_14vs_16, ID_15vs_16, ID_16vs_16]\n",
        "]\n",
        "\n",
        "# Convertir la matriz a un array de numpy para visualizaci√≥n\n",
        "matrix_data = np.array(data)\n",
        "\n",
        "# Obtener etiquetas de filas y columnas\n",
        "species_labels = [r'$\\mathit{H. sapiens}$', r'$\\mathit{G. gorilla}$', r'$\\mathit{P. paniscus}$',  r'$\\mathit{P. troglodytes}$',\n",
        "                   r'$\\mathit{P. abelii}$', r'$\\mathit{P. pygmaeus}$', r'$\\mathit{M. fascicularis}$' , r'$\\mathit{M. mulatta}$' ,r'$\\mathit{M. thibetana}$' ,\n",
        "                   r'$\\mathit{R. roxellana}$' , r'$\\mathit{T. francoisi}$' ,  r'$\\mathit{N. leucogenys}$', r'$\\mathit{S. syndactylus}$',\n",
        "                   r'$\\mathit{M. murinus}$',  r'$\\mathit{C. jacchus}$', r'$\\mathit{L. catta}$']\n",
        "\n",
        "# Crear DataFrame\n",
        "df = pd.DataFrame(data, columns=species_labels, index=species_labels)\n",
        "\n",
        "# Redondear los valores del DataFrame a dos cifras decimales\n",
        "df = df.round(2)\n",
        "\n",
        "# Crear la imagen de l0a tabla\n",
        "plt.figure(figsize=(24, 6.5))\n",
        "table = plt.table(cellText=df.values, colLabels=df.columns, rowLabels=df.index, loc='center', cellLoc='center', colColours=plt.cm.BuPu(np.full(len(df.columns), 0.1)))\n",
        "table.auto_set_font_size(False)\n",
        "table.set_fontsize(10)\n",
        "table.scale(1, 1.5)  # Ajusta la escala para mejorar la visualizaci√≥n\n",
        "\n",
        "species_labels = df.index.tolist()\n",
        "plt.title(r'Percentage Difference Table for $\\mathit{PRKN}$ Gene in Primates (IMAGE VS IMAGE) / 7-mer (unnormalized)' , fontsize=19)\n",
        "plt.xticks(np.arange(len(species_labels)), species_labels, rotation=45, ha='right', rotation_mode='anchor')\n",
        "plt.yticks(np.arange(len(species_labels)), species_labels)\n",
        "# Ocultar ejes\n",
        "plt.axis('off')\n",
        "\n",
        "output_path = \"/path/to/output/directory\"\n",
        "# Guarda la imagen en la ruta especificada\n",
        "plt.savefig(os.path.join(output_path, 'species_table_7-mer.png'), dpi=800, bbox_inches='tight')\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "a6VIfcv2AM5t"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "# Create a DataFrame with the data provided\n",
        "data = {\n",
        "    'SPECIES': [r'$\\mathit{H. sapiens}$', r'$\\mathit{G. gorilla}$', r'$\\mathit{P. paniscus}$',  r'$\\mathit{P. troglodytes}$',\n",
        "                   r'$\\mathit{P. abelii}$', r'$\\mathit{P. pygmaeus}$', r'$\\mathit{M. fascicularis}$' , r'$\\mathit{M. mulatta}$' ,r'$\\mathit{M. thibetana}$' ,\n",
        "                   r'$\\mathit{R. roxellana}$' , r'$\\mathit{T. francoisi}$' ,  r'$\\mathit{N. leucogenys}$', r'$\\mathit{S. syndactylus}$',\n",
        "                   r'$\\mathit{M. murinus}$',  r'$\\mathit{C. jacchus}$', r'$\\mathit{L. catta}$'],\n",
        "\n",
        "    r'$\\mathit{H. sapiens}$':  [ID_1vs_1, ID_2vs_1, ID_3vs_1, ID_4vs_1, ID_5vs_1, ID_6vs_1, ID_7vs_1, ID_8vs_1, ID_9vs_1, ID_10vs_1, ID_11vs_1, ID_12vs_1, ID_13vs_1, ID_14vs_1, ID_15vs_1, ID_16vs_1],\n",
        "    r'$\\mathit{G. gorilla}$': [ID_1vs_2, ID_2vs_2, ID_3vs_2, ID_4vs_2, ID_5vs_2, ID_6vs_2, ID_7vs_2, ID_8vs_2, ID_9vs_2, ID_10vs_2, ID_11vs_2, ID_12vs_2, ID_13vs_2, ID_14vs_2, ID_15vs_2, ID_16vs_2],\n",
        "    r'$\\mathit{P. paniscus}$':[ID_1vs_3, ID_2vs_3, ID_3vs_3, ID_4vs_3, ID_5vs_3, ID_6vs_3, ID_7vs_3, ID_8vs_3, ID_9vs_3, ID_10vs_3, ID_11vs_3, ID_12vs_3, ID_13vs_3, ID_14vs_3, ID_15vs_3, ID_16vs_3],\n",
        "    r'$\\mathit{P. troglodytes}$': [ID_1vs_4, ID_2vs_4, ID_3vs_4, ID_4vs_4, ID_5vs_4, ID_6vs_4, ID_7vs_4, ID_8vs_4, ID_9vs_4, ID_10vs_4, ID_11vs_4, ID_12vs_4, ID_13vs_4, ID_14vs_4, ID_15vs_4, ID_16vs_4],\n",
        "    r'$\\mathit{P. abelii}$': [ID_1vs_5, ID_2vs_5, ID_3vs_5, ID_4vs_5, ID_5vs_5, ID_6vs_5, ID_7vs_5, ID_8vs_5, ID_9vs_5, ID_10vs_5, ID_11vs_5, ID_12vs_5, ID_13vs_5, ID_14vs_5, ID_15vs_5, ID_16vs_5],\n",
        "    r'$\\mathit{P. pygmaeus}$':  [ID_1vs_6, ID_2vs_6, ID_3vs_6, ID_4vs_6, ID_5vs_6, ID_6vs_6, ID_7vs_6, ID_8vs_6, ID_9vs_6, ID_10vs_6, ID_11vs_6, ID_12vs_6, ID_13vs_6, ID_14vs_6, ID_15vs_6, ID_16vs_6],\n",
        "    r'$\\mathit{M. fascicularis}$': [ID_1vs_7, ID_2vs_7, ID_3vs_7, ID_4vs_7, ID_5vs_7, ID_6vs_7, ID_7vs_7, ID_8vs_7, ID_9vs_7, ID_10vs_7, ID_11vs_7, ID_12vs_7, ID_13vs_7, ID_14vs_7, ID_15vs_7, ID_16vs_7],\n",
        "    r'$\\mathit{M. mulatta}$': [ID_1vs_8, ID_2vs_8, ID_3vs_8, ID_4vs_8, ID_5vs_8, ID_6vs_8, ID_7vs_8, ID_8vs_8, ID_9vs_8, ID_10vs_8, ID_11vs_8, ID_12vs_8, ID_13vs_8, ID_14vs_8, ID_15vs_8, ID_16vs_8],\n",
        "    r'$\\mathit{M. thibetana}$': [ID_1vs_9, ID_2vs_9, ID_3vs_9, ID_4vs_9, ID_5vs_9, ID_6vs_9, ID_7vs_9, ID_8vs_9, ID_9vs_9, ID_10vs_9, ID_11vs_9, ID_12vs_9, ID_13vs_9, ID_14vs_9, ID_15vs_9, ID_16vs_9],\n",
        "    r'$\\mathit{R. roxellana}$': [ID_1vs_10, ID_2vs_10, ID_3vs_10, ID_4vs_10, ID_5vs_10, ID_6vs_10, ID_7vs_10, ID_8vs_10, ID_9vs_10, ID_10vs_10, ID_11vs_10, ID_12vs_10, ID_13vs_10, ID_14vs_10, ID_15vs_10, ID_16vs_10],\n",
        "    r'$\\mathit{T. francoisi}$': [ID_1vs_11, ID_2vs_11, ID_3vs_11, ID_4vs_11, ID_5vs_11, ID_6vs_11, ID_7vs_11, ID_8vs_11, ID_9vs_11, ID_10vs_11, ID_11vs_11, ID_12vs_11, ID_13vs_11, ID_14vs_11, ID_15vs_11, ID_16vs_11],\n",
        "    r'$\\mathit{N. leucogenys}$':[ID_1vs_12, ID_2vs_12, ID_3vs_12, ID_4vs_12, ID_5vs_12, ID_6vs_12, ID_7vs_12, ID_8vs_12, ID_9vs_12, ID_10vs_12, ID_11vs_12, ID_12vs_12, ID_13vs_12, ID_14vs_12, ID_15vs_12, ID_16vs_12],\n",
        "    r'$\\mathit{S. syndactylus}$': [ID_1vs_13, ID_2vs_13, ID_3vs_13, ID_4vs_13, ID_5vs_13, ID_6vs_13, ID_7vs_13, ID_8vs_13, ID_9vs_13, ID_10vs_13, ID_11vs_13, ID_12vs_13, ID_13vs_13, ID_14vs_13, ID_15vs_13, ID_16vs_13],\n",
        "    r'$\\mathit{M. murinus}$':  [ID_1vs_14, ID_2vs_14, ID_3vs_14, ID_4vs_14, ID_5vs_14, ID_6vs_14, ID_7vs_14, ID_8vs_14, ID_9vs_14, ID_10vs_14, ID_11vs_14, ID_12vs_14, ID_13vs_14, ID_14vs_14, ID_15vs_14, ID_16vs_14],\n",
        "    r'$\\mathit{C. jacchus}$': [ID_1vs_15, ID_2vs_15, ID_3vs_15, ID_4vs_15, ID_5vs_15, ID_6vs_15, ID_7vs_15, ID_8vs_15, ID_9vs_15, ID_10vs_15, ID_11vs_15, ID_12vs_15, ID_13vs_15, ID_14vs_15, ID_15vs_15, ID_16vs_15],\n",
        "    r'$\\mathit{L. catta}$': [ID_1vs_16, ID_2vs_16, ID_3vs_16, ID_4vs_16, ID_5vs_16, ID_6vs_16, ID_7vs_16, ID_8vs_16, ID_9vs_16, ID_10vs_16, ID_11vs_16, ID_12vs_16, ID_13vs_16, ID_14vs_16, ID_15vs_16, ID_16vs_16]\n",
        "}\n",
        "df = pd.DataFrame(data)\n",
        "df.set_index('SPECIES', inplace=True)\n",
        "\n",
        "# Set the size of the figure\n",
        "plt.figure(figsize=(30, 8))\n",
        "\n",
        "# Create a line chart with unique markers for each species\n",
        "markers = ['o', 's', '^', 'v', 'D', 'p', 'h', '+', 'x', '<', '>', '*', '|', '_', '.', ',']\n",
        "for i, species in enumerate(df.index):\n",
        "    plt.plot(df.columns, df.loc[species], label=species, marker=markers[i], markersize=8, linestyle='-', linewidth=1)\n",
        "\n",
        "# Set title and tags\n",
        "plt.title(r'General Comparison of the Percentage Difference in the PRKN Gene Among Primates (IMAGE VS IMAGE) / 7-mer ', fontsize=16)\n",
        "plt.xlabel('', fontsize=14)\n",
        "plt.ylabel(' % Difference', fontsize=14)\n",
        "\n",
        "# Show legend\n",
        "plt.legend(loc='upper right', bbox_to_anchor=(1.12, 1), fontsize='x-large')\n",
        "\n",
        "# Rotate x-axis labels for better readability\n",
        "plt.xticks(rotation=45)\n",
        "\n",
        "output_path = \"/path/to/output/directory\"\n",
        "#Save the image to the specified path\n",
        "plt.savefig(os.path.join(output_path, 'Differences between PRKN of primate species_7-mer.png'), dpi=500, bbox_inches='tight')\n",
        "\n",
        "# Show the graph\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "Ot83Dhnx_65Z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "# Define the data matrix\n",
        "data = [\n",
        "    [ID_1vs_1, ID_2vs_1, ID_3vs_1, ID_4vs_1, ID_5vs_1, ID_6vs_1, ID_7vs_1, ID_8vs_1, ID_9vs_1, ID_10vs_1, ID_11vs_1, ID_12vs_1, ID_13vs_1, ID_14vs_1, ID_15vs_1, ID_16vs_1],\n",
        "    [ID_1vs_2, ID_2vs_2, ID_3vs_2, ID_4vs_2, ID_5vs_2, ID_6vs_2, ID_7vs_2, ID_8vs_2, ID_9vs_2, ID_10vs_2, ID_11vs_2, ID_12vs_2, ID_13vs_2, ID_14vs_2, ID_15vs_2, ID_16vs_2],\n",
        "    [ID_1vs_3, ID_2vs_3, ID_3vs_3, ID_4vs_3, ID_5vs_3, ID_6vs_3, ID_7vs_3, ID_8vs_3, ID_9vs_3, ID_10vs_3, ID_11vs_3, ID_12vs_3, ID_13vs_3, ID_14vs_3, ID_15vs_3, ID_16vs_3],\n",
        "    [ID_1vs_4, ID_2vs_4, ID_3vs_4, ID_4vs_4, ID_5vs_4, ID_6vs_4, ID_7vs_4, ID_8vs_4, ID_9vs_4, ID_10vs_4, ID_11vs_4, ID_12vs_4, ID_13vs_4, ID_14vs_4, ID_15vs_4, ID_16vs_4],\n",
        "    [ID_1vs_5, ID_2vs_5, ID_3vs_5, ID_4vs_5, ID_5vs_5, ID_6vs_5, ID_7vs_5, ID_8vs_5, ID_9vs_5, ID_10vs_5, ID_11vs_5, ID_12vs_5, ID_13vs_5, ID_14vs_5, ID_15vs_5, ID_16vs_5],\n",
        "    [ID_1vs_6, ID_2vs_6, ID_3vs_6, ID_4vs_6, ID_5vs_6, ID_6vs_6, ID_7vs_6, ID_8vs_6, ID_9vs_6, ID_10vs_6, ID_11vs_6, ID_12vs_6, ID_13vs_6, ID_14vs_6, ID_15vs_6, ID_16vs_6],\n",
        "    [ID_1vs_7, ID_2vs_7, ID_3vs_7, ID_4vs_7, ID_5vs_7, ID_6vs_7, ID_7vs_7, ID_8vs_7, ID_9vs_7, ID_10vs_7, ID_11vs_7, ID_12vs_7, ID_13vs_7, ID_14vs_7, ID_15vs_7, ID_16vs_7],\n",
        "    [ID_1vs_8, ID_2vs_8, ID_3vs_8, ID_4vs_8, ID_5vs_8, ID_6vs_8, ID_7vs_8, ID_8vs_8, ID_9vs_8, ID_10vs_8, ID_11vs_8, ID_12vs_8, ID_13vs_8, ID_14vs_8, ID_15vs_8, ID_16vs_8],\n",
        "    [ID_1vs_9, ID_2vs_9, ID_3vs_9, ID_4vs_9, ID_5vs_9, ID_6vs_9, ID_7vs_9, ID_8vs_9, ID_9vs_9, ID_10vs_9, ID_11vs_9, ID_12vs_9, ID_13vs_9, ID_14vs_9, ID_15vs_9, ID_16vs_9],\n",
        "    [ID_1vs_10, ID_2vs_10, ID_3vs_10, ID_4vs_10, ID_5vs_10, ID_6vs_10, ID_7vs_10, ID_8vs_10, ID_9vs_10, ID_10vs_10, ID_11vs_10, ID_12vs_10, ID_13vs_10, ID_14vs_10, ID_15vs_10, ID_16vs_10],\n",
        "    [ID_1vs_11, ID_2vs_11, ID_3vs_11, ID_4vs_11, ID_5vs_11, ID_6vs_11, ID_7vs_11, ID_8vs_11, ID_9vs_11, ID_10vs_11, ID_11vs_11, ID_12vs_11, ID_13vs_11, ID_14vs_11, ID_15vs_11, ID_16vs_11],\n",
        "    [ID_1vs_12, ID_2vs_12, ID_3vs_12, ID_4vs_12, ID_5vs_12, ID_6vs_12, ID_7vs_12, ID_8vs_12, ID_9vs_12, ID_10vs_12, ID_11vs_12, ID_12vs_12, ID_13vs_12, ID_14vs_12, ID_15vs_12, ID_16vs_12],\n",
        "    [ID_1vs_13, ID_2vs_13, ID_3vs_13, ID_4vs_13, ID_5vs_13, ID_6vs_13, ID_7vs_13, ID_8vs_13, ID_9vs_13, ID_10vs_13, ID_11vs_13, ID_12vs_13, ID_13vs_13, ID_14vs_13, ID_15vs_13, ID_16vs_13],\n",
        "    [ID_1vs_14, ID_2vs_14, ID_3vs_14, ID_4vs_14, ID_5vs_14, ID_6vs_14, ID_7vs_14, ID_8vs_14, ID_9vs_14, ID_10vs_14, ID_11vs_14, ID_12vs_14, ID_13vs_14, ID_14vs_14, ID_15vs_14, ID_16vs_14],\n",
        "    [ID_1vs_15, ID_2vs_15, ID_3vs_15, ID_4vs_15, ID_5vs_15, ID_6vs_15, ID_7vs_15, ID_8vs_15, ID_9vs_15, ID_10vs_15, ID_11vs_15, ID_12vs_15, ID_13vs_15, ID_14vs_15, ID_15vs_15, ID_16vs_15],\n",
        "    [ID_1vs_16, ID_2vs_16, ID_3vs_16, ID_4vs_16, ID_5vs_16, ID_6vs_16, ID_7vs_16, ID_8vs_16, ID_9vs_16, ID_10vs_16, ID_11vs_16, ID_12vs_16, ID_13vs_16, ID_14vs_16, ID_15vs_16, ID_16vs_16]\n",
        "]\n",
        "\n",
        "# Convert the matrix to a numpy array for visualization\n",
        "matrix_data = np.array(data)\n",
        "\n",
        "# Get row and column labels\n",
        "species_labels = [\"\"] + [r'$\\mathit{H. sapiens}$', r'$\\mathit{G. gorilla}$', r'$\\mathit{P. paniscus}$', r'$\\mathit{P. troglodytes}$',\n",
        "                   r'$\\mathit{P. abelii}$', r'$\\mathit{P. pygmaeus}$', r'$\\mathit{M. fascicularis}$', r'$\\mathit{M. mulatta}$', r'$\\mathit{M. thibetana}$',\n",
        "                   r'$\\mathit{R. roxellana}$', r'$\\mathit{T. francoisi}$', r'$\\mathit{N. leucogenys}$', r'$\\mathit{S. syndactylus}$',\n",
        "                   r'$\\mathit{M. murinus}$', r'$\\mathit{C. jacchus}$', r'$\\mathit{L. catta}$']\n",
        "\n",
        "species_ls = [r'$\\mathit{H. sapiens}$', r'$\\mathit{G. gorilla}$', r'$\\mathit{P. paniscus}$', r'$\\mathit{P. troglodytes}$',\n",
        "                   r'$\\mathit{P. abelii}$', r'$\\mathit{P. pygmaeus}$', r'$\\mathit{M. fascicularis}$', r'$\\mathit{M. mulatta}$', r'$\\mathit{M. thibetana}$',\n",
        "                   r'$\\mathit{R. roxellana}$', r'$\\mathit{T. francoisi}$', r'$\\mathit{N. leucogenys}$', r'$\\mathit{S. syndactylus}$',\n",
        "                   r'$\\mathit{M. murinus}$', r'$\\mathit{C. jacchus}$', r'$\\mathit{L. catta}$']\n",
        "\n",
        "# Create the visualization\n",
        "plt.figure(figsize=(10, 8))\n",
        "plt.imshow(matrix_data, cmap='hot', interpolation='none')\n",
        "plt.colorbar(label='% Difference')\n",
        "\n",
        "# Adjust the positions of the ticks\n",
        "tick_positions = np.arange(len(species_ls))\n",
        "plt.xticks(tick_positions, species_ls, rotation=90)\n",
        "plt.yticks(tick_positions, species_ls)\n",
        "plt.title(r'Difference for the $\\mathit{PRKN}$ gene in primates (IMAGE VS IMAGE) ')\n",
        "\n",
        "# Define the output path\n",
        "output_path = \"/path/to/output/directory\"\n",
        "# Save the image to the specified path\n",
        "plt.savefig(os.path.join(output_path, 'Difference_IMAGE_VS_IMAGE_among_primate_PRKN.png'), dpi=800, bbox_inches='tight')\n",
        "\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "Og2obbUB_ttb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Create a data dictionary\n",
        "data = {\n",
        "    'SPECIES': [r'$\\mathit{H. sapiens}$', r'$\\mathit{G. gorilla}$', r'$\\mathit{P. paniscus}$', r'$\\mathit{P. troglodytes}$',\n",
        "                r'$\\mathit{P. abelii}$', r'$\\mathit{P. pygmaeus}$', r'$\\mathit{M. fascicularis}$', r'$\\mathit{M. mulatta}$', r'$\\mathit{M. thibetana}$',\n",
        "                r'$\\mathit{R. roxellana}$', r'$\\mathit{T. francoisi}$', r'$\\mathit{N. leucogenys}$', r'$\\mathit{S. syndactylus}$',\n",
        "                r'$\\mathit{M. murinus}$', r'$\\mathit{C. jacchus}$', r'$\\mathit{L. catta}$'],\n",
        "    '1': [ID_1vs_1, ID_2vs_1, ID_3vs_1, ID_4vs_1, ID_5vs_1, ID_6vs_1, ID_7vs_1, ID_8vs_1, ID_9vs_1, ID_10vs_1, ID_11vs_1, ID_12vs_1, ID_13vs_1, ID_14vs_1, ID_15vs_1, ID_16vs_1],\n",
        "    '2': [ID_1vs_2, ID_2vs_2, ID_3vs_2, ID_4vs_2, ID_5vs_2, ID_6vs_2, ID_7vs_2, ID_8vs_2, ID_9vs_2, ID_10vs_2, ID_11vs_2, ID_12vs_2, ID_13vs_2, ID_14vs_2, ID_15vs_2, ID_16vs_2],\n",
        "    '3': [ID_1vs_3, ID_2vs_3, ID_3vs_3, ID_4vs_3, ID_5vs_3, ID_6vs_3, ID_7vs_3, ID_8vs_3, ID_9vs_3, ID_10vs_3, ID_11vs_3, ID_12vs_3, ID_13vs_3, ID_14vs_3, ID_15vs_3, ID_16vs_3],\n",
        "    '4': [ID_1vs_4, ID_2vs_4, ID_3vs_4, ID_4vs_4, ID_5vs_4, ID_6vs_4, ID_7vs_4, ID_8vs_4, ID_9vs_4, ID_10vs_4, ID_11vs_4, ID_12vs_4, ID_13vs_4, ID_14vs_4, ID_15vs_4, ID_16vs_4],\n",
        "    '5': [ID_1vs_5, ID_2vs_5, ID_3vs_5, ID_4vs_5, ID_5vs_5, ID_6vs_5, ID_7vs_5, ID_8vs_5, ID_9vs_5, ID_10vs_5, ID_11vs_5, ID_12vs_5, ID_13vs_5, ID_14vs_5, ID_15vs_5, ID_16vs_5],\n",
        "    '6': [ID_1vs_6, ID_2vs_6, ID_3vs_6, ID_4vs_6, ID_5vs_6, ID_6vs_6, ID_7vs_6, ID_8vs_6, ID_9vs_6, ID_10vs_6, ID_11vs_6, ID_12vs_6, ID_13vs_6, ID_14vs_6, ID_15vs_6, ID_16vs_6],\n",
        "    '7': [ID_1vs_7, ID_2vs_7, ID_3vs_7, ID_4vs_7, ID_5vs_7, ID_6vs_7, ID_7vs_7, ID_8vs_7, ID_9vs_7, ID_10vs_7, ID_11vs_7, ID_12vs_7, ID_13vs_7, ID_14vs_7, ID_15vs_7, ID_16vs_7],\n",
        "    '8': [ID_1vs_8, ID_2vs_8, ID_3vs_8, ID_4vs_8, ID_5vs_8, ID_6vs_8, ID_7vs_8, ID_8vs_8, ID_9vs_8, ID_10vs_8, ID_11vs_8, ID_12vs_8, ID_13vs_8, ID_14vs_8, ID_15vs_8, ID_16vs_8],\n",
        "    '9': [ID_1vs_9, ID_2vs_9, ID_3vs_9, ID_4vs_9, ID_5vs_9, ID_6vs_9, ID_7vs_9, ID_8vs_9, ID_9vs_9, ID_10vs_9, ID_11vs_9, ID_12vs_9, ID_13vs_9, ID_14vs_9, ID_15vs_9, ID_16vs_9],\n",
        "    '10': [ID_1vs_10, ID_2vs_10, ID_3vs_10, ID_4vs_10, ID_5vs_10, ID_6vs_10, ID_7vs_10, ID_8vs_10, ID_9vs_10, ID_10vs_10, ID_11vs_10, ID_12vs_10, ID_13vs_10, ID_14vs_10, ID_15vs_10, ID_16vs_10],\n",
        "    '11': [ID_1vs_11, ID_2vs_11, ID_3vs_11, ID_4vs_11, ID_5vs_11, ID_6vs_11, ID_7vs_11, ID_8vs_11, ID_9vs_11, ID_10vs_11, ID_11vs_11, ID_12vs_11, ID_13vs_11, ID_14vs_11, ID_15vs_11, ID_16vs_11],\n",
        "    '12': [ID_1vs_12, ID_2vs_12, ID_3vs_12, ID_4vs_12, ID_5vs_12, ID_6vs_12, ID_7vs_12, ID_8vs_12, ID_9vs_12, ID_10vs_12, ID_11vs_12, ID_12vs_12, ID_13vs_12, ID_14vs_12, ID_15vs_12, ID_16vs_12],\n",
        "    '13': [ID_1vs_13, ID_2vs_13, ID_3vs_13, ID_4vs_13, ID_5vs_13, ID_6vs_13, ID_7vs_13, ID_8vs_13, ID_9vs_13, ID_10vs_13, ID_11vs_13, ID_12vs_13, ID_13vs_13, ID_14vs_13, ID_15vs_13, ID_16vs_13],\n",
        "    '14': [ID_1vs_14, ID_2vs_14, ID_3vs_14, ID_4vs_14, ID_5vs_14, ID_6vs_14, ID_7vs_14, ID_8vs_14, ID_9vs_14, ID_10vs_14, ID_11vs_14, ID_12vs_14, ID_13vs_14, ID_14vs_14, ID_15vs_14, ID_16vs_14],\n",
        "    '15': [ID_1vs_15, ID_2vs_15, ID_3vs_15, ID_4vs_15, ID_5vs_15, ID_6vs_15, ID_7vs_15, ID_8vs_15, ID_9vs_15, ID_10vs_15, ID_11vs_15, ID_12vs_15, ID_13vs_15, ID_14vs_15, ID_15vs_15, ID_16vs_15],\n",
        "    '16': [ID_1vs_16, ID_2vs_16, ID_3vs_16, ID_4vs_16, ID_5vs_16, ID_6vs_16, ID_7vs_16, ID_8vs_16, ID_9vs_16, ID_10vs_16, ID_11vs_16, ID_12vs_16, ID_13vs_16, ID_14vs_16, ID_15vs_16, ID_16vs_16]\n",
        "}\n",
        "\n",
        "# Create a DataFrame from the dictionary\n",
        "df = pd.DataFrame(data)\n",
        "\n",
        "# Round the values in the DataFrame to two decimal places\n",
        "df = df.round(2)\n",
        "\n",
        "# Set 'SPECIES' as the index\n",
        "df.set_index('SPECIES', inplace=True)\n",
        "\n",
        "# Set the figure size\n",
        "plt.figure(figsize=(24, 16))\n",
        "\n",
        "# Create a heatmap with seaborn\n",
        "sns.heatmap(df, annot=True, cmap='hot', linewidths=.5, fmt='g')\n",
        "\n",
        "# Set the title and labels\n",
        "plt.title('Difference for the $\\mathit{PRKN}$ gene in primates (IMAGE VS IMAGE) / 7-MER', fontsize=19)\n",
        "plt.xlabel('', fontsize=14)\n",
        "plt.ylabel('', fontsize=14)\n",
        "\n",
        "output_path = \"/path/to/output/directory\"\n",
        "# Save the image to the specified path\n",
        "plt.savefig(os.path.join(output_path, 'Difference_IMAGE_VS_IMAGE_among_Species.png'), dpi=500, bbox_inches='tight')\n",
        "\n",
        "# Show the heatmap\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "m8tCw7J--75Z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Software Matrix Difference - CODE Numerical Matrix Difference Analyzer (*NMDA*)**\n"
      ],
      "metadata": {
        "id": "oCoB6-_bfa6J"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "\n",
        "def diff_matrix(a, b, a_ref):\n",
        "    a = np.array(a)\n",
        "    b = np.array(b)\n",
        "\n",
        "    absolute_difference = np.abs(a - b)\n",
        "    percentage_difference = (np.sum(absolute_difference) / np.sum(a_ref)) * 100\n",
        "\n",
        "    # Rounding the result to 2 decimal places\n",
        "    rounded_percentage_difference = round(percentage_difference, 2)\n",
        "\n",
        "    return rounded_percentage_difference\n",
        "\n",
        "matrices_to_compare = [matrix_1, matrix_2, matrix_3, matrix_4, matrix_5,\n",
        "                       matrix_6, matrix_7, matrix_8, matrix_9, matrix_10,\n",
        "                       matrix_11, matrix_12, matrix_13, matrix_14, matrix_15, matrix_16\n",
        "                      ]\n",
        "\n",
        "# Creating a matrix to store the results\n",
        "result_matrix = np.zeros((len(matrices_to_compare), len(matrices_to_compare)))\n",
        "\n",
        "# n = matrices_to_compare.size\n",
        "a_ref = matrices_to_compare[0]\n",
        "\n",
        "# Creating a dictionary to store the individual results\n",
        "result_dict = {}\n",
        "\n",
        "# Filling the result matrix and dictionary\n",
        "for i, matrix1 in enumerate(matrices_to_compare):\n",
        "    for j, matrix2 in enumerate(matrices_to_compare):\n",
        "        result = diff_matrix(matrix1, matrix2, a_ref)\n",
        "        result_matrix[i, j] = result\n",
        "        result_dict[f\"MS_{i + 1}_{j + 1}\"] = result\n",
        "\n",
        "# Printing the matrix with labels\n",
        "for i in range(result_matrix.shape[0]):\n",
        "    for j in range(result_matrix.shape[1]):\n",
        "        print(f\"MS_{i + 1}_{j + 1}: {result_matrix[i, j]}\", end='\\t')\n",
        "    print()\n",
        "\n",
        "# Accessing individual variables using the dictionary\n",
        "print(\"\\nAccessing individual variables:\")\n",
        "for key, value in result_dict.items():\n",
        "    print(f\"{key}: {value}\")\n",
        "\n",
        "# Printing and storing the difference percentages\n",
        "for i, matrix1 in enumerate(matrices_to_compare):\n",
        "    for j, matrix2 in enumerate(matrices_to_compare):\n",
        "        percentage_diff = diff_matrix(matrix1, matrix2, a_ref)\n",
        "        variable_name = f\"MS_{i + 1}_{j + 1}\"\n",
        "        globals()[variable_name] = percentage_diff\n",
        "        print(f\"{variable_name} = {percentage_diff:.2f}\")\n",
        "\n",
        "# Accessing individual variables\n",
        "print(\"\\nAccessing individual variables:\")\n",
        "for i in range(len(matrices_to_compare)):\n",
        "    for j in range(len(matrices_to_compare)):\n",
        "        variable_name = f\"MS_{i + 1}_{j + 1}\"\n",
        "        print(f\"{variable_name} = {globals()[variable_name]:.2f}\")\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "MJEVjTs5JfdL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Define the data matrix\n",
        "data = [\n",
        "    [D_1_1, D_2_1, D_3_1, D_4_1, D_5_1, D_6_1, D_7_1, D_8_1, D_9_1, D_10_1, D_11_1, D_12_1, D_13_1, D_14_1, D_15_1, D_16_1],\n",
        "    [D_1_2, D_2_2, D_3_2, D_4_2, D_5_2, D_6_2, D_7_2, D_8_2, D_9_2, D_10_2, D_11_2, D_12_2, D_13_2, D_14_2, D_15_2, D_16_2],\n",
        "    [D_1_3, D_2_3, D_3_3, D_4_3, D_5_3, D_6_3, D_7_3, D_8_3, D_9_3, D_10_3, D_11_3, D_12_3, D_13_3, D_14_3, D_15_3, D_16_3],\n",
        "    [D_1_4, D_2_4, D_3_4, D_4_4, D_5_4, D_6_4, D_7_4, D_8_4, D_9_4, D_10_4, D_11_4, D_12_4, D_13_4, D_14_4, D_15_4, D_16_4],\n",
        "    [D_1_5, D_2_5, D_3_5, D_4_5, D_5_5, D_6_5, D_7_5, D_8_5, D_9_5, D_10_5, D_11_5, D_12_5, D_13_5, D_14_5, D_15_5, D_16_5],\n",
        "    [D_1_6, D_2_6, D_3_6, D_4_6, D_5_6, D_6_6, D_7_6, D_8_6, D_9_6, D_10_6, D_11_6, D_12_6, D_13_6, D_14_6, D_15_6, D_16_6],\n",
        "    [D_1_7, D_2_7, D_3_7, D_4_7, D_5_7, D_6_7, D_7_7, D_8_7, D_9_7, D_10_7, D_11_7, D_12_7, D_13_7, D_14_7, D_15_7, D_16_7],\n",
        "    [D_1_8, D_2_8, D_3_8, D_4_8, D_5_8, D_6_8, D_7_8, D_8_8, D_9_8, D_10_8, D_11_8, D_12_8, D_13_8, D_14_8, D_15_8, D_16_8],\n",
        "    [D_1_9, D_2_9, D_3_9, D_4_9, D_5_9, D_6_9, D_7_9, D_8_9, D_9_9, D_10_9, D_11_9, D_12_9, D_13_9, D_14_9, D_15_9, D_16_9],\n",
        "    [D_1_10, D_2_10, D_3_10, D_4_10, D_5_10, D_6_10, D_7_10, D_8_10, D_9_10, D_10_10, D_11_10, D_12_10, D_13_10, D_14_10, D_15_10, D_16_10],\n",
        "    [D_1_11, D_2_11, D_3_11, D_4_11, D_5_11, D_6_11, D_7_11, D_8_11, D_9_11, D_10_11, D_11_11, D_12_11, D_13_11, D_14_11, D_15_11, D_16_11],\n",
        "    [D_1_12, D_2_12, D_3_12, D_4_12, D_5_12, D_6_12, D_7_12, D_8_12, D_9_12, D_10_12, D_11_12, D_12_12, D_13_12, D_14_12, D_15_12, D_16_12],\n",
        "    [D_1_13, D_2_13, D_3_13, D_4_13, D_5_13, D_6_13, D_7_13, D_8_13, D_9_13, D_10_13, D_11_13, D_12_13, D_13_13, D_14_13, D_15_13, D_16_13],\n",
        "    [D_1_14, D_2_14, D_3_14, D_4_14, D_5_14, D_6_14, D_7_14, D_8_14, D_9_14, D_10_14, D_11_14, D_12_14, D_13_14, D_14_14, D_15_14, D_16_14],\n",
        "    [D_1_15, D_2_15, D_3_15, D_4_15, D_5_15, D_6_15, D_7_15, D_8_15, D_9_15, D_10_15, D_11_15, D_12_15, D_13_15, D_14_15, D_15_15, D_16_15],\n",
        "    [D_1_16, D_2_16, D_3_16, D_4_16, D_5_16, D_6_16, D_7_16, D_8_16, D_9_16, D_10_16, D_11_16, D_12_16, D_13_16, D_14_16, D_15_16, D_16_16]\n",
        "]\n",
        "\n",
        "# Convert the matrix to a numpy array for visualization\n",
        "matrix_data = np.array(data)\n",
        "\n",
        "# Get row and column labels\n",
        "labels = [f'Species_{i}' for i in range(1, 17)]\n",
        "\n",
        "# Create DataFrame\n",
        "df = pd.DataFrame(data, columns=labels, index=labels)\n",
        "\n",
        "# Round DataFrame values to two decimal places\n",
        "df = df.round(2)\n",
        "\n",
        "# Create the table image\n",
        "plt.figure(figsize=(24, 6.5))\n",
        "table = plt.table(cellText=df.values, colLabels=df.columns, rowLabels=df.index, loc='center', cellLoc='center', colColours=plt.cm.BuPu(np.full(len(df.columns), 0.1)))\n",
        "table.auto_set_font_size(False)\n",
        "table.set_fontsize(10)\n",
        "table.scale(1, 1.5)  # Adjust scale for better visualization\n",
        "\n",
        "labels = df.index.tolist()\n",
        "plt.title('Percentage Difference Table for Gene Comparison (matrix vs matrix) / 7-mer', fontsize=19)\n",
        "plt.xticks(np.arange(len(labels)), labels, rotation=45, ha='right', rotation_mode='anchor')\n",
        "plt.yticks(np.arange(len(labels)), labels)\n",
        "# Hide axes\n",
        "plt.axis('off')\n",
        "\n",
        "output_path = \"/path/to/save/location/\"\n",
        "# Save the image to the specified path\n",
        "plt.savefig(os.path.join(output_path, 'generalized_table.png'), dpi=1200, bbox_inches='tight')\n",
        "\n",
        "# Show the image\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "S6A_B0l2t31h"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import os\n",
        "\n",
        "# Define the data matrix\n",
        "data = [\n",
        "    [D_1_1, D_2_1, D_3_1, D_4_1, D_5_1, D_6_1, D_7_1, D_8_1, D_9_1, D_10_1, D_11_1, D_12_1, D_13_1, D_14_1, D_15_1, D_16_1 ],\n",
        "    [D_1_2, D_2_2, D_3_2, D_4_2, D_5_2, D_6_2, D_7_2, D_8_2, D_9_2, D_10_2, D_11_2, D_12_2, D_13_2, D_14_2, D_15_2, D_16_2 ],\n",
        "    [D_1_3, D_2_3, D_3_3, D_4_3, D_5_3, D_6_3, D_7_3, D_8_3, D_9_3, D_10_3, D_11_3, D_12_3, D_13_3, D_14_3, D_15_3, D_16_3 ],\n",
        "    [D_1_4, D_2_4, D_3_4, D_4_4, D_5_4, D_6_4, D_7_4, D_8_4, D_9_4, D_10_4, D_11_4, D_12_4, D_13_4, D_14_4, D_15_4, D_16_4 ],\n",
        "    [D_1_5, D_2_5, D_3_5, D_4_5, D_5_5, D_6_5, D_7_5, D_8_5, D_9_5, D_10_5, D_11_5, D_12_5, D_13_5, D_14_5, D_15_5, D_16_5 ],\n",
        "    [D_1_6, D_2_6, D_3_6, D_4_6, D_5_6, D_6_6, D_7_6, D_8_6, D_9_6, D_10_6, D_11_6, D_12_6, D_13_6, D_14_6, D_15_6, D_16_6 ],\n",
        "    [D_1_7, D_2_7, D_3_7, D_4_7, D_5_7, D_6_7, D_7_7, D_8_7, D_9_7, D_10_7, D_11_7, D_12_7, D_13_7, D_14_7, D_15_7, D_16_7 ],\n",
        "    [D_1_8, D_2_8, D_3_8, D_4_8, D_5_8, D_6_8, D_7_8, D_8_8, D_9_8, D_10_8, D_11_8, D_12_8, D_13_8, D_14_8, D_15_8, D_16_8 ],\n",
        "    [D_1_9, D_2_9, D_3_9, D_4_9, D_5_9, D_6_9, D_7_9, D_8_9, D_9_9, D_10_9, D_11_9, D_12_9, D_13_9, D_14_9, D_15_9, D_16_9 ],\n",
        "    [D_1_10, D_2_10, D_3_10, D_4_10, D_5_10, D_6_10, D_7_10, D_8_10, D_9_10, D_10_10, D_11_10, D_12_10, D_13_10, D_14_10, D_15_10, D_16_10 ],\n",
        "    [D_1_11, D_2_11, D_3_11, D_4_11, D_5_11, D_6_11, D_7_11, D_8_11, D_9_11, D_10_11, D_11_11, D_12_11, D_13_11, D_14_11, D_15_11, D_16_11 ],\n",
        "    [D_1_12, D_2_12, D_3_12, D_4_12, D_5_12, D_6_12, D_7_12, D_8_12, D_9_12, D_10_12, D_11_12, D_12_12, D_13_12, D_14_12, D_15_12, D_16_12 ],\n",
        "    [D_1_13, D_2_13, D_3_13, D_4_13, D_5_13, D_6_13, D_7_13, D_8_13, D_9_13, D_10_13, D_11_13, D_12_13, D_13_13, D_14_13, D_15_13, D_16_13 ],\n",
        "    [D_1_14, D_2_14, D_3_14, D_4_14, D_5_14, D_6_14, D_7_14, D_8_14, D_9_14, D_10_14, D_11_14, D_12_14, D_13_14, D_14_14, D_15_14, D_16_14 ],\n",
        "    [D_1_15, D_2_15, D_3_15, D_4_15, D_5_15, D_6_15, D_7_15, D_8_15, D_9_15, D_10_15, D_11_15, D_12_15, D_13_15, D_14_15, D_15_15, D_16_15 ],\n",
        "    [D_1_16, D_2_16, D_3_16, D_4_16, D_5_16, D_6_16, D_7_16, D_8_16, D_9_16, D_10_16, D_11_16, D_12_16, D_13_16, D_14_16, D_15_16, D_16_16 ]\n",
        "]\n",
        "\n",
        "# Convert the matrix to a numpy array for visualization\n",
        "matrix_data = np.array(data)\n",
        "\n",
        "# Obtain row and column labels\n",
        "labels = [\"\"] + [r'$\\mathit{Label_1}$', r'$\\mathit{Label_2}$', r'$\\mathit{Label_3}$',  r'$\\mathit{Label_4}$',\n",
        "                r'$\\mathit{Label_5}$', r'$\\mathit{Label_6}$', r'$\\mathit{Label_7}$', r'$\\mathit{Label_8}$', r'$\\mathit{Label_9}$',\n",
        "                r'$\\mathit{Label_10}$', r'$\\mathit{Label_11}$', r'$\\mathit{Label_12}$', r'$\\mathit{Label_13}$',\n",
        "                r'$\\mathit{Label_14}$', r'$\\mathit{Label_15}$', r'$\\mathit{Label_16}$']\n",
        "\n",
        "# Row and column labels\n",
        "species_ls = [r'$\\mathit{Label_1}$', r'$\\mathit{Label_2}$', r'$\\mathit{Label_3}$',  r'$\\mathit{Label_4}$',\n",
        "              r'$\\mathit{Label_5}$', r'$\\mathit{Label_6}$', r'$\\mathit{Label_7}$', r'$\\mathit{Label_8}$', r'$\\mathit{Label_9}$',\n",
        "              r'$\\mathit{Label_10}$', r'$\\mathit{Label_11}$', r'$\\mathit{Label_12}$', r'$\\mathit{Label_13}$',\n",
        "              r'$\\mathit{Label_14}$', r'$\\mathit{Label_15}$', r'$\\mathit{Label_16}$']\n",
        "\n",
        "# Create the visualization\n",
        "plt.figure(figsize=(10, 8))\n",
        "plt.imshow(matrix_data, cmap='hot', interpolation='none')\n",
        "plt.colorbar(label='% Difference')\n",
        "\n",
        "# Adjust the positions of the ticks\n",
        "tick_positions = np.arange(len(species_ls))\n",
        "plt.xticks(tick_positions, species_ls, rotation=90)\n",
        "plt.yticks(tick_positions, species_ls)\n",
        "plt.title(r'Difference for the $\\mathit{GENE}$ gene in species (matrix vs matrix) / 7-mer')\n",
        "\n",
        "# Define the output path\n",
        "output_path = \"/path/to/output/directory/\"\n",
        "# Save the image to the specified path\n",
        "plt.savefig(os.path.join(output_path, 'Difference_matrix_among_species_GENE.png'), dpi=400, bbox_inches='tight')\n",
        "\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "19P7awO_wFk8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Create a data dictionary\n",
        "data = {\n",
        "    'SPECIES': [r'$\\mathit{Species1}$', r'$\\mathit{Species2}$', r'$\\mathit{Species3}$', r'$\\mathit{Species4}$',\n",
        "                r'$\\mathit{Species5}$', r'$\\mathit{Species6}$', r'$\\mathit{Species7}$', r'$\\mathit{Species8}$', r'$\\mathit{Species9}$',\n",
        "                r'$\\mathit{Species10}$', r'$\\mathit{Species11}$', r'$\\mathit{Species12}$', r'$\\mathit{Species13}$',\n",
        "                r'$\\mathit{Species14}$', r'$\\mathit{Species15}$', r'$\\mathit{Species16}$'],\n",
        "    '1': [Value_1_1, Value_2_1, Value_3_1, Value_4_1, Value_5_1, Value_6_1, Value_7_1, Value_8_1, Value_9_1, Value_10_1, Value_11_1, Value_12_1, Value_13_1, Value_14_1, Value_15_1, Value_16_1],\n",
        "    '2': [Value_1_2, Value_2_2, Value_3_2, Value_4_2, Value_5_2, Value_6_2, Value_7_2, Value_8_2, Value_9_2, Value_10_2, Value_11_2, Value_12_2, Value_13_2, Value_14_2, Value_15_2, Value_16_2],\n",
        "    '3': [Value_1_3, Value_2_3, Value_3_3, Value_4_3, Value_5_3, Value_6_3, Value_7_3, Value_8_3, Value_9_3, Value_10_3, Value_11_3, Value_12_3, Value_13_3, Value_14_3, Value_15_3, Value_16_3],\n",
        "    '4': [Value_1_4, Value_2_4, Value_3_4, Value_4_4, Value_5_4, Value_6_4, Value_7_4, Value_8_4, Value_9_4, Value_10_4, Value_11_4, Value_12_4, Value_13_4, Value_14_4, Value_15_4, Value_16_4],\n",
        "    '5': [Value_1_5, Value_2_5, Value_3_5, Value_4_5, Value_5_5, Value_6_5, Value_7_5, Value_8_5, Value_9_5, Value_10_5, Value_11_5, Value_12_5, Value_13_5, Value_14_5, Value_15_5, Value_16_5],\n",
        "    '6': [Value_1_6, Value_2_6, Value_3_6, Value_4_6, Value_5_6, Value_6_6, Value_7_6, Value_8_6, Value_9_6, Value_10_6, Value_11_6, Value_12_6, Value_13_6, Value_14_6, Value_15_6, Value_16_6],\n",
        "    '7': [Value_1_7, Value_2_7, Value_3_7, Value_4_7, Value_5_7, Value_6_7, Value_7_7, Value_8_7, Value_9_7, Value_10_7, Value_11_7, Value_12_7, Value_13_7, Value_14_7, Value_15_7, Value_16_7],\n",
        "    '8': [Value_1_8, Value_2_8, Value_3_8, Value_4_8, Value_5_8, Value_6_8, Value_7_8, Value_8_8, Value_9_8, Value_10_8, Value_11_8, Value_12_8, Value_13_8, Value_14_8, Value_15_8, Value_16_8],\n",
        "    '9': [Value_1_9, Value_2_9, Value_3_9, Value_4_9, Value_5_9, Value_6_9, Value_7_9, Value_8_9, Value_9_9, Value_10_9, Value_11_9, Value_12_9, Value_13_9, Value_14_9, Value_15_9, Value_16_9],\n",
        "    '10': [Value_1_10, Value_2_10, Value_3_10, Value_4_10, Value_5_10, Value_6_10, Value_7_10, Value_8_10, Value_9_10, Value_10_10, Value_11_10, Value_12_10, Value_13_10, Value_14_10, Value_15_10, Value_16_10],\n",
        "    '11': [Value_1_11, Value_2_11, Value_3_11, Value_4_11, Value_5_11, Value_6_11, Value_7_11, Value_8_11, Value_9_11, Value_10_11, Value_11_11, Value_12_11, Value_13_11, Value_14_11, Value_15_11, Value_16_11],\n",
        "    '12': [Value_1_12, Value_2_12, Value_3_12, Value_4_12, Value_5_12, Value_6_12, Value_7_12, Value_8_12, Value_9_12, Value_10_12, Value_11_12, Value_12_12, Value_13_12, Value_14_12, Value_15_12, Value_16_12],\n",
        "    '13': [Value_1_13, Value_2_13, Value_3_13, Value_4_13, Value_5_13, Value_6_13, Value_7_13, Value_8_13, Value_9_13, Value_10_13, Value_11_13, Value_12_13, Value_13_13, Value_14_13, Value_15_13, Value_16_13],\n",
        "    '14': [Value_1_14, Value_2_14, Value_3_14, Value_4_14, Value_5_14, Value_6_14, Value_7_14, Value_8_14, Value_9_14, Value_10_14, Value_11_14, Value_12_14, Value_13_14, Value_14_14, Value_15_14, Value_16_14],\n",
        "    '15': [Value_1_15, Value_2_15, Value_3_15, Value_4_15, Value_5_15, Value_6_15, Value_7_15, Value_8_15, Value_9_15, Value_10_15, Value_11_15, Value_12_15, Value_13_15, Value_14_15, Value_15_15, Value_16_15],\n",
        "    '16': [Value_1_16, Value_2_16, Value_3_16, Value_4_16, Value_5_16, Value_6_16, Value_7_16, Value_8_16, Value_9_16, Value_10_16, Value_11_16, Value_12_16, Value_13_16, Value_14_16, Value_15_16, Value_16_16]\n",
        "}\n",
        "\n",
        "# Create a DataFrame from the dictionary\n",
        "df = pd.DataFrame(data)\n",
        "\n",
        "# Round the values in the DataFrame to two decimal places\n",
        "df = df.round(2)\n",
        "\n",
        "# Set 'SPECIES' as the index\n",
        "df.set_index('SPECIES', inplace=True)\n",
        "\n",
        "# Set the figure size\n",
        "plt.figure(figsize=(24, 16))\n",
        "\n",
        "# Create a heatmap with seaborn\n",
        "sns.heatmap(df, annot=True, cmap='hot', linewidths=.5, fmt='g')\n",
        "\n",
        "# Set the title and labels\n",
        "plt.title('Difference for the $\\mathit{GENE}$ gene in species (matrix vs matrix) / 7-mer', fontsize=19)\n",
        "plt.xlabel('', fontsize=14)\n",
        "plt.ylabel('', fontsize=14)\n",
        "\n",
        "output_path = \"/your/path/to/save/location/\"\n",
        "# Save the image to the specified path\n",
        "plt.savefig(os.path.join(output_path, 'Difference_Matrix_among_Species.png'), dpi=400, bbox_inches='tight')\n",
        "\n",
        "# Show the heatmap\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "OULrnwbJ2KQG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import os\n",
        "\n",
        "# Create a DataFrame with the provided data\n",
        "data = {\n",
        "    'SPECIES': [r'$\\mathit{H. sapiens}$', r'$\\mathit{G. gorilla}$', r'$\\mathit{P. paniscus}$', r'$\\mathit{P. troglodytes}$',\n",
        "                r'$\\mathit{P. abelii}$', r'$\\mathit{P. pygmaeus}$', r'$\\mathit{M. fascicularis}$', r'$\\mathit{M. mulatta}$', r'$\\mathit{M. thibetana}$',\n",
        "                r'$\\mathit{R. roxellana}$', r'$\\mathit{T. francoisi}$', r'$\\mathit{N. leucogenys}$', r'$\\mathit{S. syndactylus}$',\n",
        "                r'$\\mathit{M. murinus}$', r'$\\mathit{C. jacchus}$', r'$\\mathit{L. catta}$'],\n",
        "\n",
        "    r'$\\mathit{H. sapiens}$': [MS_1_1, MS_2_1, MS_3_1, MS_4_1, MS_5_1, MS_6_1, MS_7_1, MS_8_1, MS_9_1, MS_10_1, MS_11_1, MS_12_1, MS_13_1, MS_14_1, MS_15_1, MS_16_1],\n",
        "    r'$\\mathit{G. gorilla}$': [MS_1_2, MS_2_2, MS_3_2, MS_4_2, MS_5_2, MS_6_2, MS_7_2, MS_8_2, MS_9_2, MS_10_2, MS_11_2, MS_12_2, MS_13_2, MS_14_2, MS_15_2, MS_16_2],\n",
        "    r'$\\mathit{P. paniscus}$': [MS_1_3, MS_2_3, MS_3_3, MS_4_3, MS_5_3, MS_6_3, MS_7_3, MS_8_3, MS_9_3, MS_10_3, MS_11_3, MS_12_3, MS_13_3, MS_14_3, MS_15_3, MS_16_3],\n",
        "    r'$\\mathit{P. troglodytes}$': [MS_1_4, MS_2_4, MS_3_4, MS_4_4, MS_5_4, MS_6_4, MS_7_4, MS_8_4, MS_9_4, MS_10_4, MS_11_4, MS_12_4, MS_13_4, MS_14_4, MS_15_4, MS_16_4],\n",
        "    r'$\\mathit{P. abelii}$': [MS_1_5, MS_2_5, MS_3_5, MS_4_5, MS_5_5, MS_6_5, MS_7_5, MS_8_5, MS_9_5, MS_10_5, MS_11_5, MS_12_5, MS_13_5, MS_14_5, MS_15_5, MS_16_5],\n",
        "    r'$\\mathit{P. pygmaeus}$': [MS_1_6, MS_2_6, MS_3_6, MS_4_6, MS_5_6, MS_6_6, MS_7_6, MS_8_6, MS_9_6, MS_10_6, MS_11_6, MS_12_6, MS_13_6, MS_14_6, MS_15_6, MS_16_6],\n",
        "    r'$\\mathit{M. fascicularis}$': [MS_1_7, MS_2_7, MS_3_7, MS_4_7, MS_5_7, MS_6_7, MS_7_7, MS_8_7, MS_9_7, MS_10_7, MS_11_7, MS_12_7, MS_13_7, MS_14_7, MS_15_7, MS_16_7],\n",
        "    r'$\\mathit{M. mulatta}$': [MS_1_8, MS_2_8, MS_3_8, MS_4_8, MS_5_8, MS_6_8, MS_7_8, MS_8_8, MS_9_8, MS_10_8, MS_11_8, MS_12_8, MS_13_8, MS_14_8, MS_15_8, MS_16_8],\n",
        "    r'$\\mathit{M. thibetana}$': [MS_1_9, MS_2_9, MS_3_9, MS_4_9, MS_5_9, MS_6_9, MS_7_9, MS_8_9, MS_9_9, MS_10_9, MS_11_9, MS_12_9, MS_13_9, MS_14_9, MS_15_9, MS_16_9],\n",
        "    r'$\\mathit{R. roxellana}$': [MS_1_10, MS_2_10, MS_3_10, MS_4_10, MS_5_10, MS_6_10, MS_7_10, MS_8_10, MS_9_10, MS_10_10, MS_11_10, MS_12_10, MS_13_10, MS_14_10, MS_15_10, MS_16_10],\n",
        "    r'$\\mathit{T. francoisi}$': [MS_1_11, MS_2_11, MS_3_11, MS_4_11, MS_5_11, MS_6_11, MS_7_11, MS_8_11, MS_9_11, MS_10_11, MS_11_11, MS_12_11, MS_13_11, MS_14_11, MS_15_11, MS_16_11],\n",
        "    r'$\\mathit{N. leucogenys}$': [MS_1_12, MS_2_12, MS_3_12, MS_4_12, MS_5_12, MS_6_12, MS_7_12, MS_8_12, MS_9_12, MS_10_12, MS_11_12, MS_12_12, MS_13_12, MS_14_12, MS_15_12, MS_16_12],\n",
        "    r'$\\mathit{S. syndactylus}$': [MS_1_13, MS_2_13, MS_3_13, MS_4_13, MS_5_13, MS_6_13, MS_7_13, MS_8_13, MS_9_13, MS_10_13, MS_11_13, MS_12_13, MS_13_13, MS_14_13, MS_15_13, MS_16_13],\n",
        "    r'$\\mathit{M. murinus}$': [MS_1_14, MS_2_14, MS_3_14, MS_4_14, MS_5_14, MS_6_14, MS_7_14, MS_8_14, MS_9_14, MS_10_14, MS_11_14, MS_12_14, MS_13_14, MS_14_14, MS_15_14, MS_16_14],\n",
        "    r'$\\mathit{C. jacchus}$': [MS_1_15, MS_2_15, MS_3_15, MS_4_15, MS_5_15, MS_6_15, MS_7_15, MS_8_15, MS_9_15, MS_10_15, MS_11_15, MS_12_15, MS_13_15, MS_14_15, MS_15_15, MS_16_15],\n",
        "    r'$\\mathit{L. catta}$': [MS_1_16, MS_2_16, MS_3_16, MS_4_16, MS_5_16, MS_6_16, MS_7_16, MS_8_16, MS_9_16, MS_10_16, MS_11_16, MS_12_16, MS_13_16, MS_14_16, MS_15_16, MS_16_16]\n",
        "}\n",
        "\n",
        "df = pd.DataFrame(data)\n",
        "df.set_index('SPECIES', inplace=True)\n",
        "\n",
        "# Set figure size\n",
        "plt.figure(figsize=(30, 8))\n",
        "\n",
        "# Create a line plot with unique markers for each species\n",
        "markers = ['o', 's', '^', 'v', 'D', 'p', 'h', '+', 'x', '<', '>', '*', '|', '_', '.', ',']\n",
        "for i, species in enumerate(df.index):\n",
        "    plt.plot(df.columns, df.loc[species], label=species, marker=markers[i], markersize=8, linestyle='-', linewidth=1)\n",
        "\n",
        "# Set title and labels\n",
        "plt.title(r'General Comparison of the Percentage Difference in the PRKN Gene Among Primates (matrix vs matrix) / 7-mer', fontsize=16)\n",
        "plt.xlabel('', fontsize=14)\n",
        "plt.ylabel(' % Difference', fontsize=14)\n",
        "\n",
        "# Show legend\n",
        "plt.legend(loc='upper right', bbox_to_anchor=(1.12, 1), fontsize='x-large')\n",
        "\n",
        "# Rotate x-axis labels for better readability\n",
        "plt.xticks(rotation=45)\n",
        "\n",
        "output_path = \"/path/to/save/directory\"\n",
        "# Save the image to the specified path\n",
        "plt.savefig(os.path.join(output_path, 'Differences_between_PRKN_of_primate_species.png'), dpi=400, bbox_inches='tight')\n",
        "\n",
        "# Display the plot\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "77GZMOcu4dPi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE Random DNA generator (*RandDNA*)**"
      ],
      "metadata": {
        "id": "ytuTygnW4fqB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import random\n",
        "\n",
        "# Length of the sequence you want to generate (in this case, the length of DATA)\n",
        "sequence_length = len(DATA)\n",
        "\n",
        "# Possible nucleotide bases\n",
        "nucleotides = \"ATCG\"\n",
        "\n",
        "# Generate the initial random sequence\n",
        "random_sequence = ''.join(random.choice(nucleotides) for _ in range(sequence_length))\n",
        "\n",
        "# Iterate over the bases and change if it matches DATA\n",
        "for i, base in enumerate(random_sequence):\n",
        "    if base == DATA[i]:\n",
        "        # Find a different base from DATA\n",
        "        new_base = random.choice(nucleotides.replace(DATA[i], ''))\n",
        "        # Replace the base in the random sequence\n",
        "        random_sequence = random_sequence[:i] + new_base + random_sequence[i+1:]\n",
        "\n",
        "# Print the generated random DNA sequence with changes if it matches DATA\n",
        "print(\"Random DNA sequence with changes if it matches DATA:\")\n",
        "print(random_sequence)\n"
      ],
      "metadata": {
        "id": "nMgrrNaaWaxk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE DNA Subsequence Foward (*DSF*)**"
      ],
      "metadata": {
        "id": "-N_EDU_QLU0H"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define your DNA sequences\n",
        "SEQ1\n",
        "SEQ2\n",
        "SEQ3\n",
        "SEQ4\n",
        "SEQ5\n",
        "SEQ6\n",
        "SEQ7\n",
        "SEQ8\n",
        "SEQ9\n",
        "SEQ10\n",
        "SEQ11\n",
        "SEQ12\n",
        "SEQ13\n",
        "SEQ14\n",
        "SEQ15\n",
        "SEQ16\n",
        "\n",
        "# Length of the subsequence you want to extract (in this case, the length of SEQUENCE)\n",
        "subsequence_length = len(SEQ)\n",
        "\n",
        "# Iterate over the sequences and extract the subsequence of SEQUENCE14's length\n",
        "for i in range(1, 17):\n",
        "    variable_name = f\"SUBSEQUENCE_{i}\"\n",
        "    original_sequence = globals()[f\"SEQUENCE{i}\"]\n",
        "\n",
        "    # Ensure the original sequence is at least as long as SEQUENCE\n",
        "    if len(original_sequence) >= subsequence_length:\n",
        "        new_sequence = original_sequence[:subsequence_length]\n",
        "        # Create a new variable with the desired name\n",
        "        globals()[variable_name] = new_sequence\n",
        "    else:\n",
        "        print(f\"The sequence SEQUENCE{i} is not long enough to extract the subsequence.\")\n",
        "\n",
        "# Print the new sequences (optional)\n",
        "for i in range(1, 17):\n",
        "    variable_name = f\"SUBSEQUENCE_{i}\"\n",
        "    print(f\"{variable_name}: \")\n"
      ],
      "metadata": {
        "id": "9G_Yj7pPWkPS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE DNA Subsequence Reverse (*DSR*)**"
      ],
      "metadata": {
        "id": "g7LOsjLtMiDQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define your DNA sequences\n",
        "SEQ1\n",
        "SEQ2\n",
        "SEQ3\n",
        "SEQ4\n",
        "SEQ5\n",
        "SEQ6\n",
        "SEQ7\n",
        "SEQ8\n",
        "SEQ9\n",
        "SEQ10\n",
        "SEQ11\n",
        "SEQ12\n",
        "SEQ13\n",
        "SEQ14\n",
        "SEQ15\n",
        "SEQ16\n",
        "\n",
        "# Size of the subsequence you want to extract (in this case, the size of SEQ14)\n",
        "subsequence_size = len(SEQ14)\n",
        "\n",
        "# Iterate over the sequences and extract the subsequence of SEQ14 size\n",
        "for i in range(1, 17):\n",
        "    variable_name = f\"SUBSEQ_R14_{i}\"\n",
        "    original_sequence = globals()[f\"SEQ{i}\"]\n",
        "\n",
        "    # Ensure that the original sequence is at least the size of SEQ14\n",
        "    if len(original_sequence) >= subsequence_size:\n",
        "        # Extract the subsequence starting from the last base backwards\n",
        "        new_sequence = original_sequence[-subsequence_size:]\n",
        "        # Create a new variable with the desired name\n",
        "        globals()[variable_name] = new_sequence\n",
        "    else:\n",
        "        print(f\"The sequence SEQ{i} is not long enough to extract the subsequence.\")\n",
        "\n",
        "# Print the new sequences (optional)\n",
        "for i in range(1, 17):\n",
        "    variable_name = f\"SUBSEQ_R14_{i}\"\n",
        "    print(f\"{variable_name}: \")\n"
      ],
      "metadata": {
        "id": "9ECW0hk-fcOQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE DNA extraction random (*DNAxtra*)**"
      ],
      "metadata": {
        "id": "6_o2qgpULlie"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import random\n",
        "import collections\n",
        "import math\n",
        "\n",
        "num_subsequences = n\n",
        "subsequence_size = len(DATA14)\n",
        "\n",
        "# Sequence names\n",
        "sequence_names = [\n",
        "    r'$\\mathit{Sequence \\ 1}$',\n",
        "    r'$\\mathit{Sequence \\ 2}$',\n",
        "    r'$\\mathit{Sequence \\ 3}$',\n",
        "    r'$\\mathit{Sequence \\ 4}$',\n",
        "    r'$\\mathit{Sequence \\ 5}$',\n",
        "    r'$\\mathit{Sequence \\ 6}$',\n",
        "    r'$\\mathit{Sequence \\ 7}$',\n",
        "    r'$\\mathit{Sequence \\ 8}$',\n",
        "    r'$\\mathit{Sequence \\ 9}$',\n",
        "    r'$\\mathit{Sequence \\ 10}$',\n",
        "    r'$\\mathit{Sequence \\ 11}$',\n",
        "    r'$\\mathit{Sequence \\ 12}$',\n",
        "    r'$\\mathit{Sequence \\ 13}$',\n",
        "    r'$\\mathit{Sequence \\ 14}$',\n",
        "    r'$\\mathit{Sequence \\ 15}$',\n",
        "    r'$\\mathit{Sequence \\ 16}$',\n",
        "    r'$\\mathit{Homolog \\ Sequence}$',\n",
        "    r'$\\mathit{Random 100\\% Sequence}$'\n",
        "]\n",
        "\n",
        "# Dictionary to store subsequences grouped by original sequence\n",
        "subsequences_by_sequence = {}\n",
        "\n",
        "# List to store average entropies and standard deviations\n",
        "average_entropies = []\n",
        "standard_deviations = []\n",
        "\n",
        "# Function to calculate Shannon entropy\n",
        "def entropy(sequence):\n",
        "    # Calculate the probability distribution of characters in the sequence.\n",
        "    counts = collections.Counter(sequence)\n",
        "    probabilities = {char: count / len(sequence) for char, count in counts.items()}\n",
        "\n",
        "    # Calculate Shannon entropy.\n",
        "    entropy = 0\n",
        "    for probability in probabilities.values():\n",
        "        if probability > 0:\n",
        "            entropy -= probability * math.log2(probability)\n",
        "\n",
        "    return round(entropy, 5)  # Round the result to 4 decimal places\n",
        "\n",
        "# Iterate over the sequences\n",
        "for i in range(1, 17):\n",
        "    sequence_name = f\"DATA{i}\"\n",
        "    original_sequence = globals()[sequence_name]\n",
        "\n",
        "    # List to store subsequences of this original sequence\n",
        "    subsequences_of_this_sequence = []\n",
        "\n",
        "    # Iterate to extract the desired number of subsequences\n",
        "    for j in range(num_subsequences):\n",
        "        variable_name = f\"SEQUENCE_A_{i}_14_{j+1}\"\n",
        "\n",
        "        # Ensure the original sequence is at least the size of DATA14\n",
        "        if len(original_sequence) >= subsequence_size:\n",
        "            # Generate a random start index\n",
        "            random_start = random.randint(0, len(original_sequence) - subsequence_size)\n",
        "            new_sequence = original_sequence[random_start: random_start + subsequence_size]\n",
        "            # Add the new subsequence to the corresponding list\n",
        "            subsequences_of_this_sequence.append(new_sequence)\n",
        "        else:\n",
        "            print(f\"The sequence {sequence_name} is not large enough to extract subsequence {j+1}.\")\n",
        "\n",
        "    # Calculate the entropies for all subsequences of this sequence\n",
        "    entropies = [entropy(subsequence) for subsequence in subsequences_of_this_sequence]\n",
        "    average_entropies.append(sum(entropies) / len(entropies))\n",
        "    standard_deviations.append(math.sqrt(sum((x - average_entropies[-1]) ** 2 for x in entropies) / len(entropies)))\n",
        "    # Store the subsequences of this original sequence in the dictionary\n",
        "    subsequences_by_sequence[sequence_name] = subsequences_of_this_sequence\n",
        "\n",
        "# Add predefined entropies\n",
        "entropy17 = entropy17\n",
        "entropy18 = entropy18\n",
        "average_entropies.extend([entropy17, entropy18])\n",
        "standard_deviations.extend([0, 0])  # No standard deviation for predefined entropies\n",
        "\n",
        "# Sort entropies from highest to lowest\n",
        "average_entropies_sorted, standard_deviations_sorted, sequence_names_sorted = zip(*sorted(zip(average_entropies, standard_deviations, sequence_names), reverse=False))\n"
      ],
      "metadata": {
        "id": "PMKcw5dDfe58"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE Random Mutation Entropy Analysis (*RMEA*)**"
      ],
      "metadata": {
        "id": "cBV1tPnYLyDz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import random\n",
        "import math\n",
        "\n",
        "def generate_mutated_sequence(original_sequence, num_changes):\n",
        "    mutated_sequence = list(original_sequence)\n",
        "    bases = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
        "\n",
        "    for _ in range(num_changes):\n",
        "        index = random.randint(0, len(mutated_sequence) - 1)\n",
        "        mutated_sequence[index] = bases[mutated_sequence[index]]\n",
        "\n",
        "    return ''.join(mutated_sequence)\n",
        "\n",
        "# Number of changes to make in each sequence\n",
        "num_changes = n\n",
        "# List to store the generated sequences\n",
        "generated_sequences = []\n",
        "\n",
        "for _ in range(num_mutated_sequences):\n",
        "    generated_sequences.append(generate_mutated_sequence(original_sequence, num_changes))\n",
        "\n",
        "# Assign individual sequences to variables\n",
        "for i, sequence in enumerate(generated_sequences, start=1):\n",
        "    globals()[f\"DATA1_M1_{i}\"] = sequence\n",
        "\n",
        "# Generate mutated sequences and calculate entropy for each\n",
        "entropies = []\n",
        "for _ in range(num_mutated_sequences):\n",
        "    mutated_sequence = generate_mutated_sequence(original_sequence, num_changes)\n",
        "    generated_sequences.append(mutated_sequence)\n",
        "    entropies.append(entropy(mutated_sequence))\n",
        "\n",
        "# Calculate the average entropy\n",
        "average_entropy = sum(entropies) / len(entropies)\n",
        "\n",
        "# Calculate the standard deviation of the average entropy\n",
        "standard_deviation = math.sqrt(sum((x - average_entropy) ** 2 for x in entropies) / len(entropies))\n",
        "\n",
        "print(\"Average entropy:\", average_entropy)\n",
        "print(\"Standard deviation of the average entropy:\", standard_deviation)\n"
      ],
      "metadata": {
        "id": "GIHBkbUkF3WX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE Complementary Nucleotide Sequence Mutation (*CNSMut*)**"
      ],
      "metadata": {
        "id": "rERx74W_MHfO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import random\n",
        "import math\n",
        "\n",
        "def generate_mutated_sequence(original_sequence, num_changes):\n",
        "    mutated_sequence = list(original_sequence)\n",
        "    bases = ['A', 'T', 'C', 'G']\n",
        "\n",
        "    # Get unique positions to make changes\n",
        "    change_positions = random.sample(range(len(original_sequence)), num_changes)\n",
        "\n",
        "    # Change the bases at the selected positions\n",
        "    for pos in change_positions:\n",
        "        original_base = original_sequence[pos]\n",
        "        bases_without_orig = [b for b in bases if b != original_base]\n",
        "        mutated_sequence[pos] = random.choice(bases_without_orig)\n",
        "\n",
        "    return ''.join(mutated_sequence)\n",
        "\n",
        "# Number of changes to make in each sequence\n",
        "num_changes = n\n",
        "\n",
        "# List to store the generated sequences\n",
        "generated_sequences = []\n",
        "\n",
        "for _ in range(num_mutated_sequences):\n",
        "    generated_sequences.append(generate_mutated_sequence(original_sequence, num_changes))\n",
        "\n",
        "# Individual assignment to variables\n",
        "for i, sequence in enumerate(generated_sequences, start=1):\n",
        "    globals()[f\"SEQUENCE_{i}\"] = sequence\n",
        "\n",
        "# Generate mutated sequences and calculate entropy for each one\n",
        "entropies = []\n",
        "for _ in range(num_mutated_sequences):\n",
        "    mutated_sequence = generate_mutated_sequence(original_sequence, num_changes)\n",
        "    generated_sequences.append(mutated_sequence)\n",
        "    entropies.append(entropy(mutated_sequence))\n",
        "\n",
        "# Calculate the average entropy\n",
        "average_entropy = sum(entropies) / len(entropies)\n",
        "\n",
        "# Calculate the standard deviation of the average entropy\n",
        "std_dev_entropy = math.sqrt(sum((x - average_entropy) ** 2 for x in entropies) / len(entropies))\n",
        "\n",
        "print(\"Average entropy:\", average_entropy)\n",
        "print(\"Standard deviation of the average entropy:\", std_dev_entropy)\n"
      ],
      "metadata": {
        "id": "bPl3adcvJU0L"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CODE Chemical Swap Entropy Analysis (*ChemSwap*)**"
      ],
      "metadata": {
        "id": "qQZi3ykgMNVZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import random\n",
        "import math\n",
        "\n",
        "def generate_mutated_sequence(original_sequence, num_changes):\n",
        "    mutated_sequence = list(original_sequence)\n",
        "    purine_bases = ['A', 'G']  # Purines\n",
        "    pyrimidine_bases = ['T', 'C']  # Pyrimidines\n",
        "\n",
        "    for _ in range(num_changes):\n",
        "        index = random.randint(0, len(mutated_sequence)-1)\n",
        "        original_base = mutated_sequence[index]\n",
        "        if original_base in purine_bases:\n",
        "            new_base = random.choice(pyrimidine_bases)\n",
        "        elif original_base in pyrimidine_bases:\n",
        "            new_base = random.choice(purine_bases)\n",
        "        else:\n",
        "            # In case it is neither a purine nor a pyrimidine, no change is made\n",
        "            continue\n",
        "\n",
        "        mutated_sequence[index] = new_base\n",
        "\n",
        "    return ''.join(mutated_sequence)\n",
        "\n",
        "# Number of changes to make in each sequence\n",
        "num_changes = n\n",
        "\n",
        "# List to store generated sequences\n",
        "generated_sequences = []\n",
        "\n",
        "for _ in range(num_mutated_sequences):\n",
        "    generated_sequences.append(generate_mutated_sequence(original_sequence, num_changes))\n",
        "\n",
        "# Individual assignment to variables\n",
        "for i, sequence in enumerate(generated_sequences, start=1):\n",
        "    globals()[f\"DATA1_G1_{i}\"] = sequence\n",
        "\n",
        "# Generate mutated sequences and calculate entropy for each\n",
        "entropies = []\n",
        "for _ in range(num_mutated_sequences):\n",
        "    mutated_sequence = generate_mutated_sequence(original_sequence, num_changes)\n",
        "    generated_sequences.append(mutated_sequence)\n",
        "    entropies.append(entropy(mutated_sequence))\n",
        "\n",
        "# Calculate the average entropy\n",
        "average_entropy = sum(entropies) / len(entropies)\n",
        "\n",
        "# Calculate the standard deviation of the average entropy\n",
        "entropy_std_dev = math.sqrt(sum((x - average_entropy) ** 2 for x in entropies) / len(entropies))\n",
        "\n",
        "print(\"Average entropy:\", average_entropy)\n",
        "print(\"Standard deviation of the average entropy:\", entropy_std_dev)\n"
      ],
      "metadata": {
        "id": "u-bQsujaKoHP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **IMAGE GROUPER(*IMG*)**"
      ],
      "metadata": {
        "id": "weEPAp5KLz6Q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "from PIL import Image\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Define the path to the image folder\n",
        "folder_path = IMAGE_FOLDER_PATH\n",
        "\n",
        "# Get a list of files in the folder and sort them by name\n",
        "image_files = sorted([f for f in os.listdir(folder_path) if f.endswith(('png', 'jpg', 'jpeg', 'bmp', 'gif'))])\n",
        "\n",
        "# Create the figure and axes for the image grid\n",
        "fig, axs = plt.subplots(2, 5, figsize=(20, 10))\n",
        "\n",
        "# Iterate over the images and display them in the grid\n",
        "for i, ax in enumerate(axs.flatten()):\n",
        "    if i < len(image_files):\n",
        "        image_path = os.path.join(folder_path, image_files[i])\n",
        "        img = Image.open(image_path)\n",
        "        ax.imshow(img)\n",
        "        ax.axis('off')  # Hide the axes\n",
        "        # Get the file name without the extension\n",
        "        title = os.path.splitext(image_files[i])[0]\n",
        "        ax.set_title(title, fontsize=10)  # Optional: set the file name as title\n",
        "\n",
        "plt.tight_layout()\n",
        "# Adjust space between subplots and space for the title\n",
        "plt.subplots_adjust(top=0.9)\n",
        "plt.suptitle('Clustering IMAGES', fontsize=16)\n",
        "\n",
        "plt.tight_layout()\n",
        "# Save the figure to the specified path\n",
        "plt.savefig('output_path/IMAGE_GROUP.png')\n",
        "\n",
        "# Show the figure\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "KLBJbaqoL8n7"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}