{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-2Nss7qjGgyN"
      },
      "source": [
        "# **LIBRERIAS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eTnAvveoGZa0"
      },
      "outputs": [],
      "source": [
        "!pip install xlsxwriter"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hzDTgd9lGkI5"
      },
      "outputs": [],
      "source": [
        "import collections\n",
        "from collections import OrderedDict\n",
        "from matplotlib import pyplot as plt\n",
        "from matplotlib import cm\n",
        "import pylab\n",
        "import math\n",
        "from PIL import Image\n",
        "import numpy as np\n",
        "from itertools import count\n",
        "import os\n",
        "import cv2\n",
        "import matplotlib.pyplot as plt\n",
        "from google.colab.patches import cv2_imshow\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive/')\n",
        "from scipy.spatial.distance import pdist, squareform\n",
        "from scipy.cluster.hierarchy import linkage, dendrogram\n",
        "import shutil\n",
        "from scipy.ndimage import binary_erosion\n",
        "from scipy.spatial import ConvexHull\n",
        "import imageio\n",
        "from mpl_toolkits.mplot3d import Axes3D  # Importa Axes3D para el gráfico 3D\n",
        "import pandas as pd\n",
        "import gc\n",
        "import os\n",
        "import seaborn as sns\n",
        "\n",
        "\n",
        "import matplotlib.backends.backend_pdf\n",
        "import matplotlib.pyplot as plt\n",
        "from collections import defaultdict\n",
        "from operator import itemgetter\n",
        "import shutil\n",
        "\n",
        "import imageio\n",
        "from mpl_toolkits.mplot3d import Axes3D  # Importa Axes3D para el gráfico 3D\n",
        "import pandas as pd\n",
        "import gc\n",
        "import os\n",
        "import seaborn as sns\n",
        "from matplotlib.colors import Normalize\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "p6B6HOu2GmPn"
      },
      "source": [
        "# **ARCHIVOS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lfxSDTNSGqGg"
      },
      "outputs": [],
      "source": [
        "fasta_1 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Homo sapiens1.fasta\") #Homo sapiens (humano)\n",
        "fasta_2 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Homo Gorilla.fasta\") #Gorilla gorilla gorilla (gorila occidental de llanura)\n",
        "fasta_3 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Pan paniscus.fasta\") #Pan paniscus (bonobo)\n",
        "fasta_4 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Pan troglodytes.fasta\") #Pan troglodytes (chimpancé común)\n",
        "fasta_5 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Pongo abelii.fasta\") #Pongo abelii (orangután de Sumatra)\n",
        "fasta_6 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Pongo pygmaeus.fasta\") #Pongo pygmaeus (orangután de Borneo)\n",
        "fasta_7 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Macaca fascicularis.fasta\") #Macaca fascicularis (macaco cangrejero)\n",
        "fasta_8 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Macaca mulatta.fasta\") #Macaca mulatta (macaco rhesus)\n",
        "fasta_9 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Macaca thibetana thibetana.fasta\") #Macaca thibetana thibetana (macaco tibetano)\n",
        "fasta_10 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Rhinopithecus roxellana.fasta\") #Rhinopithecus roxellana (macaco de nariz dorada)\n",
        "fasta_11 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Trachypithecus francoisi.fasta\") #Trachypithecus francoisi (langur de François)\n",
        "fasta_12 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Nomascus leucogenys.fasta\") #Nomascus leucogenys (gibón de mejillas blancas)\n",
        "fasta_13 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Symphalangus syndactylus.fasta\") #Symphalangus syndactylus (gibón siamés)\n",
        "fasta_14 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Microcebus murinus.fasta\") #Microcebus murinus (ratón lémur gris)\n",
        "fasta_15 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Callithrix jacchus.fasta\") #Callithrix jacchus (tití común)\n",
        "fasta_16 = open(\"/content/drive/MyDrive/fractal_python/SECUENCIAS_FASTA/TRABAJO/nuevas/Lemur catta.fasta\") #Lemur catta (lémur de cola anillada)\n",
        "\n",
        "#se lee el archivo fasta_ y se le asigna a la variable  f\n",
        "f1 = fasta_1.read()\n",
        "f2 = fasta_2.read()\n",
        "f3 = fasta_3.read()\n",
        "f4 = fasta_4.read()\n",
        "f5 = fasta_5.read()\n",
        "f6 = fasta_6.read()\n",
        "f7 = fasta_7.read()\n",
        "f8 = fasta_8.read()\n",
        "f9 = fasta_9.read()\n",
        "f10 = fasta_10.read()\n",
        "f11 = fasta_11.read()\n",
        "f12 = fasta_12.read()\n",
        "f13 = fasta_13.read()\n",
        "f14 = fasta_14.read()\n",
        "f15 = fasta_15.read()\n",
        "f16 = fasta_16.read()\n",
        "\n",
        "\n",
        "#carga la variable s1  y le da lectura y asigna el nombre a una variable DATA\n",
        "DATA1 = \"\".join(f1.split(\"\\n\")[0:])\n",
        "DATA2 = \"\".join(f2.split(\"\\n\")[0:])\n",
        "DATA3 = \"\".join(f3.split(\"\\n\")[0:])\n",
        "DATA4 = \"\".join(f4.split(\"\\n\")[0:])\n",
        "DATA5 = \"\".join(f5.split(\"\\n\")[0:])\n",
        "DATA6 = \"\".join(f6.split(\"\\n\")[0:])\n",
        "DATA7 = \"\".join(f7.split(\"\\n\")[0:])\n",
        "DATA8 = \"\".join(f8.split(\"\\n\")[0:])\n",
        "DATA9 = \"\".join(f9.split(\"\\n\")[0:])\n",
        "DATA10 = \"\".join(f10.split(\"\\n\")[0:])\n",
        "DATA11 = \"\".join(f11.split(\"\\n\")[0:])\n",
        "DATA12 = \"\".join(f12.split(\"\\n\")[0:])\n",
        "DATA13 = \"\".join(f13.split(\"\\n\")[0:])\n",
        "DATA14 = \"\".join(f14.split(\"\\n\")[0:])\n",
        "DATA15 = \"\".join(f15.split(\"\\n\")[0:])\n",
        "DATA16 = \"\".join(f16.split(\"\\n\")[0:])\n",
        "\n",
        "\n",
        "#se elimina el primer elemento de la lista (la línea en blanco) y se concatenan todas las líneas restantes en una sola cadena de texto.\n",
        "DATA_FRAME1 = \"\".join(f1.split(\"\\n\")[1:])\n",
        "DATA_FRAME2 = \"\".join(f2.split(\"\\n\")[1:])\n",
        "DATA_FRAME3 = \"\".join(f3.split(\"\\n\")[1:])\n",
        "DATA_FRAME4 = \"\".join(f4.split(\"\\n\")[1:])\n",
        "DATA_FRAME5 = \"\".join(f5.split(\"\\n\")[1:])\n",
        "DATA_FRAME6 = \"\".join(f6.split(\"\\n\")[1:])\n",
        "DATA_FRAME7 = \"\".join(f7.split(\"\\n\")[1:])\n",
        "DATA_FRAME8 = \"\".join(f8.split(\"\\n\")[1:])\n",
        "DATA_FRAME9 = \"\".join(f9.split(\"\\n\")[1:])\n",
        "DATA_FRAME10 = \"\".join(f10.split(\"\\n\")[1:])\n",
        "DATA_FRAME11 = \"\".join(f11.split(\"\\n\")[1:])\n",
        "DATA_FRAME12 = \"\".join(f12.split(\"\\n\")[1:])\n",
        "DATA_FRAME13 = \"\".join(f13.split(\"\\n\")[1:])\n",
        "DATA_FRAME14 = \"\".join(f14.split(\"\\n\")[1:])\n",
        "DATA_FRAME15 = \"\".join(f15.split(\"\\n\")[1:])\n",
        "DATA_FRAME16 = \"\".join(f16.split(\"\\n\")[1:])\n",
        "\n",
        "\n",
        "\n",
        "DATA_1 = DATA_FRAME1\n",
        "DATA_2 = DATA_FRAME2\n",
        "DATA_3 = DATA_FRAME3\n",
        "DATA_4 = DATA_FRAME4\n",
        "DATA_5 = DATA_FRAME5\n",
        "DATA_6 = DATA_FRAME6\n",
        "DATA_7 = DATA_FRAME7\n",
        "DATA_8 = DATA_FRAME8\n",
        "DATA_9 = DATA_FRAME9\n",
        "DATA_10 = DATA_FRAME10\n",
        "DATA_11 = DATA_FRAME11\n",
        "DATA_12 = DATA_FRAME12\n",
        "DATA_13 = DATA_FRAME13\n",
        "DATA_14 = DATA_FRAME14\n",
        "DATA_15 = DATA_FRAME15\n",
        "DATA_16 = DATA_FRAME16\n",
        "\n",
        "DATOS1 = DATA_1\n",
        "DATOS2 = DATA_2\n",
        "DATOS3 = DATA_3\n",
        "DATOS4 = DATA_4\n",
        "DATOS5 = DATA_5\n",
        "DATOS6 = DATA_6\n",
        "DATOS7 = DATA_7\n",
        "DATOS8 = DATA_8\n",
        "DATOS9 = DATA_9\n",
        "DATOS10 = DATA_10\n",
        "DATOS11 = DATA_11\n",
        "DATOS12 = DATA_12\n",
        "DATOS13 = DATA_13\n",
        "DATOS14 = DATA_14\n",
        "DATOS15 = DATA_15\n",
        "DATOS16 = DATA_16\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H-0WQom8GvlB"
      },
      "source": [
        "# **FUNCIONES**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vYEzBB8cHnk0"
      },
      "source": [
        "## **CONTEO FRECUENCIA RELATIVA  Y TOTAL DE DNA**\n",
        "\n",
        "---\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "C8J5NsM9HxGv"
      },
      "outputs": [],
      "source": [
        "def calculate_base_content(dna_sequence):\n",
        "    length = len(dna_sequence)\n",
        "    a_count = dna_sequence.count('A')\n",
        "    t_count = dna_sequence.count('T')\n",
        "    g_count = dna_sequence.count('G')\n",
        "    c_count = dna_sequence.count('C')\n",
        "    total_bases = a_count + t_count + g_count + c_count\n",
        "    a_percent = (a_count / total_bases) * 100\n",
        "    t_percent = (t_count / total_bases) * 100\n",
        "    g_percent = (g_count / total_bases) * 100\n",
        "    c_percent = (c_count / total_bases) * 100\n",
        "    return a_percent, t_percent, g_percent, c_percent\n",
        "\n",
        "def plot_base_content(a_percent, t_percent, g_percent, c_percent, species_name,save_path):\n",
        "    bases = ['A', 'T', 'G', 'C']\n",
        "    percentages = [a_percent, t_percent, g_percent, c_percent]\n",
        "    colors = ['green', 'red', 'black', 'blue']\n",
        "\n",
        "    plt.figure(figsize=(8, 6))\n",
        "    bars = plt.bar(bases, percentages, color=colors)\n",
        "    plt.title(f'Relative Frequency of bases in {species_name} gene')\n",
        "    plt.xlabel('Bases')\n",
        "    plt.ylabel('Relative Percentage (%)')\n",
        "    plt.ylim(0, 100)\n",
        "\n",
        "    for bar, percentage in zip(bars, percentages):\n",
        "        plt.text(bar.get_x() + bar.get_width()/2, bar.get_height(), f'{percentage:.2f}%',\n",
        "                 ha='center', va='bottom')\n",
        "\n",
        "    plt.savefig(save_path)\n",
        "    plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z1Q0DtUTPdeB"
      },
      "source": [
        "## **TOTALES**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "copvKP2oPjKg"
      },
      "outputs": [],
      "source": [
        "def calculate_base_content_T(dna_sequence):\n",
        "    length = len(dna_sequence)\n",
        "    a_count = dna_sequence.count('A')\n",
        "    t_count = dna_sequence.count('T')\n",
        "    g_count = dna_sequence.count('G')\n",
        "    c_count = dna_sequence.count('C')\n",
        "    total_bases = length\n",
        "    return a_count, t_count, g_count, c_count\n",
        "\n",
        "def plot_base_content_T(a_count, t_count, g_count, c_count, species_name, save_path):\n",
        "    bases = ['A', 'T', 'G', 'C']\n",
        "    counts = [a_count, t_count, g_count, c_count]\n",
        "    colors = ['green', 'red', 'black', 'blue']\n",
        "\n",
        "    plt.figure(figsize=(8, 6))\n",
        "    bars = plt.bar(bases, counts, color=colors)\n",
        "    plt.title(f'Total Frequency of bases in {species_name} gene')\n",
        "    plt.xlabel('Bases')\n",
        "    plt.ylabel('Total Count')\n",
        "\n",
        "    for bar, count in zip(bars, counts):\n",
        "        plt.text(bar.get_x() + bar.get_width()/2, bar.get_height(), f'{count}',\n",
        "                 ha='center', va='bottom')\n",
        "\n",
        "    # Guardar la imagen en la ruta especificada\n",
        "    plt.savefig(save_path)\n",
        "    plt.show()\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_zpb3pwtHne6"
      },
      "source": [
        "## **CONTEO DE K MER Y GENERACION DE MATRIZ**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rx_K15lcG0gx"
      },
      "outputs": [],
      "source": [
        "# Define la función count_kmers\n",
        "def count_kmers(seq, k):\n",
        "    d = collections.defaultdict(int)\n",
        "    for i in range(len(seq) - (k - 1)):\n",
        "        d[seq[i:i+k]] += 1\n",
        "    return d\n",
        "\n",
        "# Define la función chaos_game_representation\n",
        "def chaos_game_representation(frequencies, k):\n",
        "    array_size = int(math.sqrt(4**k))\n",
        "    chaos = []\n",
        "    for i in range(array_size):\n",
        "        chaos.append([0] * array_size)\n",
        "\n",
        "    maxx = array_size\n",
        "    maxy = array_size\n",
        "    posx = 1\n",
        "    posy = 1\n",
        "    for key, value in frequencies.items():\n",
        "        for char in key:\n",
        "            if char == \"A\":\n",
        "                posx += maxx / 2\n",
        "            elif char == \"T\":\n",
        "                posy += maxy / 2\n",
        "            elif char == \"G\":\n",
        "                posx += maxx / 2\n",
        "                posy += maxy / 2\n",
        "            maxx = maxx / 2\n",
        "            maxy /= 2\n",
        "        chaos[int(posy)-1][int(posx)-1] = value\n",
        "        maxx = array_size\n",
        "        maxy = array_size\n",
        "        posx = 1\n",
        "        posy = 1\n",
        "\n",
        "    return chaos"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CugJV9mZHtw8"
      },
      "source": [
        "## **GENERACION DE IMAGENES SIN ETIQUETAS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qU733kyNHWuv"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.axis('off')  # Elimina el eje\n",
        "    fig.savefig(os.path.join(output2_directory, f'{name}.png'), bbox_inches='tight', pad_inches=0)\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VkK5d5A8HyxM"
      },
      "source": [
        "## **COMPARATIVA DE MATRICES**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4nXUO7_9HXPi"
      },
      "outputs": [],
      "source": [
        "def compare_matrices(matrices):\n",
        "    matrix1 = np.array(matrices[0])\n",
        "    percentage_differences = []\n",
        "\n",
        "    for matrix in matrices[1:]:\n",
        "        # Convierte la matriz a un array de NumPy\n",
        "        matrix = np.array(matrix)\n",
        "\n",
        "        # Calcula la diferencia absoluta entre las dos matrices\n",
        "        absolute_difference = np.abs(matrix1 - matrix)\n",
        "\n",
        "        # Calcula el porcentaje de diferencia en comparación con la primera matriz\n",
        "        percentage_difference = (np.sum(absolute_difference) / np.sum(matrix1)) * 100\n",
        "        percentage_differences.append(percentage_difference)\n",
        "\n",
        "    return percentage_differences"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vJdXW1ggKa_3"
      },
      "source": [
        "## **nombres**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-zJunM4kIiY5"
      },
      "outputs": [],
      "source": [
        "NAME_1 = '01 H. sapiens'\n",
        "NAME_2 = '02 G. gorilla'\n",
        "NAME_3 = '03 P. paniscus'\n",
        "NAME_4 = '04 P. troglodytes'\n",
        "NAME_5 = '05 P. abelii'\n",
        "NAME_6 = '06 P. pygmaeus'\n",
        "NAME_7 = '07 M. fascicularis'\n",
        "NAME_8 = '08 M. mulatta'\n",
        "NAME_9 = '09 M. thibetana'\n",
        "NAME_10 = '10 R. roxellana'\n",
        "NAME_11 = '11 T. francoisi'\n",
        "NAME_12 = '12 N. leucogenys'\n",
        "NAME_13 = '13 S. syndactylus'\n",
        "NAME_14 = '14 M. murinus'\n",
        "NAME_15 = '15 C. jacchus'\n",
        "NAME_16 = '16 L. catta'\n",
        "\n",
        "NAME1_1 = 'H. sapiens'\n",
        "NAME1_2 = 'G. gorilla'\n",
        "NAME1_3 = 'P. paniscus'\n",
        "NAME1_4 = 'P. troglodytes'\n",
        "NAME1_5 = 'P. abelii'\n",
        "NAME1_6 = 'P. pygmaeus'\n",
        "NAME1_7 = 'M. fascicularis'\n",
        "NAME1_8 = 'M. mulatta'\n",
        "NAME1_9 = 'M. thibetana'\n",
        "NAME1_10 = 'R. roxellana'\n",
        "NAME1_11 = 'T. francoisi'\n",
        "NAME1_12 = 'N. leucogenys'\n",
        "NAME1_13 = 'S. syndactylus'\n",
        "NAME1_14 = 'M. murinus'\n",
        "NAME1_15 = 'C. jacchus'\n",
        "NAME1_16 = 'L. catta'\n",
        "\n",
        "NAME1 = r'$\\mathit{H. sapiens \\ PRKN}$'\n",
        "NAME2 = r'$\\mathit{G. gorilla \\ PRKN}$'\n",
        "NAME3 = r'$\\mathit{P. paniscus \\ PRKN}$'\n",
        "NAME4 = r'$\\mathit{P. troglodytes \\ PRKN}$'\n",
        "NAME5 = r'$\\mathit{P. abelii \\ PRKN}$'\n",
        "NAME6 = r'$\\mathit{P. pygmaeus \\ PRKN}$'\n",
        "NAME7 = r'$\\mathit{M. fascicularis \\ PRKN}$'\n",
        "NAME8 = r'$\\mathit{M. mulatta \\ PRKN}$'\n",
        "NAME9 = r'$\\mathit{M. thibetana \\ PRKN}$'\n",
        "NAME10 = r'$\\mathit{R. roxellana \\ PRKN}$'\n",
        "NAME11 = r'$\\mathit{T. francoisi \\ PRKN}$'\n",
        "NAME12 = r'$\\mathit{N. leucogenys \\ PRKN}$'\n",
        "NAME13 = r'$\\mathit{S. syndactylus \\ PRKN}$'\n",
        "NAME14 = r'$\\mathit{M. murinus \\ PRKN}$'\n",
        "NAME15 = r'$\\mathit{C. jacchus \\ PRKN}$'\n",
        "NAME16 = r'$\\mathit{L. catta \\ PRKN}$'\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GGGKpv4dH2bZ"
      },
      "source": [
        "# **CONTEO DE FRECUENCIAS RELATIVAS Y TOTALES DNA**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4IDiFGOqIQpg"
      },
      "source": [
        "## **Freecuencias relativas** **texto en negrita**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "84eIEu5FMoDE"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS1\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_1}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME1,save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tN7nE2C9M-d9"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS2\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_2}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME2, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "24gS2PAuM-7A"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS3\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_3}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME3, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KxgZcF0_M_Q0"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS4\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_4}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME4, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "RbNPO0tQM_rq"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS5\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_5}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME5, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "U8R_v_14NABo"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS6\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_6}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME6, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "T-NvIlH1NEY2"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS7\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_7}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME7, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "O7ASuViBNEz0"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS8\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_8}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME8, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iYLZPl1zNFLe"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS9\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_9}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME9, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "F2J3I2WINFcC"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS10\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_10}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME10, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xQiibcA9NFt0"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS11\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_11}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME11, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "v77oEIisNGEo"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS12\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_12}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME12, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "c7keYuM1NGja"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS13\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_13}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME13, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "uZGy4S4XNG6k"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS14\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_14}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME14, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "St0f5o8eNHU7"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS15\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_15}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME15, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lbz_JDdpNHp6"
      },
      "outputs": [],
      "source": [
        "dna_sequence = DATOS16\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/01 Supplementary material/00 PRKN nucleotide frequency/{NAME_16}.png\"\n",
        "# Calcular el contenido relativo de bases\n",
        "a_percent, t_percent, g_percent, c_percent = calculate_base_content(dna_sequence)\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content(a_percent, t_percent, g_percent, c_percent, NAME16, save_path)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WedrUmZpOzZl"
      },
      "source": [
        "## **FRECUENCIA TOTALES**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jzWM9C4VO2GD"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS1  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_1}/{NAME_1}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME1, save_path)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qisjHHCHRNJH"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS2  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_2}/{NAME_2}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME2, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Tl4Zx5KxRNoJ"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS3  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_3}/{NAME_3}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME3, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "pqYWxjH8ROBx"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS4  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_4}/{NAME_4}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME4, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rrKY0Jv5ROcX"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS5  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_5}/{NAME_5}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME5, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5qOklZTMRO1h"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS6  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_6}/{NAME_6}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME6, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xiXRCBzpRPUp"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS7  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_7}/{NAME_7}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME7, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cMPjj_DwTdws"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS8  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_8}/{NAME_8}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME8, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3P9eLBKGRPt7"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS9  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_9}/{NAME_9}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME_9, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "slH-ddsKRQQX"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS10  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_10}/{NAME_10}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME10, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xxBtZlwrRQou"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS11  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_11}/{NAME_11}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME11, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QFPVqFMfRRAj"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS12  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_12}/{NAME_12}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME12, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Uz_fmd-fRRc_"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS13  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_13}/{NAME_13}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME13, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "geY4cYz_RR08"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS14  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_14}/{NAME_14}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME14, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "aL6p48NLRSH7"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS15  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_15}/{NAME_15}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME15, save_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7-9R2l4gRSl9"
      },
      "outputs": [],
      "source": [
        "# Ejemplo de secuencia de ADN\n",
        "dna_sequence = DATOS16  # Asegúrate de definir tu secuencia de ADN correctamente\n",
        "# Definir la ruta donde deseas guardar la imagen\n",
        "save_path = f\"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/{NAME_16}/{NAME_16}.png\"\n",
        "\n",
        "# Calcular el conteo total de bases\n",
        "a_count, t_count, g_count, c_count = calculate_base_content_T(dna_sequence)\n",
        "\n",
        "# Generar el gráfico de barras\n",
        "plot_base_content_T(a_count, t_count, g_count, c_count, NAME16, save_path)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3i3GwnclX6Dd"
      },
      "source": [
        "# **CONTEO DE K-MER Y MATRIZ DE FRECUENCIAS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "d40SYph5YBPl"
      },
      "outputs": [],
      "source": [
        " #data1\n",
        "d1_7 = count_kmers(DATOS1, 7)\n",
        "#data2\n",
        "d2_7 = count_kmers(DATOS2, 7)\n",
        "#data3\n",
        "d3_7 = count_kmers(DATOS3, 7)\n",
        "#data4\n",
        "d4_7 = count_kmers(DATOS4, 7)\n",
        "#data5\n",
        "d5_7 = count_kmers(DATOS5, 7)\n",
        "#data6\n",
        "d6_7 = count_kmers(DATOS6, 7)\n",
        "#data7\n",
        "d7_7 = count_kmers(DATOS7, 7)\n",
        "#data8\n",
        "d8_7 = count_kmers(DATOS8, 7)\n",
        "#data9\n",
        "d9_7 = count_kmers(DATOS9, 7)\n",
        "#data10\n",
        "d10_7 = count_kmers(DATOS10, 7)\n",
        "#data11\n",
        "d11_7 = count_kmers(DATOS11, 7)\n",
        "#data12\n",
        "d12_7 = count_kmers(DATOS12, 7)\n",
        "#data13\n",
        "d13_7 = count_kmers(DATOS13, 7)\n",
        "#data14\n",
        "d14_7 = count_kmers(DATOS14, 7)\n",
        "#data15\n",
        "d15_7 = count_kmers(DATOS15, 7)\n",
        "#data16\n",
        "d16_7 = count_kmers(DATOS16, 7)\n",
        "\n",
        "d17_7 = count_kmers(DATOS16, 7)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "POecPaz0YEf0"
      },
      "outputs": [],
      "source": [
        "chaos_1_k7 = chaos_game_representation(d1_7, 7)\n",
        "chaos_2_k7 = chaos_game_representation(d2_7, 7)\n",
        "chaos_3_k7 = chaos_game_representation(d3_7, 7)\n",
        "chaos_4_k7 = chaos_game_representation(d4_7, 7)\n",
        "chaos_5_k7 = chaos_game_representation(d5_7, 7)\n",
        "chaos_6_k7 = chaos_game_representation(d6_7, 7)\n",
        "chaos_7_k7 = chaos_game_representation(d7_7, 7)\n",
        "chaos_8_k7 = chaos_game_representation(d8_7, 7)\n",
        "chaos_9_k7 = chaos_game_representation(d9_7, 7)\n",
        "chaos_10_k7 = chaos_game_representation(d10_7, 7)\n",
        "chaos_11_k7 = chaos_game_representation(d11_7, 7)\n",
        "chaos_12_k7 = chaos_game_representation(d12_7, 7)\n",
        "chaos_13_k7 = chaos_game_representation(d13_7, 7)\n",
        "chaos_14_k7 = chaos_game_representation(d14_7, 7)\n",
        "chaos_15_k7 = chaos_game_representation(d15_7, 7)\n",
        "chaos_16_k7 = chaos_game_representation(d16_7, 7)\n",
        "chaos_17_k7 = chaos_game_representation(d17_7, 7)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0O6vkNcGXVKN"
      },
      "source": [
        "# **FRECUENCIA DE K-MERS**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "D6n6nd4bBXGj"
      },
      "source": [
        "## **IMAGEN FRECUENCIA**"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Lista de tus conjuntos de datos\n",
        "conjuntos_de_datos = [d1_7, d2_7, d3_7, d4_7, d5_7, d6_7, d7_7, d8_7,\n",
        "                      d9_7, d10_7, d11_7, d12_7, d13_7, d14_7, d15_7, d16_7]\n",
        "\n",
        "nombres_especies = [\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{G. gorilla \\ PRKN}$',\n",
        "    r'$\\mathit{P. paniscus \\ PRKN}$',\n",
        "    r'$\\mathit{P. troglodytes \\ PRKN}$',\n",
        "    r'$\\mathit{P. abelii \\ PRKN}$',\n",
        "    r'$\\mathit{P. pygmaeus \\ PRKN}$',\n",
        "    r'$\\mathit{M. fascicularis \\ PRKN}$',\n",
        "     r'$\\mathit{M. mulatta \\ PRKN}$',\n",
        "    r'$\\mathit{M. thibetana \\ PRKN}$',\n",
        "    r'$\\mathit{R. roxellana \\ PRKN}$',\n",
        "    r'$\\mathit{T. francoisi \\ PRKN}$',\n",
        "    r'$\\mathit{N. leucogenys \\ PRKN}$',\n",
        "    r'$\\mathit{S. syndactylus \\ PRKN}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$',\n",
        "]\n",
        "\n",
        "# Lista de nombres de especies correspondientes a cada conjunto de datos\n",
        "etiquetas_especies = [\n",
        "    '01 H. sapiens',\n",
        "    '02 G. gorilla',\n",
        "    '03 P. paniscus',\n",
        "    '04 P. troglodytes',\n",
        "    '05 P. abelii',\n",
        "    '06 P. pygmaeus',\n",
        "    '07 M. fascicularis',\n",
        "    '08 M. mulatta',\n",
        "    '09 M. thibetana',\n",
        "    '10 R. roxellana',\n",
        "    '11 T. francoisi',\n",
        "    '12 N. leucogenys',\n",
        "    '13 S. syndactylus',\n",
        "    '14 M. murinus',\n",
        "    '15 C. jacchus',\n",
        "    '16 L. catta',\n",
        "]\n",
        "\n",
        "# Ruta base de la carpeta donde se guardarán las imágenes\n",
        "ruta_base = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/\"\n",
        "\n",
        "# Para cada conjunto de datos, genera y guarda el gráfico\n",
        "for i, datos in enumerate(conjuntos_de_datos, start=1):\n",
        "    # Ruta de la carpeta específica para esta especie\n",
        "    carpeta_especie = os.path.join(ruta_base, etiquetas_especies[i-1])\n",
        "\n",
        "    # Si la carpeta no existe, se crea\n",
        "    if not os.path.exists(carpeta_especie):\n",
        "        os.makedirs(carpeta_especie)\n",
        "\n",
        "    # Cambiar el directorio actual a la carpeta de la especie\n",
        "    os.chdir(carpeta_especie)\n",
        "\n",
        "    # Crear una nueva figura con el tamaño deseado\n",
        "    plt.figure(figsize=(20, 11))\n",
        "    # Calcular frecuencias y ordenar por frecuencia\n",
        "    freq_list = sorted(list(datos.values()), reverse=True)\n",
        "\n",
        "    # Crear un gráfico de barras\n",
        "    plt.bar(\n",
        "        range(1, len(freq_list) + 1),\n",
        "        freq_list,\n",
        "        label=f'{nombres_especies[i-1]}',  # Usar el nombre de la especie en lugar de \"Datos\"\n",
        "        color='blue',    # Color de las barras\n",
        "        alpha=0.7        # Transparencia\n",
        "    )\n",
        "\n",
        "    # Configurar etiquetas y título\n",
        "    plt.xlabel('Range')\n",
        "    plt.ylabel('Frequency')\n",
        "    plt.title(f'Frequency Plot for {nombres_especies[i-1]} of size 7-mers')\n",
        "\n",
        "    # Mostrar leyenda\n",
        "    plt.legend()\n",
        "\n",
        "    # Guardar el gráfico con el nombre deseado\n",
        "    nombre_archivo = etiquetas_especies[i-1] + \"_k7.png\"\n",
        "    plt.savefig(nombre_archivo)\n",
        "\n",
        "    # Cerrar la figura para liberar memoria\n",
        "    plt.close()\n",
        "\n",
        "    # Cambiar el directorio de trabajo de vuelta al original\n",
        "    os.chdir(ruta_base)\n"
      ],
      "metadata": {
        "id": "c0ofVYb57jCc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "628WHvRz_vPe"
      },
      "source": [
        "## **LISTA DE  K-MERS ORDENADOS DE MAYOR A MENOR**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ymeRr9gsAjTs"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d1_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/01 H. sapiens/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_1}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_1}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_1}.txt', ruta_destino + f'{NAME_1}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_1}.txt'\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Xy4GZBS_BLJg"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d2_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/02 G. gorilla/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_2}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_2}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_2}.txt', ruta_destino + f'{NAME_2}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_2}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qUYzeTh-BLpe"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d3_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/03 P. paniscus/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_3}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_3}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_3}.txt', ruta_destino + f'{NAME_3}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_3}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lLSYROJCBMH4"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d4_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/04 P. troglodytes/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_4}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_4}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_4}.txt', ruta_destino + f'{NAME_4}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_4}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZqbyhvbdBMgy"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d5_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/05 P. abelii/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_5}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_5}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_5}.txt', ruta_destino + f'{NAME_5}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_5}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "prfn4nveBM6o"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d6_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/06 P. pygmaeus/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_6}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_6}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_6}.txt', ruta_destino + f'{NAME_6}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_6}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "IaObY32RBNVQ"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d7_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/07 M. fascicularis/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_7}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_7}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_7}.txt', ruta_destino + f'{NAME_7}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_7}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wlDY482DBNvW"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d8_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/08 M. mulatta/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_8}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_8}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_8}.txt', ruta_destino + f'{NAME_8}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_8}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "m962FY9-BOJW"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d9_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/09 M. thibetana/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_9}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_9}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_9}.txt', ruta_destino + f'{NAME_9}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_9}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vJBg2Wv9BOj2"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d10_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/10 R. roxellana/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_10}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_10}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_10}.txt', ruta_destino + f'{NAME_10}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_10}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "i0D6vhqZBO82"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d11_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/11 T. francoisi/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_11}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_11}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_11}.txt', ruta_destino + f'{NAME_11}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_11}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZXjn0wDLBPWO"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d12_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/12 N. leucogenys/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_12}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_12}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_12}.txt', ruta_destino + f'{NAME_12}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_12}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "2sECt_XzBPuw"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d13_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/13 S. syndactylus/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_13}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_13}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_13}.txt', ruta_destino + f'{NAME_13}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_13}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_33uwUdTBQGC"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d14_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/14 M. murinus/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_14}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_14}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_14}.txt', ruta_destino + f'{NAME_14}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_14}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bj4Smy8zBQd8"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d15_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/15 C. jacchus/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_15}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_15}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_15}.txt', ruta_destino + f'{NAME_15}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_15}.txt'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "G2GekeUbBQ38"
      },
      "outputs": [],
      "source": [
        "# Supongamos que tienes el diccionario 'dero'\n",
        "dero = d16_7\n",
        "\n",
        "# Crear un diccionario defaultdict para agrupar valores similares\n",
        "grupos = defaultdict(list)\n",
        "\n",
        "# Llenar el diccionario con los valores similares\n",
        "for clave, valor in dero.items():\n",
        "    grupos[valor].append(clave)\n",
        "\n",
        "# Crear un diccionario de conteo para contar las ocurrencias en cada grupo\n",
        "conteo_grupos = {valor: len(claves) for valor, claves in grupos.items()}\n",
        "\n",
        "# Ordenar el diccionario de conteo de mayor a menor por el valor\n",
        "conteo_grupos_ordenado = dict(sorted(conteo_grupos.items(), key=itemgetter(0), reverse=True))\n",
        "\n",
        "\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/K-mer nucleotide appearance frequency and composition/16 L. catta/\"\n",
        "\n",
        "# Guardar la salida en un archivo de texto con formato Markdown\n",
        "with open(f'{NAME_16}.txt', 'w') as archivo:\n",
        "    # Agregar título de encabezado\n",
        "    archivo.write(f\"## Frequency of occurrence of 7-mers for {NAME_16}\\n\\n\")\n",
        "    for valor, conteo in conteo_grupos_ordenado.items():\n",
        "        # Utilizar formato Markdown para texto en negrita\n",
        "        linea = f\"**Frequency of appearance {valor}:** {conteo} **Nucleotide k-mer Composition** - {grupos[valor]}\\n\"\n",
        "        archivo.write(linea)\n",
        "\n",
        "# Mover el archivo a la ruta de destino\n",
        "shutil.move(f'{NAME_16}.txt', ruta_destino + f'{NAME_16}.txt')\n",
        "\n",
        "print(f\"Resultados guardados en '{ruta_destino}{NAME_16}.txt'\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dx3gQdLGDILX"
      },
      "source": [
        "# **GENERACION DE FRACTALES**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "A121eF6RLPZt"
      },
      "source": [
        "## **COLOR Y RUTA DE GUADADO**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ezX-D-eILWaj"
      },
      "outputs": [],
      "source": [
        "color_1 = color_general = cm.flag_r\n",
        "color_2 = cm.tab20c\n",
        "output_directory_1 = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material\"\n",
        "output_directory_2 = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/03 Fractal color normalized/\"\n",
        "output_directory_3 = \"/content/drive/MyDrive/TESIS_FINAL/Suplementary/02 Supplementari material/04 Landscape/\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F6jkzoUINXU4"
      },
      "source": [
        "## **NORMALIZADO**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "natGzEbaNae9"
      },
      "outputs": [],
      "source": [
        "# Convertir las listas a arrays NumPy\n",
        "chaos_1_k7 = np.array(chaos_1_k7)\n",
        "chaos_2_k7 = np.array(chaos_2_k7)\n",
        "chaos_3_k7 = np.array(chaos_3_k7)\n",
        "chaos_4_k7 = np.array(chaos_4_k7)\n",
        "chaos_5_k7 = np.array(chaos_5_k7)\n",
        "chaos_6_k7 = np.array(chaos_6_k7)\n",
        "chaos_7_k7 = np.array(chaos_7_k7)\n",
        "chaos_8_k7 = np.array(chaos_8_k7)\n",
        "chaos_9_k7 = np.array(chaos_9_k7)\n",
        "chaos_10_k7 = np.array(chaos_10_k7)\n",
        "chaos_11_k7 = np.array(chaos_11_k7)\n",
        "chaos_12_k7 = np.array(chaos_12_k7)\n",
        "chaos_13_k7 = np.array(chaos_13_k7)\n",
        "chaos_14_k7 = np.array(chaos_14_k7)\n",
        "chaos_15_k7 = np.array(chaos_15_k7)\n",
        "chaos_16_k7 = np.array(chaos_16_k7)\n",
        "# Crear un objeto Normalize para ambas imágenes (usando los valores mínimo y máximo de todas las imágenes)\n",
        "global_min = min(chaos_1_k7.min(), chaos_2_k7.min(), chaos_3_k7.min(), chaos_4_k7.min(), chaos_5_k7.min(), chaos_6_k7.min(), chaos_7_k7.min(), chaos_8_k7.min(), chaos_9_k7.min(), chaos_10_k7.min(), chaos_11_k7.min(), chaos_12_k7.min(), chaos_13_k7.min(), chaos_14_k7.min(), chaos_15_k7.min(), chaos_16_k7.min() )\n",
        "global_max = max(chaos_1_k7.max(), chaos_2_k7.max(), chaos_3_k7.max(), chaos_4_k7.max(), chaos_5_k7.max(), chaos_6_k7.max(), chaos_7_k7.max(), chaos_8_k7.max(), chaos_9_k7.max(), chaos_10_k7.max(), chaos_11_k7.max(), chaos_12_k7.max(), chaos_13_k7.max(), chaos_14_k7.max(), chaos_15_k7.max(), chaos_16_k7.max())\n",
        "global_normalization = Normalize(vmin=global_min, vmax=global_max)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lregAz5GKztF"
      },
      "source": [
        "## **FRACTALES NO NORMALIZADOS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "czZHFw-sMAzF"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$ ')\n",
        "    ax.set_ylabel('Total pixels for $2^7$ ')\n",
        "    ax.set_title(f'FCGR for {NAME1} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_1}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_1_k7, f'{NAME1}_k7')\n",
        "pylab.title(f'FCGR for {NAME1} / 7-mers')\n",
        "pylab.imshow(chaos_1_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CfiW_pphMMuL"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME2} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_2}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_2_k7, f'{NAME2}_k7')\n",
        "pylab.title(f'FCGR for {NAME2} / 7-mers')\n",
        "pylab.imshow(chaos_2_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "pylab.show()\n",
        "gc.collect()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6JG2t8FnMVlV"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME3} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_3}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_3_k7, f'{NAME3}_k7')\n",
        "pylab.title(f'FCGR for {NAME3} / 7-mers')\n",
        "pylab.imshow(chaos_3_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_MaI56V4PGQh"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME4} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_4}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_4_k7, f'{NAME4}_k7')\n",
        "pylab.title(f'FCGR for {NAME4} / 7-mers')\n",
        "pylab.imshow(chaos_4_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iAznpUnwPIep"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME5} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_5}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_5_k7, f'{NAME5}_k7')\n",
        "pylab.title(f'FCGR for {NAME5} / 7-mers')\n",
        "pylab.imshow(chaos_5_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rn3NclCpPLrp"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME6} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_6}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_6_k7, f'{NAME6}_k7')\n",
        "pylab.title(f'FCGR for {NAME6} / 7-mers')\n",
        "pylab.imshow(chaos_6_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3fwnwr99PUdt"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME7} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_7}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "save_image(chaos_7_k7, f'{NAME7}_k7')\n",
        "pylab.title(f'FCGR for {NAME7} / 7-mers')\n",
        "pylab.imshow(chaos_7_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OObGwirWPaPD"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for  {NAME8} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_8}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_8_k7, f'{NAME8}_k7')\n",
        "pylab.title(f'FCGR for  {NAME8} / 7-mers')\n",
        "pylab.imshow(chaos_8_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jrCoyR3uPc9M"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for  {NAME9} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_9}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_9_k7, f'{NAME9}_k7')\n",
        "pylab.title(f'FCGR for {NAME9} / 7-mers')\n",
        "pylab.imshow(chaos_9_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-gdmjxjYPf19"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for  {NAME10} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_10}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "save_image(chaos_10_k7, f'{NAME10}_k7')\n",
        "pylab.title(f'FCGR for {NAME10} / 7-mers')\n",
        "pylab.imshow(chaos_10_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "C8aJkeBkPixK"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME11} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_11}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_11_k7, f'{NAME11}_k7')\n",
        "pylab.title(f'FCGR for {NAME11} / 7-mers')\n",
        "pylab.imshow(chaos_11_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "9y2j2FURPpa8"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for  {NAME12} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_12}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_12_k7, f'{NAME12}_k7')\n",
        "pylab.title(f'FCGR for {NAME12} / 7-mers')\n",
        "pylab.imshow(chaos_12_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dvvGgjw7PvSU"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for  {NAME13} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_13}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_13_k7, f'{NAME13}_k7')\n",
        "pylab.title(f'FCGR for {NAME13} / 7-mers')\n",
        "pylab.imshow(chaos_13_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "GFxmcda8Pzwv"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for  {NAME14} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_14}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_14_k7, f'{NAME14}_k7')\n",
        "pylab.title(f'FCGR for {NAME14} / 7-mers')\n",
        "pylab.imshow(chaos_14_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WMLIOk_WP3j6"
      },
      "outputs": [],
      "source": [
        "\n",
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME15} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_15}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_15_k7, f'{NAME15}_k7')\n",
        "pylab.title(f'FCGR for {NAME15} / 7-mers')\n",
        "pylab.imshow(chaos_15_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tw0HR9a1P7Yg"
      },
      "outputs": [],
      "source": [
        "# Función para guardar una imagen con un nombre específico\n",
        "def save_image(image, name):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)  # Ajusta el tamaño y la resolución aquí\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME16} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory_1, f'{NAME_16}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_16_k7, f'{NAME16}_k7')\n",
        "pylab.title(f'FCGR for {NAME16} / 7-mers')\n",
        "pylab.imshow(chaos_16_k7, interpolation='nearest', cmap=color_1)\n",
        "plt.xlabel('Total pixels for $2^7$')\n",
        "plt.ylabel('Total pixels for $2^7$')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "roY9jbf1K3rV"
      },
      "source": [
        "## **FRACTALES NORMALIZADOS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0oTChpl7RC49"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME1} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_1}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_1_k7, f'{NAME1}_k7', f'FCGR for {NAME1} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME1} / 7-mers')\n",
        "pylab.imshow(chaos_1_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LkD7akAAW7Tw"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME2} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_2}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_2_k7, f'{NAME2}_k7', f'FCGR for {NAME2} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME2} / 7-mers')\n",
        "pylab.imshow(chaos_2_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lMF_FQ0JW_ON"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME3} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_3}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_3_k7, f'{NAME3}_k7', f'FCGR for {NAME3} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME3} / 7-mers')\n",
        "pylab.imshow(chaos_3_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ly-KLl-qXKxa"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME4} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_4}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_4_k7, f'{NAME4}_k7', f'FCGR for {NAME1} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME1} / 7-mers')\n",
        "pylab.imshow(chaos_4_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JEp0_b4JXN1s"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME5} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_5}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_5_k7, f'{NAME5}_k7', f'FCGR for {NAME5} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME1} / 7-mers')\n",
        "pylab.imshow(chaos_5_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wI8mnNrUXQnx"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME6} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_6}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_6_k7, f'{NAME6}_k7', f'FCGR for {NAME6} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME6} / 7-mers')\n",
        "pylab.imshow(chaos_6_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PPTP4F0pYfTl"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME7} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_7}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_7_k7, f'{NAME7}_k7', f'FCGR for {NAME7} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME7} / 7-mers')\n",
        "pylab.imshow(chaos_7_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vfqua5vEYirW"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME8} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_8}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_8_k7, f'{NAME8}_k7', f'FCGR for {NAME8} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME8} / 7-mers')\n",
        "pylab.imshow(chaos_8_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ntfe11vhYnbx"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME9} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_9}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_9_k7, f'{NAME9}_k7', f'FCGR for {NAME9} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME9} / 7-mers')\n",
        "pylab.imshow(chaos_9_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "nIGii8XwYqa2"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME10} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_10}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_10_k7, f'{NAME10}_k7', f'FCGR for {NAME10} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME10} / 7-mers')\n",
        "pylab.imshow(chaos_10_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KPxQftDhYtuy"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME11} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_11}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_11_k7, f'{NAME11}_k7', f'FCGR for {NAME11} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME11} / 7-mers')\n",
        "pylab.imshow(chaos_11_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "VzE14uylYveW"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME12} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_12}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_12_k7, f'{NAME12}_k7', f'FCGR for {NAME12} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME12} / 7-mers')\n",
        "pylab.imshow(chaos_12_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vGA1wGatYxdz"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME13} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_13}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_13_k7, f'{NAME13}_k7', f'FCGR for {NAME13} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME13} / 7-mers')\n",
        "pylab.imshow(chaos_13_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "RQAdxQrGYzFG"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME14} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_14}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_14_k7, f'{NAME14}_k7', f'FCGR for {NAME14} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME14} / 7-mers')\n",
        "pylab.imshow(chaos_14_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "TjlQQNOgY1YI"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME15} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_15}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_15_k7, f'{NAME15}_k7', f'FCGR for {NAME15} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME15} / 7-mers')\n",
        "pylab.imshow(chaos_15_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qBQYjMWtY3VU"
      },
      "outputs": [],
      "source": [
        "def save_image(image, name, title, output_directory, color_map, normalization):\n",
        "    fig, ax = plt.subplots(figsize=(6, 6), dpi=1200)\n",
        "    ax.imshow(image, interpolation='nearest', cmap=color_1, norm=normalization)\n",
        "    ax.set_xlabel('Total pixels for $2^7$')\n",
        "    ax.set_ylabel('Total pixels for $2^7$')\n",
        "    ax.set_title(f'FCGR for {NAME16} / 7-mers')\n",
        "    fig.savefig(os.path.join(output_directory, f'{NAME_16}.png'))\n",
        "    plt.close(fig)  # Cierra la figura actual para liberar memoria\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "save_image(chaos_16_k7, f'{NAME16}_k7', f'FCGR for {NAME16} / 7-mers', output_directory_2, color_1, global_normalization)\n",
        "pylab.title(f'FCGR for {NAME16} / 7-mers')\n",
        "pylab.imshow(chaos_16_k7, interpolation='nearest', cmap=color_1, norm=global_normalization)\n",
        "plt.xlabel('Total pixels for $2^7$ ')\n",
        "plt.ylabel('Total pixels for $2^7$ ')\n",
        "pylab.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PqC3NaYrK7nA"
      },
      "source": [
        "## **LANDSCAPE**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EYYEVwS3Z8ol"
      },
      "outputs": [],
      "source": [
        "#DPI\n",
        "res = 800"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "B29gadZ_ZguK"
      },
      "outputs": [],
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_1_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_1_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME1} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_1} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "fNDjKhUUacF8"
      },
      "outputs": [],
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_2_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_2_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME2} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_2} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WgRDDO3waeNF"
      },
      "outputs": [],
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_3_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_3_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME3} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_3} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0fQJ9LzyaiIx"
      },
      "outputs": [],
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_4_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_4_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME4} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_4} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_5_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_5_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME5} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_5} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()\n"
      ],
      "metadata": {
        "id": "uvDNR99ybb7q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_6_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_6_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME6} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_6} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "gmSpOtzDbd_2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_7_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_7_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME7} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_7} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()\n"
      ],
      "metadata": {
        "id": "0JnAgUeFbgDJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_8_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_8_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME8} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_8} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "KfKWwAXDbiUF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_9_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_9_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME9} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_9} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "T-qbSuvQbk0-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_10_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_10_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME10} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_10} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "26Yu5oGybnfU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_11_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_11_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME11} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_11} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()\n"
      ],
      "metadata": {
        "id": "53zsuhd2brUT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_12_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_12_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME12} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_12} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "7g8kEYBubucM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_13_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_13_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME13} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_13} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()\n"
      ],
      "metadata": {
        "id": "PdzjitsNbwTf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_14_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_14_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME14} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_14} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "EyHrHmqTbyRi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_15_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_15_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME15} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_15} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "482XXMcMb01K"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#GENERACION DEL LANDSCAPE\n",
        "# Definir el tamaño del paisaje (resolución)\n",
        "resolution1 = len(chaos_16_k7)\n",
        "# Generar coordenadas en el rango [0, 1)\n",
        "x = np.linspace(0, 1, resolution1)\n",
        "y = np.linspace(0, 1, resolution1)\n",
        "\n",
        "# Crear una cuadrícula de coordenadas\n",
        "X, Y = np.meshgrid(x, y)\n",
        "\n",
        "# Utilizar las alturas de chaos_X_k7\n",
        "Z = np.array(chaos_16_k7).reshape(resolution1, resolution1)\n",
        "Z_normalized = global_normalization(Z)\n",
        "\n",
        "# Configurar el colormap y la visualización en 3D\n",
        "fig = plt.figure(figsize=(20, 15), dpi= res)\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "# Ajustar la elevación y el ángulo de visión\n",
        "ax.view_init(elev=45, azim=45)\n",
        "\n",
        "# Crear la superficie tridimensional\n",
        "surf = ax.plot_surface(X, Y, Z_normalized, cmap=color_2, edgecolor='k', linewidth=0.01)\n",
        "\n",
        "# Configurar etiquetas\n",
        "ax.set_xlabel('X')\n",
        "ax.set_ylabel('Y')\n",
        "ax.set_zlabel('Z')\n",
        "ax.set_title(f'Landscape in 3D for {NAME16} / 7-mers')\n",
        "\n",
        "# Añadir una barra de color\n",
        "norm = Normalize(vmin=global_min, vmax=global_max)\n",
        "mappable = plt.cm.ScalarMappable(cmap=color_2, norm=norm)\n",
        "mappable.set_array(Z_normalized)\n",
        "barra_color = fig.colorbar(mappable, ax=ax, shrink=0.5, aspect=10)\n",
        "\n",
        "\n",
        "nombre_archivo = f'{NAME_16} Landscape 7-mers.png'\n",
        "ruta_completa = os.path.join(output_directory_3, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar la imagen\n",
        "plt.show()\n",
        "# Recolectar y liberar memoria\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "p89dbuwFb3YQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **DENDROGRAMA**"
      ],
      "metadata": {
        "id": "aO90-chlgChg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Suponiendo que chaos_1_k3 y chaos_14_k3 son listas que deben convertirse a arrays de NumPy\n",
        "chaos_1_k7 = np.array(chaos_1_k7)\n",
        "chaos_2_k7 = np.array(chaos_2_k7)\n",
        "chaos_3_k7 = np.array(chaos_3_k7)\n",
        "chaos_4_k7 = np.array(chaos_4_k7)\n",
        "chaos_5_k7 = np.array(chaos_5_k7)\n",
        "chaos_6_k7 = np.array(chaos_6_k7)\n",
        "chaos_7_k7 = np.array(chaos_7_k7)\n",
        "chaos_8_k7 = np.array(chaos_8_k7)\n",
        "chaos_9_k7 = np.array(chaos_9_k7)\n",
        "chaos_10_k7 = np.array(chaos_10_k7)\n",
        "chaos_11_k7 = np.array(chaos_11_k7)\n",
        "chaos_12_k7 = np.array(chaos_12_k7)\n",
        "chaos_13_k7 = np.array(chaos_13_k7)\n",
        "chaos_14_k7 = np.array(chaos_14_k7)\n",
        "chaos_15_k7 = np.array(chaos_15_k7)\n",
        "chaos_16_k7 = np.array(chaos_16_k7)\n",
        "\n",
        "# Concatenar las matrices aplanadas\n",
        "flattened_matrices = np.vstack([\n",
        "    chaos_1_k7.flatten(),\n",
        "    chaos_2_k7.flatten(),\n",
        "    chaos_3_k7.flatten(),\n",
        "    chaos_4_k7.flatten(),\n",
        "    chaos_5_k7.flatten(),\n",
        "    chaos_6_k7.flatten(),\n",
        "    chaos_7_k7.flatten(),\n",
        "    chaos_8_k7.flatten(),\n",
        "    chaos_9_k7.flatten(),\n",
        "    chaos_10_k7.flatten(),\n",
        "    chaos_11_k7.flatten(),\n",
        "    chaos_12_k7.flatten(),\n",
        "    chaos_13_k7.flatten(),\n",
        "    chaos_14_k7.flatten(),\n",
        "    chaos_15_k7.flatten(),\n",
        "    chaos_16_k7.flatten()\n",
        "\n",
        "])\n",
        "\n",
        "# Cálculo de la distancia euclidiana entre las matrices aplanadas\n",
        "distance_matrix = pdist(flattened_matrices, metric='euclidean')\n",
        "\n",
        "\n",
        "# Construcción del dendrograma sin squareform\n",
        "plt.figure(figsize=(12, 15))\n",
        "dendrogram(linkage(distance_matrix, method= 'average'), labels=[\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M.a fascicularis}$',\n",
        "    r'$\\mathit{M. thibetana thibetana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$'\n",
        "])\n",
        "# Rotar las etiquetas del eje x a 45 grados\n",
        "plt.xticks(rotation=45, ha='right')\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/DENDROGRAMA/\"\n",
        "\n",
        "# Guardar la imagen en la ruta de destino\n",
        "plt.savefig(ruta_destino + \"dendrograma1.png\")\n",
        "\n",
        "# Mostrar el dendrograma\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "Cyg7H_MrHCAm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Suponiendo que chaos_1_k3 y chaos_14_k3 son listas que deben convertirse a arrays de NumPy\n",
        "chaos_1_k7 = np.array(chaos_1_k7)\n",
        "chaos_2_k7 = np.array(chaos_2_k7)\n",
        "chaos_3_k7 = np.array(chaos_3_k7)\n",
        "chaos_4_k7 = np.array(chaos_4_k7)\n",
        "chaos_5_k7 = np.array(chaos_5_k7)\n",
        "chaos_6_k7 = np.array(chaos_6_k7)\n",
        "chaos_7_k7 = np.array(chaos_7_k7)\n",
        "chaos_8_k7 = np.array(chaos_8_k7)\n",
        "chaos_9_k7 = np.array(chaos_9_k7)\n",
        "chaos_10_k7 = np.array(chaos_10_k7)\n",
        "chaos_11_k7 = np.array(chaos_11_k7)\n",
        "chaos_12_k7 = np.array(chaos_12_k7)\n",
        "chaos_13_k7 = np.array(chaos_13_k7)\n",
        "chaos_14_k7 = np.array(chaos_14_k7)\n",
        "chaos_15_k7 = np.array(chaos_15_k7)\n",
        "chaos_16_k7 = np.array(chaos_16_k7)\n",
        "\n",
        "# Concatenar las matrices aplanadas\n",
        "flattened_matrices = np.vstack([\n",
        "    chaos_1_k7.flatten(),\n",
        "    chaos_2_k7.flatten(),\n",
        "    chaos_3_k7.flatten(),\n",
        "    chaos_4_k7.flatten(),\n",
        "    chaos_5_k7.flatten(),\n",
        "    chaos_6_k7.flatten(),\n",
        "    chaos_7_k7.flatten(),\n",
        "    chaos_8_k7.flatten(),\n",
        "    chaos_9_k7.flatten(),\n",
        "    chaos_10_k7.flatten(),\n",
        "    chaos_11_k7.flatten(),\n",
        "    chaos_12_k7.flatten(),\n",
        "    chaos_13_k7.flatten(),\n",
        "    chaos_14_k7.flatten(),\n",
        "    chaos_15_k7.flatten(),\n",
        "    chaos_16_k7.flatten()\n",
        "\n",
        "])\n",
        "\n",
        "# Cálculo de la distancia euclidiana entre las matrices aplanadas\n",
        "distance_matrix = pdist(flattened_matrices, metric='euclidean')\n",
        "\n",
        "# Construcción del dendrograma sin squareform\n",
        "plt.figure(figsize=(15, 12))\n",
        "linkage_matrix = linkage(distance_matrix, method='average')\n",
        "\n",
        "# Etiquetas con números e intervalos\n",
        "labels = [\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M.a fascicularis}$',\n",
        "    r'$\\mathit{M. thibetana thibetana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$'\n",
        "]\n",
        "\n",
        "\n",
        "\n",
        "# Dendrograma\n",
        "dendrogram(linkage_matrix, labels=labels, orientation='right')\n",
        "\n",
        "# Anotaciones con números e intervalos\n",
        "for i, (label, color) in enumerate(zip(labels, linkage_matrix[:, 2])):\n",
        "    plt.annotate(f\"{i + 1}: {label}\", (10, color), textcoords=\"offset points\", xytext=(-10,0), ha='center')\n",
        "\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/DENDROGRAMA/\"\n",
        "\n",
        "# Guardar la imagen en la ruta de destino\n",
        "plt.savefig(ruta_destino + \"dendrograma2.png\")\n",
        "\n",
        "# Mostrar el dendrograma\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "RJD_Vs4uyvOM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### prueba que medio ajusta"
      ],
      "metadata": {
        "id": "th1o7TUMHCVx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Suponiendo que chaos_1_k3 y chaos_14_k3 son listas que deben convertirse a arrays de NumPy\n",
        "chaos_1_k7 = np.array(chaos_1_k7)\n",
        "chaos_2_k7 = np.array(chaos_2_k7)\n",
        "chaos_3_k7 = np.array(chaos_3_k7)\n",
        "chaos_4_k7 = np.array(chaos_4_k7)\n",
        "chaos_5_k7 = np.array(chaos_5_k7)\n",
        "chaos_6_k7 = np.array(chaos_6_k7)\n",
        "chaos_7_k7 = np.array(chaos_7_k7)\n",
        "chaos_8_k7 = np.array(chaos_8_k7)\n",
        "chaos_9_k7 = np.array(chaos_9_k7)\n",
        "chaos_10_k7 = np.array(chaos_10_k7)\n",
        "chaos_11_k7 = np.array(chaos_11_k7)\n",
        "chaos_12_k7 = np.array(chaos_12_k7)\n",
        "chaos_13_k7 = np.array(chaos_13_k7)\n",
        "chaos_14_k7 = np.array(chaos_14_k7)\n",
        "chaos_15_k7 = np.array(chaos_15_k7)\n",
        "chaos_16_k7 = np.array(chaos_16_k7)\n",
        "\n",
        "# Concatenar las matrices aplanadas\n",
        "flattened_matrices = np.vstack([\n",
        "    chaos_1_k7.flatten(),\n",
        "    chaos_2_k7.flatten(),\n",
        "    chaos_3_k7.flatten(),\n",
        "    chaos_4_k7.flatten(),\n",
        "    chaos_5_k7.flatten(),\n",
        "    chaos_6_k7.flatten(),\n",
        "    chaos_7_k7.flatten(),\n",
        "    chaos_8_k7.flatten(),\n",
        "    chaos_9_k7.flatten(),\n",
        "    chaos_10_k7.flatten(),\n",
        "    chaos_11_k7.flatten(),\n",
        "    chaos_12_k7.flatten(),\n",
        "    chaos_13_k7.flatten(),\n",
        "    chaos_14_k7.flatten(),\n",
        "    chaos_15_k7.flatten(),\n",
        "    chaos_16_k7.flatten()\n",
        "\n",
        "])\n",
        "\n",
        "# Cálculo de la distancia euclidiana entre las matrices aplanadas\n",
        "distance_matrix = pdist(flattened_matrices, metric='euclidean')\n",
        "\n",
        "\n",
        "# Construcción del dendrograma sin squareform\n",
        "plt.figure(figsize=(12, 15))\n",
        "dendrogram(linkage(distance_matrix, method= 'single'), labels=[\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M.a fascicularis}$',\n",
        "    r'$\\mathit{M. thibetana thibetana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$'\n",
        "])\n",
        "# Rotar las etiquetas del eje x a 45 grados\n",
        "plt.xticks(rotation=45, ha='right')\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/DENDROGRAMA/\"\n",
        "\n",
        "# Guardar la imagen en la ruta de destino\n",
        "plt.savefig(ruta_destino + \"dendrograma3.png\")\n",
        "\n",
        "# Mostrar el dendrograma\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "PrIii3KzyrNu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import matplotlib.pyplot as plt\n",
        "from scipy.cluster.hierarchy import dendrogram, linkage\n",
        "from scipy.spatial.distance import pdist\n",
        "import numpy as np\n",
        "\n",
        "# Suponiendo que chaos_1_k3 y chaos_14_k3 son listas que deben convertirse a arrays de NumPy\n",
        "chaos_1_k7 = np.array(chaos_1_k7)\n",
        "chaos_2_k7 = np.array(chaos_2_k7)\n",
        "chaos_3_k7 = np.array(chaos_3_k7)\n",
        "chaos_4_k7 = np.array(chaos_4_k7)\n",
        "chaos_5_k7 = np.array(chaos_5_k7)\n",
        "chaos_6_k7 = np.array(chaos_6_k7)\n",
        "chaos_7_k7 = np.array(chaos_7_k7)\n",
        "chaos_8_k7 = np.array(chaos_8_k7)\n",
        "chaos_9_k7 = np.array(chaos_9_k7)\n",
        "chaos_10_k7 = np.array(chaos_10_k7)\n",
        "chaos_11_k7 = np.array(chaos_11_k7)\n",
        "chaos_12_k7 = np.array(chaos_12_k7)\n",
        "chaos_13_k7 = np.array(chaos_13_k7)\n",
        "chaos_14_k7 = np.array(chaos_14_k7)\n",
        "chaos_15_k7 = np.array(chaos_15_k7)\n",
        "chaos_16_k7 = np.array(chaos_16_k7)\n",
        "\n",
        "# Concatenar las matrices aplanadas\n",
        "flattened_matrices = np.vstack([\n",
        "    chaos_1_k7.flatten(),\n",
        "    chaos_2_k7.flatten(),\n",
        "    chaos_3_k7.flatten(),\n",
        "    chaos_4_k7.flatten(),\n",
        "    chaos_5_k7.flatten(),\n",
        "    chaos_6_k7.flatten(),\n",
        "    chaos_7_k7.flatten(),\n",
        "    chaos_8_k7.flatten(),\n",
        "    chaos_9_k7.flatten(),\n",
        "    chaos_10_k7.flatten(),\n",
        "    chaos_11_k7.flatten(),\n",
        "    chaos_12_k7.flatten(),\n",
        "    chaos_13_k7.flatten(),\n",
        "    chaos_14_k7.flatten(),\n",
        "    chaos_15_k7.flatten(),\n",
        "    chaos_16_k7.flatten()\n",
        "])\n",
        "\n",
        "# Cálculo de la distancia euclidiana entre las matrices aplanadas\n",
        "distance_matrix = pdist(flattened_matrices, metric='euclidean')\n",
        "\n",
        "# Construcción del dendrograma sin squareform\n",
        "plt.figure(figsize=(8, 15))  # Tamaño ajustable de la figura\n",
        "dendrogram(linkage(distance_matrix, method='average'), labels=[\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M.a fascicularis}$',\n",
        "    r'$\\mathit{M. thibetana thibetana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$'\n",
        "])\n",
        "# Rotar las etiquetas del eje x a 45 grados\n",
        "plt.xticks(rotation=45, ha='right')\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "\n",
        "# Ruta de destino\n",
        "ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/DENDROGRAMA/\"\n",
        "\n",
        "# Guardar la imagen en la ruta de destino\n",
        "plt.savefig(ruta_destino + \"dendrograma.png\")\n",
        "\n",
        "# Mostrar el dendrograma\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "E0OTtRMNCobD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "from scipy.cluster.hierarchy import linkage, dendrogram, cophenet\n",
        "from scipy.spatial.distance import pdist\n",
        "\n",
        "# Suponiendo que chaos_1_k3 y chaos_14_k3 son listas que deben convertirse a arrays de NumPy\n",
        "chaos_1_k7 = np.array(chaos_1_k7)\n",
        "chaos_2_k7 = np.array(chaos_2_k7)\n",
        "chaos_3_k7 = np.array(chaos_3_k7)\n",
        "chaos_4_k7 = np.array(chaos_4_k7)\n",
        "chaos_5_k7 = np.array(chaos_5_k7)\n",
        "chaos_6_k7 = np.array(chaos_6_k7)\n",
        "chaos_7_k7 = np.array(chaos_7_k7)\n",
        "chaos_8_k7 = np.array(chaos_8_k7)\n",
        "chaos_9_k7 = np.array(chaos_9_k7)\n",
        "chaos_10_k7 = np.array(chaos_10_k7)\n",
        "chaos_11_k7 = np.array(chaos_11_k7)\n",
        "chaos_12_k7 = np.array(chaos_12_k7)\n",
        "chaos_13_k7 = np.array(chaos_13_k7)\n",
        "chaos_14_k7 = np.array(chaos_14_k7)\n",
        "chaos_15_k7 = np.array(chaos_15_k7)\n",
        "chaos_16_k7 = np.array(chaos_16_k7)\n",
        "\n",
        "# Concatenar las matrices aplanadas\n",
        "flattened_matrices = np.vstack([\n",
        "    chaos_1_k7.flatten(),\n",
        "    chaos_2_k7.flatten(),\n",
        "    chaos_3_k7.flatten(),\n",
        "    chaos_4_k7.flatten(),\n",
        "    chaos_5_k7.flatten(),\n",
        "    chaos_6_k7.flatten(),\n",
        "    chaos_7_k7.flatten(),\n",
        "    chaos_8_k7.flatten(),\n",
        "    chaos_9_k7.flatten(),\n",
        "    chaos_10_k7.flatten(),\n",
        "    chaos_11_k7.flatten(),\n",
        "    chaos_12_k7.flatten(),\n",
        "    chaos_13_k7.flatten(),\n",
        "    chaos_14_k7.flatten(),\n",
        "    chaos_15_k7.flatten(),\n",
        "    chaos_16_k7.flatten()\n",
        "])\n",
        "\n",
        "# Cálculo de la distancia euclidiana entre las matrices aplanadas\n",
        "distance_matrix = pdist(flattened_matrices, metric='euclidean')\n",
        "\n",
        "# Generar la matriz de enlace\n",
        "Z = linkage(distance_matrix, method='average')\n",
        "\n",
        "# Convertir la matriz de enlace a tipo double\n",
        "Z = Z.astype('double')\n",
        "\n",
        "# Construcción del dendrograma sin squareform\n",
        "#plt.figure(figsize=(12, 15))\n",
        "dendrogram_object = dendrogram(Z, labels=[\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{M. fascicularis}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M. thibetana}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$',\n",
        "])\n",
        "# Rotar las etiquetas del eje x a 45 grados\n",
        "plt.xticks(rotation=90, ha='center')\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "# Ruta de destino\n",
        "#ruta_destino = \"/content/drive/MyDrive/TESIS_FINAL/DENDROGRAMA/\"\n",
        "\n",
        "# Guardar la imagen en la ruta de destino\n",
        "#plt.savefig(ruta_destino + \"dendrograma1.png\")\n",
        "\n",
        "# Calcular el índice cophenético\n",
        "cophenet_index, _ = cophenet(Z, distance_matrix)\n",
        "\n",
        "# Agregar el índice cophenético como texto en la figura\n",
        "plt.text(0.5, 0.95, f'Cophenetic index: {cophenet_index:.3f}', horizontalalignment='left', verticalalignment='center', transform=plt.gca().transAxes, fontsize=12)\n",
        "\n",
        "plt.title(r'Similarity dendrogram of $\\mathit{PRKN}$ gene to primate')\n",
        "\n",
        "# Mostrar el dendrograma\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "gt12TBOXMzLn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **POWER LAW**"
      ],
      "metadata": {
        "id": "mrqEE3dBHDPK"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **LOG-LOG**"
      ],
      "metadata": {
        "id": "DYKxqcKXHYG2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Lista de nombres correspondientes a cada conjunto de datos\n",
        "nombres_especies = [\n",
        "    r'$\\mathit{H. sapiens \\ PRKN}$',\n",
        "    r'$\\mathit{G. gorilla \\ PRKN}$',\n",
        "    r'$\\mathit{P. paniscus \\ PRKN}$',\n",
        "    r'$\\mathit{P. troglodytes \\ PRKN}$',\n",
        "    r'$\\mathit{P. abelii \\ PRKN}$',\n",
        "    r'$\\mathit{P. pygmaeus \\ PRKN}$',\n",
        "    r'$\\mathit{M. fascicularis \\ PRKN}$',\n",
        "    r'$\\mathit{M. mulatta \\ PRKN}$',\n",
        "    r'$\\mathit{M. thibetana \\ PRKN}$',\n",
        "    r'$\\mathit{R. roxellana \\ PRKN}$',\n",
        "    r'$\\mathit{T. francoisi \\ PRKN}$',\n",
        "    r'$\\mathit{N. leucogenys \\ PRKN}$',\n",
        "    r'$\\mathit{S. syndactylus \\ PRKN}$',\n",
        "    r'$\\mathit{M. murinus \\ PRKN}$',\n",
        "    r'$\\mathit{C. jacchus \\ PRKN}$',\n",
        "    r'$\\mathit{L. catta \\ PRKN}$',\n",
        "]\n",
        "\n",
        "# Lista de tus conjuntos de datos\n",
        "conjuntos_de_datos = [d1_7, d2_7, d3_7, d4_7, d5_7, d6_7, d7_7, d8_7,\n",
        "                      d9_7, d10_7, d11_7, d12_7, d13_7, d14_7, d15_7, d16_7]\n",
        "\n",
        "# Ruta para guardar la imagen\n",
        "ruta_guardado = '/content/drive/MyDrive/TESIS_FINAL/Suplementary/05 Supplementary material'\n",
        "\n",
        "# Configurar una figura\n",
        "plt.figure(figsize=(20, 11))\n",
        "\n",
        "# Para cada conjunto de datos, genera y muestra el gráfico\n",
        "for i, datos in enumerate(conjuntos_de_datos, start=0):\n",
        "    # Calcular frecuencias y ordenar por frecuencia\n",
        "    freq_list = sorted(list(datos.values()), reverse=True)\n",
        "\n",
        "    # Colores diferentes para cada conjunto de datos\n",
        "    colores = plt.cm.jet(np.linspace(0, 1, len(conjuntos_de_datos)))\n",
        "\n",
        "    # Ajustar valores en el eje x y y\n",
        "    x_values = np.log(range(1, len(freq_list) + 1)) * 1.5\n",
        "    y_values = np.log(freq_list) + 1\n",
        "\n",
        "    # Seleccionar un marcador diferente para cada curva\n",
        "    marcadores = ['o', 's', '^', 'v', 'D', 'p', 'h', '+', 'x', '<', '>', '*', '|', '_', '.', ',']\n",
        "\n",
        "    # Crear un gráfico log-log con líneas y asignar un color y marcador a cada curva\n",
        "    plt.plot(\n",
        "        x_values,\n",
        "        y_values,\n",
        "        label=nombres_especies[i],\n",
        "        marker=marcadores[i % len(marcadores)],\n",
        "        markersize=6,\n",
        "        linestyle='-',\n",
        "        color=colores[i],\n",
        "        alpha=0.7\n",
        "    )\n",
        "\n",
        "# Configurar etiquetas y título\n",
        "plt.ylabel('ln(Frequency)')\n",
        "plt.xlabel('ln(Range)')\n",
        "plt.title(r'General graph for log-log of the $\\mathit{PRKN}$ gene of 16 primates')\n",
        "# Mostrar leyenda\n",
        "plt.legend()\n",
        "\n",
        "# Guardar la imagen en la ruta especificada\n",
        "nombre_archivo = \"00 Log-Log Frequency Histogram of size 7-mers.png\"\n",
        "ruta_completa = os.path.join(ruta_guardado, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar el gráfico\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "7qtk8lrbHGjE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Lista de tus conjuntos de datos\n",
        "conjuntos_de_datos = [d1_7, d2_7, d3_7, d4_7, d5_7, d6_7, d7_7, d8_7,\n",
        "                      d9_7, d10_7, d11_7, d12_7, d13_7, d14_7, d15_7, d16_7]\n",
        "\n",
        "\n",
        "nombres_especies = [\n",
        "    r'$\\mathit{H. sapiens \\ PRKN}$',\n",
        "    r'$\\mathit{G. gorilla \\ PRKN}$',\n",
        "    r'$\\mathit{P. paniscus \\ PRKN}$',\n",
        "    r'$\\mathit{P. troglodytes \\ PRKN}$',\n",
        "    r'$\\mathit{P. abelii \\ PRKN}$',\n",
        "    r'$\\mathit{P. pygmaeus \\ PRKN}$',\n",
        "    r'$\\mathit{M. fascicularis \\ PRKN}$',\n",
        "    r'$\\mathit{M. mulatta \\ PRKN}$',\n",
        "    r'$\\mathit{M. thibetana \\ PRKN}$',\n",
        "    r'$\\mathit{R. roxellana \\ PRKN}$',\n",
        "    r'$\\mathit{T. francoisi \\ PRKN}$',\n",
        "    r'$\\mathit{N. leucogenys \\ PRKN}$',\n",
        "    r'$\\mathit{S. syndactylus \\ PRKN}$',\n",
        "    r'$\\mathit{M. murinus \\ PRKN}$',\n",
        "    r'$\\mathit{C. jacchus \\ PRKN}$',\n",
        "    r'$\\mathit{L. catta \\ PRKN}$',\n",
        "]\n",
        "\n",
        "\n",
        "# Lista de nombres de especies correspondientes a cada conjunto de datos\n",
        "etiquetas_especies = [\n",
        "    '01 H. sapiens',\n",
        "    '02 G. gorilla',\n",
        "    '03 P. paniscus',\n",
        "    '04 P. troglodytes',\n",
        "    '05 P. abelii',\n",
        "    '06 P. pygmaeus',\n",
        "    '07 M. fascicularis',\n",
        "    '08 M. mulatta',\n",
        "    '09 M. thibetana',\n",
        "    '10 R. roxellana',\n",
        "    '11 T. francoisi',\n",
        "    '12 N. leucogenys',\n",
        "    '13 S. syndactylus',\n",
        "    '14 M. murinus',\n",
        "    '15 C. jacchus',\n",
        "    '16 L. catta',\n",
        "\n",
        "]\n",
        "\n",
        "# Ruta de la carpeta donde se guardarán las imágenes\n",
        "ruta_guardado = '/content/drive/MyDrive/TESIS_FINAL/Suplementary/05 Supplementary material/'\n",
        "\n",
        "# Para cada conjunto de datos, genera y guarda el gráfico\n",
        "for i, datos in enumerate(conjuntos_de_datos, start=1):\n",
        "    # Crear una nueva figura con el tamaño deseado\n",
        "    plt.figure(figsize=(20, 11))\n",
        "    # Calcular frecuencias y ordenar por frecuencia\n",
        "    freq_list = sorted(list(datos.values()), reverse=True)\n",
        "\n",
        "    # Crear un histograma log-log con puntos personalizados\n",
        "    plt.scatter(\n",
        "        np.log(range(1, len(freq_list) + 1)),\n",
        "        np.log(freq_list),\n",
        "        label=f'{nombres_especies[i-1]}',  # Usar el nombre de la especie en lugar de \"Datos\"\n",
        "        s=2,        # Tamaño de los puntos\n",
        "        c='blue',    # Color de los puntos\n",
        "        alpha=0.7     # Transparencia\n",
        "    )\n",
        "\n",
        "    # Configurar etiquetas y título\n",
        "    plt.xlabel('ln(Range)')\n",
        "    plt.ylabel('ln(Frequency)')\n",
        "    plt.title(f'Log-Log for {nombres_especies[i-1]} of size 7-mers')\n",
        "\n",
        "    # Mostrar leyenda\n",
        "    plt.legend()\n",
        "\n",
        "    # Guardar el gráfico con el nombre deseado\n",
        "    nombre_archivo = etiquetas_especies[i-1].replace(\" \", \" \").title() + \"_k7.png\"\n",
        "\n",
        "    ruta_completa = ruta_guardado + nombre_archivo\n",
        "    plt.savefig(ruta_completa)\n",
        "\n",
        "    # Mostrar el gráfico\n",
        "    plt.show()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "print(\"Imágenes guardadas exitosamente en:\", ruta_guardado)"
      ],
      "metadata": {
        "id": "GWGZCBnWIfSw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **SEMI-LOG**"
      ],
      "metadata": {
        "id": "4QDTwIy5Iqii"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Lista de nombres correspondientes a cada conjunto de datos\n",
        "nombres_especies = [\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M.a fascicularis}$',\n",
        "    r'$\\mathit{M. thibetana thibetana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$'\n",
        "]\n",
        "\n",
        "# Lista de tus conjuntos de datos\n",
        "conjuntos_de_datos = [d1_7, d2_7, d3_7, d4_7, d5_7, d6_7, d7_7, d8_7,\n",
        "                      d9_7, d10_7, d11_7, d12_7, d13_7, d14_7, d15_7, d16_7]\n",
        "\n",
        "# Ruta para guardar la imagen\n",
        "ruta_guardado = \"/content/drive/MyDrive/TESIS_FINAL/POWER LAW/SEMI-LOG/\"\n",
        "\n",
        "# Configurar una figura\n",
        "plt.figure(figsize=(20, 11))\n",
        "\n",
        "# Para cada conjunto de datos, genera y muestra el gráfico\n",
        "for i, datos in enumerate(conjuntos_de_datos, start=0):\n",
        "    # Calcular frecuencias y ordenar por frecuencia\n",
        "    freq_list = sorted(list(datos.values()), reverse=True)\n",
        "\n",
        "    # Colores diferentes para cada conjunto de datos\n",
        "    colores = plt.cm.jet(np.linspace(0, 1, len(conjuntos_de_datos)))\n",
        "\n",
        "    # Ajustar valores en el eje x y y\n",
        "    x_values = (range(1, len(freq_list) + 1))\n",
        "    y_values = np.log(freq_list) + 1\n",
        "\n",
        "    # Seleccionar un marcador diferente para cada curva\n",
        "    marcadores = ['o', 's', '^', 'v', 'D', 'p', 'h', '+', 'x', '<', '>', '*', '|', '_', '.', ',']\n",
        "\n",
        "    # Crear un gráfico log-log con líneas y asignar un color y marcador a cada curva\n",
        "    plt.plot(\n",
        "        x_values,\n",
        "        y_values,\n",
        "        label=nombres_especies[i],\n",
        "        marker=marcadores[i % len(marcadores)],\n",
        "        markersize=6,\n",
        "        linestyle='-',\n",
        "        color=colores[i],\n",
        "        alpha=0.7\n",
        "    )\n",
        "\n",
        "# Configurar etiquetas y título\n",
        "plt.ylabel('ln(Frequency)')\n",
        "plt.xlabel('ln(Range)')\n",
        "plt.title(r'General graph for Semi-log of the $\\mathit{PRKN}$ gene of 16 primates')\n",
        "# Mostrar leyenda\n",
        "plt.legend()\n",
        "\n",
        "# Guardar la imagen en la ruta especificada\n",
        "nombre_archivo = \"Semi-Log Frequency Histogram of size 7-mers.png\"\n",
        "ruta_completa = os.path.join(ruta_guardado, nombre_archivo)\n",
        "plt.savefig(ruta_completa)\n",
        "\n",
        "# Mostrar el gráfico\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "4mZR64TnIslj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Lista de tus conjuntos de datos\n",
        "conjuntos_de_datos = [d1_7, d2_7, d3_7, d4_7, d5_7, d6_7, d7_7, d8_7,\n",
        "                      d9_7, d10_7, d11_7, d12_7, d13_7, d14_7, d15_7, d16_7]\n",
        "\n",
        "\n",
        "nombres_especies = [\n",
        "    r'$\\mathit{H. sapiens}$',\n",
        "    r'$\\mathit{P. troglodytes}$',\n",
        "    r'$\\mathit{P. paniscus}$',\n",
        "    r'$\\mathit{G. gorilla}$',\n",
        "    r'$\\mathit{P. abelii}$',\n",
        "    r'$\\mathit{P. pygmaeus}$',\n",
        "    r'$\\mathit{N. leucogenys}$',\n",
        "    r'$\\mathit{S. syndactylus}$',\n",
        "    r'$\\mathit{M. mulatta}$',\n",
        "    r'$\\mathit{M.a fascicularis}$',\n",
        "    r'$\\mathit{M. thibetana thibetana}$',\n",
        "    r'$\\mathit{T. francoisi}$',\n",
        "    r'$\\mathit{R. roxellana}$',\n",
        "    r'$\\mathit{M. murinus}$',\n",
        "    r'$\\mathit{C. jacchus}$',\n",
        "    r'$\\mathit{L. catta}$'\n",
        "]\n",
        "\n",
        "\n",
        "# Lista de nombres de especies correspondientes a cada conjunto de datos\n",
        "etiquetas_especies = [\n",
        "  'H. sapiens',\n",
        "  'P. troglodytes',\n",
        "  'P. paniscus',\n",
        "  'G. gorilla',\n",
        "  'P. abelii',\n",
        "  'P. pygmaeus',\n",
        "  'N. leucogenys',\n",
        "  'S. syndactylus',\n",
        "  'M. mulatta',\n",
        "  'M. fascicularis',\n",
        "  'M. thibetana',\n",
        "  'T. francoisi',\n",
        "  'R. roxellana',\n",
        "  'M. murinus',\n",
        "  'C. jacchus',\n",
        "  'L. catta'\n",
        "\n",
        "]\n",
        "\n",
        "# Ruta de la carpeta donde se guardarán las imágenes\n",
        "ruta_guardado = \"/content/drive/MyDrive/TESIS_FINAL/POWER LAW/SEMI-LOG/\"\n",
        "\n",
        "# Para cada conjunto de datos, genera y guarda el gráfico\n",
        "for i, datos in enumerate(conjuntos_de_datos, start=1):\n",
        "    # Crear una nueva figura con el tamaño deseado\n",
        "    plt.figure(figsize=(20, 11))\n",
        "    # Calcular frecuencias y ordenar por frecuencia\n",
        "    freq_list = sorted(list(datos.values()), reverse=True)\n",
        "\n",
        "    # Crear un histograma log-log con puntos personalizados\n",
        "    plt.scatter(\n",
        "        (range(1, len(freq_list) + 1)),\n",
        "        np.log(freq_list),\n",
        "        label=f'{nombres_especies[i-1]}',  # Usar el nombre de la especie en lugar de \"Datos\"\n",
        "        s=2,        # Tamaño de los puntos\n",
        "        linestyle='-',\n",
        "        c='blue',    # Color de los puntos\n",
        "        alpha=0.7     # Transparencia\n",
        "    )\n",
        "\n",
        "\n",
        "    # Configurar etiquetas y título\n",
        "    plt.xlabel('ln(Range)')\n",
        "    plt.ylabel('ln(Frequency)')\n",
        "    plt.title(f'Semi-Log for {nombres_especies[i-1]} of size 7-mers')\n",
        "\n",
        "    # Mostrar leyenda\n",
        "    plt.legend()\n",
        "\n",
        "    # Guardar el gráfico con el nombre deseado\n",
        "    nombre_archivo = etiquetas_especies [i-1].replace(\" \", \"_\").lower() + \"_k7.png\"\n",
        "    ruta_completa = ruta_guardado + nombre_archivo\n",
        "    plt.savefig(ruta_completa)\n",
        "\n",
        "    # Mostrar el gráfico\n",
        "    plt.show()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "print(\"Imágenes guardadas exitosamente en:\", ruta_guardado)"
      ],
      "metadata": {
        "id": "BNrGz40xI67i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **ENTROPIA DE SHANNON**"
      ],
      "metadata": {
        "id": "nRwfpjjVLocA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def entropy(sequence):\n",
        "  # Calcula la distribución de probabilidad de los caracteres en la secuencia.\n",
        "  counts = collections.Counter(sequence)\n",
        "  probabilities = {char: count / len(sequence) for char, count in counts.items()}\n",
        "\n",
        "  # Calcula la entropía de Shannon.\n",
        "  entropy = 0\n",
        "  for probability in probabilities.values():\n",
        "    if probability > 0:\n",
        "      entropy -= probability * math.log2(probability)\n",
        "\n",
        "  return entropy"
      ],
      "metadata": {
        "id": "AvgHtVgGL217"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def entropy(sequence):\n",
        "    # Calcula la distribución de probabilidad de los caracteres en la secuencia.\n",
        "    counts = collections.Counter(sequence)\n",
        "    probabilities = {char: count / len(sequence) for char, count in counts.items()}\n",
        "\n",
        "    # Calcula la entropía de Shannon.\n",
        "    entropy = 0\n",
        "    for probability in probabilities.values():\n",
        "        if probability > 0:\n",
        "            entropy -= probability * math.log2(probability)\n",
        "\n",
        "    return round(entropy, 5)  # Reducir el resultado a 4 decimales\n"
      ],
      "metadata": {
        "id": "oTrF8kgeNAzp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "entropy1 = entropy(DATOS1)\n",
        "entropy2 = entropy(DATOS2)\n",
        "entropy3 = entropy(DATOS3)\n",
        "entropy4 = entropy(DATOS4)\n",
        "entropy5 = entropy(DATOS5)\n",
        "entropy6 = entropy(DATOS6)\n",
        "entropy7 = entropy(DATOS7)\n",
        "entropy8 = entropy(DATOS8)\n",
        "entropy9 = entropy(DATOS9)\n",
        "entropy10 = entropy(DATOS10)\n",
        "entropy11 = entropy(DATOS11)\n",
        "entropy12 = entropy(DATOS12)\n",
        "entropy13 = entropy(DATOS13)\n",
        "entropy14 = entropy(DATOS14)\n",
        "entropy15 = entropy(DATOS15)\n",
        "entropy16 = entropy(DATOS16)"
      ],
      "metadata": {
        "id": "P6k97UwsL9Dj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(entropy1)\n",
        "print(entropy2)\n",
        "print(entropy3)\n",
        "print(entropy4)\n",
        "print(entropy5)\n",
        "print(entropy6)\n",
        "print(entropy7)\n",
        "print(entropy8)\n",
        "print(entropy9)\n",
        "print(entropy10)\n",
        "print(entropy11)\n",
        "print(entropy12)\n",
        "print(entropy13)\n",
        "print(entropy14)\n",
        "print(entropy15)\n",
        "print(entropy16)\n"
      ],
      "metadata": {
        "id": "5I3P7MZoMAnQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Ruta deseada\n",
        "ruta_deseada = '/content/drive/MyDrive/TESIS_FINAL/Suplementary/03 Supplementary material/'\n",
        "\n",
        "# Datos de entropía\n",
        "entropias = {\n",
        "    'H. sapiens': entropy1,\n",
        "    'G. gorilla': entropy2,\n",
        "    'P. paniscus': entropy3,\n",
        "    'P. troglodytes': entropy4,\n",
        "    'P. abeli': entropy5,\n",
        "    'P. pygmaeus': entropy6,\n",
        "    'M. fascicularis': entropy7,\n",
        "    'M. mulatta': entropy8,\n",
        "    'M. thibetana': entropy9,\n",
        "    'R. roxellana': entropy10,\n",
        "    'T. francoisi': entropy11,\n",
        "    'N. leucogenys': entropy12,\n",
        "    'S. syndactylus': entropy13,\n",
        "    'M. murinus': entropy14,\n",
        "    'C. jacchus': entropy15,\n",
        "    'L. catta': entropy16,\n",
        "}\n",
        "\n",
        "# Crear un DataFrame de pandas\n",
        "df = pd.DataFrame(list(entropias.items()), columns=['Primate species', 'Shannon entropy'])\n",
        "\n",
        "# Ordenar el DataFrame por entropía de menor a mayor\n",
        "df = df.sort_values(by='Shannon entropy')\n",
        "\n",
        "# Ruta completa para guardar el archivo Excel\n",
        "excel_ruta = os.path.join(ruta_deseada, 'Shannon_entropy_Primate_species.xlsx')\n",
        "\n",
        "# Guardar el DataFrame en un archivo Excel con nombres en negrita y cursiva\n",
        "with pd.ExcelWriter(excel_ruta, engine='xlsxwriter') as writer:\n",
        "    df.style.applymap(lambda x: 'font-weight: bold; font-style: italic', subset=['Primate species']).to_excel(writer, index=False, sheet_name='Entropías')\n",
        "\n",
        "# Mostrar el mensaje de éxito\n",
        "print(f\"Archivo Excel guardado en: {excel_ruta}\")"
      ],
      "metadata": {
        "id": "SM44zU5EME3G"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Define la ruta de destino\n",
        "ruta_destino = '/content/drive/MyDrive/TESIS_FINAL/Suplementary/03 Supplementary material/'\n",
        "\n",
        "# Define las entropías\n",
        "entropias = {\n",
        "    'H. sapiens': entropy1,\n",
        "    'G. gorilla': entropy2,\n",
        "    'P. paniscus': entropy3,\n",
        "    'P. troglodytes': entropy4,\n",
        "    'P. abeli': entropy5,\n",
        "    'P. pygmaeus': entropy6,\n",
        "    'M. fascicularis': entropy7,\n",
        "    'M. mulatta': entropy8,\n",
        "    'M. thibetana': entropy9,\n",
        "    'R. roxellana': entropy10,\n",
        "    'T. francoisi': entropy11,\n",
        "    'N. leucogenys': entropy12,\n",
        "    'S. syndactylus': entropy13,\n",
        "    'M. murinus': entropy14,\n",
        "    'C. jacchus': entropy15,\n",
        "    'L. catta': entropy16,\n",
        "}\n",
        "\n",
        "# Ordena las entropías de menor a mayor\n",
        "entropias_ordenadas = sorted(entropias.items(), key=lambda x: x[1])\n",
        "\n",
        "# Extrae los nombres de las especies y las entropías ordenadas\n",
        "especies_ordenadas, entropias_ordenadas = zip(*entropias_ordenadas)\n",
        "\n",
        "# Calcula el promedio y la desviación estándar\n",
        "promedio = np.mean(entropias_ordenadas)\n",
        "desviacion_estandar = np.std(entropias_ordenadas)\n",
        "\n",
        "\n",
        "# Grafica las entropías\n",
        "plt.figure(figsize=(13, 9))\n",
        "plt.plot(especies_ordenadas, entropias_ordenadas, marker='o', color='blue', linestyle='-')\n",
        "\n",
        "# Agregar línea para el promedio\n",
        "plt.axhline(y=promedio, color='red', linestyle='--', label=f'Average: {promedio:.5f}')\n",
        "\n",
        "# Agregar líneas para la desviación estándar\n",
        "plt.axhline(y=promedio + desviacion_estandar, color='green', linestyle='--', label=f'Average + Standard deviation: {promedio + desviacion_estandar:.5f}')\n",
        "plt.axhline(y=promedio - desviacion_estandar, color='green', linestyle='--', label=f'Average - Standard deviation: {promedio - desviacion_estandar:.5f}')\n",
        "\n",
        "# Etiqueta los puntos con los valores de las entropías\n",
        "for i, txt in enumerate(entropias_ordenadas):\n",
        "    plt.annotate(f'{txt:.5f}', (i, entropias_ordenadas[i]), textcoords=\"offset points\", xytext=(-4,-4), ha='left', va='top', rotation=340, fontsize=8)\n",
        "\n",
        "# Configura el eje x con los nombres de las especies\n",
        "plt.xticks(range(len(especies_ordenadas)), especies_ordenadas, rotation=45, ha='right')\n",
        "plt.ylabel('Shannon Entropy', fontweight='bold')\n",
        "plt.title(r'Shannon Entropy for Different $\\mathit{PRKN}$ Gene from Primate Species')\n",
        "\n",
        "# Añade una leyenda\n",
        "plt.legend()\n",
        "\n",
        "# Guarda la gráfica en la ruta de destino\n",
        "if not os.path.exists(ruta_destino):\n",
        "    os.makedirs(ruta_destino)\n",
        "plt.savefig(os.path.join(ruta_destino, 'Entropy_primate_species.png'))\n",
        "\n",
        "# Muestra la gráfica\n",
        "plt.tight_layout()\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "Jv73UUojc8P6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **CGR JEFFREY**"
      ],
      "metadata": {
        "id": "pXI3GeIWPmQE"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KmiDX3Xp2fp_"
      },
      "source": [
        "## **SE DEFINE LA FUNCION  GENERACION DE PUNTOS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Fdqcg0-12szH"
      },
      "outputs": [],
      "source": [
        "def generate_points(dna_sequence):\n",
        "    # Define las esquinas del cuadrado\n",
        "    corners = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "\n",
        "    # Punto inicial\n",
        "    current_point = (0.5, 0.5)\n",
        "\n",
        "    # Guarda las coordenadas de los puntos para graficar\n",
        "    x_values = [current_point[0]]\n",
        "    y_values = [current_point[1]]\n",
        "\n",
        "    # Itera sobre la secuencia de ADN\n",
        "    for nucleotide in dna_sequence:\n",
        "        # Obtiene la coordenada de la esquina correspondiente al nucleótido actual\n",
        "        target_corner = corners[nucleotide]\n",
        "\n",
        "        # Calcula el punto a la mitad de la distancia entre el punto actual y la esquina objetivo\n",
        "        current_point = ((current_point[0] + target_corner[0]) / 2, (current_point[1] + target_corner[1]) / 2)\n",
        "\n",
        "        # Guarda las coordenadas del nuevo punto\n",
        "        x_values.append(current_point[0])\n",
        "        y_values.append(current_point[1])\n",
        "\n",
        "    return x_values, y_values"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def normalize_scale(data, reference_scale):\n",
        "    min_data = min(data)\n",
        "    max_data = max(data)\n",
        "    normalized_data = [(x - min_data) / (max_data - min_data) * (reference_scale[1] - reference_scale[0]) + reference_scale[0] for x in data]\n",
        "    return normalized_data"
      ],
      "metadata": {
        "id": "xs0lyTN71r8O"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "reference_scale = (0, 100)"
      ],
      "metadata": {
        "id": "LmOggOTP6BQU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sDnwcW7S5Ugu"
      },
      "source": [
        "## **SE DEFINE EL COLOR**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YAenZoY290ua"
      },
      "outputs": [],
      "source": [
        "color_general = cm.tab20b_r\n",
        "ruta_salida = '/content/drive/MyDrive/TESIS_FINAL/Suplementary/04  Supplementary material/'\n",
        "#RESOLUCION\n",
        "R = 500\n",
        "RT = 500\n",
        "#TAMAÑO DE FIGURA\n",
        "T = (7, 6)\n",
        "#TAMAÑO DE PUNTO\n",
        "P= 0.1\n",
        "#MARKADOR\n",
        "M = '.'"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "NAME1 = r'$\\mathit{H. sapiens \\ PRKN}$'\n",
        "NAME2 = r'$\\mathit{G. gorilla \\ PRKN}$'\n",
        "NAME3 = r'$\\mathit{P. paniscus \\ PRKN}$'\n",
        "NAME4 = r'$\\mathit{P. troglodytes \\ PRKN}$'\n",
        "NAME5 = r'$\\mathit{P. abelii \\ PRKN}$'\n",
        "NAME6 = r'$\\mathit{P. pygmaeus \\ PRKN}$'\n",
        "NAME7 = r'$\\mathit{M. fascicularis \\ PRKN}$'\n",
        "NAME8 = r'$\\mathit{M. mulatta \\ PRKN}$'\n",
        "NAME9 = r'$\\mathit{M. thibetana \\ PRKN}$'\n",
        "NAME10 = r'$\\mathit{R. roxellana \\ PRKN}$'\n",
        "NAME11 = r'$\\mathit{T. francoisi \\ PRKN}$'\n",
        "NAME12 = r'$\\mathit{N. leucogenys \\ PRKN}$'\n",
        "NAME13 = r'$\\mathit{S. syndactylus \\ PRKN}$'\n",
        "NAME14 = r'$\\mathit{M. murinus \\ PRKN}$'\n",
        "NAME15 = r'$\\mathit{C. jacchus \\ PRKN}$'\n",
        "NAME16 = r'$\\mathit{L. catta \\ PRKN}$'\n",
        "\n",
        "NAME_1 = 'H. sapiens'\n",
        "NAME_2 = 'G. gorilla'\n",
        "NAME_3 = 'P. paniscus'\n",
        "NAME_4 = 'P. troglodytes'\n",
        "NAME_5 = 'P. abelii'\n",
        "NAME_6 = 'P. pygmaeus'\n",
        "NAME_7 = 'M. fascicularis'\n",
        "NAME_8 = 'M. mulatta'\n",
        "NAME_9 = 'M. thibetana'\n",
        "NAME_10 = 'R. roxellana'\n",
        "NAME_11 = 'T. francoisi'\n",
        "NAME_12 = 'N. leucogenys'\n",
        "NAME_13 = 'S. syndactylus'\n",
        "NAME_14 = 'M. murinus'\n",
        "NAME_15 = 'C. jacchus'\n",
        "NAME_16 = 'L. catta'\n"
      ],
      "metadata": {
        "id": "fYuCTbs3fICT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RQT-mRi220n0"
      },
      "source": [
        "## **SE GENERAN LA IMAGENES FRACTALES**"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "gc.collect()"
      ],
      "metadata": {
        "id": "_2K2v0ATfL5w"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_bxvPAbF3NV3"
      },
      "source": [
        "## **01 H. sapiens**"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "x1, y1 = generate_points(DATOS1)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "reference_scale = (0, 100)\n",
        "normalized_indices = normalize_scale(range(len(x1)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x1, y1, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x1, y1) in corners_labels.items():\n",
        "    plt.text(x1, y1, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR of {NAME1}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "\n",
        "\n",
        "plt.savefig(os.path.join(ruta_salida, f\"01 CGR_of_{NAME_1}.png\"), dpi=RT, bbox_inches='tight')\n",
        "\n",
        "plt.show()\n",
        "gc.collect()"
      ],
      "metadata": {
        "id": "oxM8fZd0E-9i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xySAV8fN3NNf"
      },
      "source": [
        "## **02 G. gorilla**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "VxdFGprb7sS1"
      },
      "outputs": [],
      "source": [
        "x2, y2 = generate_points(DATOS2)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "reference_scale = (0, 100)\n",
        "normalized_indices = normalize_scale(range(len(x2)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x2, y2, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x2, y2) in corners_labels.items():\n",
        "    plt.text(x2, y2, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR of {NAME2}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "\n",
        "\n",
        "plt.savefig(os.path.join(ruta_salida, f\"02 CGR_of_{NAME_2}.png\"), dpi=RT, bbox_inches='tight')\n",
        "\n",
        "plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TjxKWllU3ND7"
      },
      "source": [
        "## **'03 P. paniscus**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-sQGa3mU8Vxl"
      },
      "outputs": [],
      "source": [
        "x3, y3 = generate_points(DATOS3)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x3)), reference_scale)\n",
        "\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x3, y3, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x3, y3) in corners_labels.items():\n",
        "    plt.text(x3, y3, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME3}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "\n",
        "plt.savefig(os.path.join(ruta_salida, f\"03 CGR_of_{NAME_3}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yKTyh8lu3M7W"
      },
      "source": [
        "## **04 P. troglodytes**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Zcx15LIr8xvb"
      },
      "outputs": [],
      "source": [
        "x4, y4 = generate_points(DATOS4)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x4)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x4, y4, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x4, y4) in corners_labels.items():\n",
        "    plt.text(x4, y4, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME4}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"04 CGR_of_{NAME_4}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZpXcv4BQ3MzP"
      },
      "source": [
        "## **05 P. abelii**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "nvddRP6Y86hk"
      },
      "outputs": [],
      "source": [
        "x5, y5 = generate_points(DATOS5)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x5)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x5, y5, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x5, y5) in corners_labels.items():\n",
        "    plt.text(x5, y5, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME5}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"05 CGR_of_{NAME_5}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show())\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wC2ttsM13Mqi"
      },
      "source": [
        "## **06 P. pygmaeus**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ey7XxUu09iW_"
      },
      "outputs": [],
      "source": [
        "x6, y6 = generate_points(DATOS6)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x6)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x6, y6, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x6, y6) in corners_labels.items():\n",
        "    plt.text(x6, y6, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME6}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"06 CGR_of_{NAME_6}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z80xI4q63Mib"
      },
      "source": [
        "## **07 M. fascicularis**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3A3Irbmw99Jp"
      },
      "outputs": [],
      "source": [
        "x7, y7 = generate_points(DATOS7)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x7)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x7, y7, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x7, y7) in corners_labels.items():\n",
        "    plt.text(x7, y7, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME7}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"07 CGR_of_{NAME_7}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6gDtRrad3Mar"
      },
      "source": [
        "## **08 M. mulatta**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AnZL9msv_O3Q"
      },
      "outputs": [],
      "source": [
        "x8, y8 = generate_points(DATOS8)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x8)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x8, y8, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x8, y8) in corners_labels.items():\n",
        "    plt.text(x8, y8, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME8}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "\n",
        "plt.savefig(os.path.join(ruta_salida, f\"08 CGR_of_{NAME_8}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6aoSD8ax3MRg"
      },
      "source": [
        "## **09 M. thibetana**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ON8CQ_Ot_gue"
      },
      "outputs": [],
      "source": [
        "x9, y9 = generate_points(DATOS9)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x9)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x9, y9, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x9, y9) in corners_labels.items():\n",
        "    plt.text(x9, y9, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME9}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"09 CGR_of_{NAME_9}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GkbLOonx3MJX"
      },
      "source": [
        "## **10 R. roxellana**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "FgZcN9k9_0wQ"
      },
      "outputs": [],
      "source": [
        "x10, y10 = generate_points(DATOS10)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x10)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x10, y10, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x10, y10) in corners_labels.items():\n",
        "    plt.text(x10, y10, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME10}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"10 CGR_of_{NAME_10}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "r9fRlHNG3L-1"
      },
      "source": [
        "## **11 T. francoisi**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "GrLx9d--__Ng"
      },
      "outputs": [],
      "source": [
        "x11, y11 = generate_points(DATOS11)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x11)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x11, y11, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x11, y11) in corners_labels.items():\n",
        "    plt.text(x11, y11, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME11}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"11 CGR_of_{NAME_11}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8QrfK_kB3Lrj"
      },
      "source": [
        "## **12 N. leucogenys**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "FZHAxLVkADPU"
      },
      "outputs": [],
      "source": [
        "x12, y12 = generate_points(DATOS12)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x12)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x12, y12, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x12, y12) in corners_labels.items():\n",
        "    plt.text(x12, y12, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME12}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"12 CGR_of_{NAME_12}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GxPPPN2H3LgR"
      },
      "source": [
        "## **S. syndactylus**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "9NFij2q7ASP6"
      },
      "outputs": [],
      "source": [
        "x13, y13 = generate_points(DATOS13)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x13)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x13, y13, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x13, y13) in corners_labels.items():\n",
        "    plt.text(x13, y13, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME13}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"13 CGR_of_{NAME_13}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cNE1oE2rxQyX"
      },
      "source": [
        "## **M. murinus**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6b-PsJzKxw5G"
      },
      "outputs": [],
      "source": [
        "x14, y14 = generate_points(DATOS14)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x14)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x14, y14, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x14, y14) in corners_labels.items():\n",
        "    plt.text(x14, y14, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME14}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"14 CGR_of_{NAME_14}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jRkLEVJfyKod"
      },
      "source": [
        "##**15 C. jacchus**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JtzB0M-Vxxh8"
      },
      "outputs": [],
      "source": [
        "x15, y15 = generate_points(DATOS13)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x15)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x15, y15, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x15, y15) in corners_labels.items():\n",
        "    plt.text(x15, y15, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME15}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"15 CGR_of_{NAME_15}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rvhAkN7hyP_H"
      },
      "source": [
        "## **16 L. catta**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4-Hy8rzyxPcs"
      },
      "outputs": [],
      "source": [
        "x16, y16 = generate_points(DATOS16)\n",
        "# Normaliza los índices de punto utilizando la escala de referencia\n",
        "normalized_indices = normalize_scale(range(len(x16)), reference_scale)\n",
        "# Define las esquinas con letras\n",
        "corners_labels = {'G': (0, 100), 'A': (100, 100), 'T': (0, 0), 'C': (100, 0)}\n",
        "# Grafica solo los puntos sin líneas, con tamaño reducido\n",
        "fig, ax = plt.subplots(figsize= T, dpi=R)\n",
        "scatter_plot = plt.scatter(x16, y16, c=normalized_indices, cmap= color_general , s= P, marker= M)    # Utiliza 'c' para especificar el color y 'cmap' para el mapa de colores\n",
        "\n",
        "\n",
        "# Muestra las esquinas con letras\n",
        "for nucleotide, (x16, y16) in corners_labels.items():\n",
        "    plt.text(x16, y16, nucleotide, fontsize=5, ha='center', va='center', color='black')\n",
        "\n",
        "\n",
        "plt.title(f\"CGR OF {NAME16}\")\n",
        "plt.xlabel(\"\")\n",
        "plt.ylabel(\"\")\n",
        "\n",
        "# Agrega la barra de color\n",
        "colorbar = plt.colorbar(scatter_plot, label='Point index')\n",
        "# Agrega la barra de color\n",
        "plt.savefig(os.path.join(ruta_salida, f\"16 CGR_of_{NAME_16}.png\"), dpi=RT, bbox_inches='tight')\n",
        "#plt.show()\n",
        "gc.collect()"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [
        "-2Nss7qjGgyN",
        "p6B6HOu2GmPn",
        "vYEzBB8cHnk0",
        "Z1Q0DtUTPdeB",
        "_zpb3pwtHne6",
        "CugJV9mZHtw8",
        "VkK5d5A8HyxM",
        "vJdXW1ggKa_3",
        "GGGKpv4dH2bZ",
        "4IDiFGOqIQpg",
        "WedrUmZpOzZl",
        "3i3GwnclX6Dd",
        "0O6vkNcGXVKN",
        "D6n6nd4bBXGj",
        "628WHvRz_vPe",
        "A121eF6RLPZt",
        "F6jkzoUINXU4",
        "lregAz5GKztF",
        "roY9jbf1K3rV",
        "PqC3NaYrK7nA",
        "aO90-chlgChg",
        "mrqEE3dBHDPK",
        "4QDTwIy5Iqii",
        "nRwfpjjVLocA",
        "pXI3GeIWPmQE",
        "KmiDX3Xp2fp_",
        "sDnwcW7S5Ugu",
        "RQT-mRi220n0",
        "_bxvPAbF3NV3",
        "xySAV8fN3NNf",
        "TjxKWllU3ND7",
        "yKTyh8lu3M7W",
        "ZpXcv4BQ3MzP",
        "wC2ttsM13Mqi",
        "Z80xI4q63Mib",
        "6gDtRrad3Mar",
        "6aoSD8ax3MRg",
        "GkbLOonx3MJX",
        "r9fRlHNG3L-1",
        "8QrfK_kB3Lrj",
        "GxPPPN2H3LgR",
        "cNE1oE2rxQyX",
        "jRkLEVJfyKod",
        "rvhAkN7hyP_H"
      ],
      "provenance": [],
      "authorship_tag": "ABX9TyPP305fIgVgjf3p3Nrb9hAu"
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}